// @aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-target-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codedeploy = require("@aws-cdk/aws-codedeploy-api");
const cdk = require("@aws-cdk/cdk");
const elasticloadbalancingv2_generated_1 = require("../elasticloadbalancingv2.generated");
const util_1 = require("./util");
/**
 * Define the target of a load balancer
 */
class BaseTargetGroup extends cdk.Construct {
    constructor(parent, id, baseProps, additionalProps) {
        super(parent, id);
        /**
         * List of dependables that need to be depended on to ensure the TargetGroup is associated to a load balancer
         */
        this.loadBalancerAssociationDependencies = new Array();
        /**
         * Attributes of this target group
         */
        this.attributes = {};
        /**
         * The JSON objects returned by the directly registered members of this target group
         */
        this.targetsJson = new Array();
        if (baseProps.deregistrationDelaySec !== undefined) {
            this.setAttribute('deregistration_delay.timeout_seconds', baseProps.deregistrationDelaySec.toString());
        }
        this.healthCheck = baseProps.healthCheck || {};
        this.targetType = baseProps.targetType;
        this.resource = new elasticloadbalancingv2_generated_1.cloudformation.TargetGroupResource(this, 'Resource', Object.assign({ targetGroupName: baseProps.targetGroupName, targetGroupAttributes: new cdk.Token(() => util_1.renderAttributes(this.attributes)), targetType: new cdk.Token(() => this.targetType), targets: new cdk.Token(() => this.targetsJson), vpcId: baseProps.vpc.vpcId, 
            // HEALTH CHECK
            healthCheckIntervalSeconds: new cdk.Token(() => this.healthCheck && this.healthCheck.intervalSecs), healthCheckPath: new cdk.Token(() => this.healthCheck && this.healthCheck.path), healthCheckPort: new cdk.Token(() => this.healthCheck && this.healthCheck.port), healthCheckProtocol: new cdk.Token(() => this.healthCheck && this.healthCheck.protocol), healthCheckTimeoutSeconds: new cdk.Token(() => this.healthCheck && this.healthCheck.timeoutSeconds), healthyThresholdCount: new cdk.Token(() => this.healthCheck && this.healthCheck.healthyThresholdCount), unhealthyThresholdCount: new cdk.Token(() => this.healthCheck && this.healthCheck.unhealthyThresholdCount), matcher: new cdk.Token(() => this.healthCheck && this.healthCheck.healthyHttpCodes !== undefined ? {
                httpCode: this.healthCheck.healthyHttpCodes
            } : undefined) }, additionalProps));
        this.targetGroupLoadBalancerArns = this.resource.targetGroupLoadBalancerArns.toList();
        this.targetGroupArn = this.resource.ref;
        this.targetGroupFullName = this.resource.targetGroupFullName;
        this.loadBalancerArns = this.resource.targetGroupLoadBalancerArns.toString();
        this.targetGroupName = this.resource.targetGroupName;
        this.defaultPort = `${additionalProps.port}`;
    }
    /**
     * Set/replace the target group's health check
     */
    configureHealthCheck(healthCheck) {
        this.healthCheck = healthCheck;
    }
    /**
     * Set a non-standard attribute on the target group
     *
     * @see https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html#target-group-attributes
     */
    setAttribute(key, value) {
        this.attributes[key] = value;
    }
    /**
     * Export this target group
     */
    export() {
        return {
            targetGroupArn: new cdk.Output(this, 'TargetGroupArn', { value: this.targetGroupArn }).makeImportValue().toString(),
            defaultPort: new cdk.Output(this, 'Port', { value: this.defaultPort }).makeImportValue().toString(),
        };
    }
    /**
     * Add a dependency between this target group and the indicated resources
     */
    addDependency(...other) {
        this.resource.addDependency(...other);
    }
    asCodeDeployLoadBalancer() {
        return {
            generation: codedeploy.LoadBalancerGeneration.Second,
            name: this.targetGroupName,
        };
    }
    /**
     * Return an object to depend on this TargetGroup being attached to a load balancer
     */
    loadBalancerDependency() {
        return new util_1.LazyDependable(this.loadBalancerAssociationDependencies);
    }
    /**
     * Register the given load balancing target as part of this group
     */
    addLoadBalancerTarget(props) {
        if (this.targetType !== undefined && this.targetType !== props.targetType) {
            throw new Error(`Already have a of type '${this.targetType}', adding '${props.targetType}'; make all targets the same type.`);
        }
        this.targetType = props.targetType;
        if (props.targetJson) {
            this.targetsJson.push(props.targetJson);
        }
    }
}
exports.BaseTargetGroup = BaseTargetGroup;
/**
 * Extract the full load balancer name (used for metrics) from the listener ARN:
 *
 * Turns
 *
 *     arn:aws:elasticloadbalancing:us-west-2:123456789012:listener/app/my-load-balancer/50dc6c495c0c9188/f2f7dc8efc522ab2
 *
 * Into
 *
 *     app/my-load-balancer/50dc6c495c0c9188
 */
function loadBalancerNameFromListenerArn(listenerArn) {
    const arnParts = new cdk.FnSplit('/', listenerArn);
    return `${new cdk.FnSelect(1, arnParts)}/${new cdk.FnSelect(2, arnParts)}/${new cdk.FnSelect(3, arnParts)}`;
}
exports.loadBalancerNameFromListenerArn = loadBalancerNameFromListenerArn;
