// @aws-cdk/aws-elasticloadbalancingv2/lib/elasticloadbalancingv2.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:04:41.350Z","fingerprint":"/5B7yIrSC1k3GiF3eHAVQiAO5oXKukx7NM09d0ckeLY="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ListenerResourceProps``
     *
     * @param properties the TypeScript properties of a ``ListenerResourceProps``
     *
     * @returns the result of the validation.
     */
    function ListenerResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('certificates', cdk.listValidator(ListenerResource_CertificatePropertyValidator))(properties.certificates));
        errors.collect(cdk.propertyValidator('defaultActions', cdk.requiredValidator)(properties.defaultActions));
        errors.collect(cdk.propertyValidator('defaultActions', cdk.listValidator(ListenerResource_ActionPropertyValidator))(properties.defaultActions));
        errors.collect(cdk.propertyValidator('loadBalancerArn', cdk.requiredValidator)(properties.loadBalancerArn));
        errors.collect(cdk.propertyValidator('loadBalancerArn', cdk.validateString)(properties.loadBalancerArn));
        errors.collect(cdk.propertyValidator('port', cdk.requiredValidator)(properties.port));
        errors.collect(cdk.propertyValidator('port', cdk.validateNumber)(properties.port));
        errors.collect(cdk.propertyValidator('protocol', cdk.requiredValidator)(properties.protocol));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        errors.collect(cdk.propertyValidator('sslPolicy', cdk.validateString)(properties.sslPolicy));
        return errors.wrap('supplied properties not correct for "ListenerResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener`` resource
     *
     * @param properties the TypeScript properties of a ``ListenerResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener`` resource.
     */
    function listenerResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResourcePropsValidator(properties).assertSuccess();
        return {
            DefaultActions: cdk.listMapper(listenerResourceActionPropertyToCloudFormation)(properties.defaultActions),
            LoadBalancerArn: cdk.stringToCloudFormation(properties.loadBalancerArn),
            Port: cdk.numberToCloudFormation(properties.port),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
            Certificates: cdk.listMapper(listenerResourceCertificatePropertyToCloudFormation)(properties.certificates),
            SslPolicy: cdk.stringToCloudFormation(properties.sslPolicy),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listener.html
     */
    class ListenerResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancingV2::Listener``.
         *
         * @param parent   the ``cdk.Construct`` this ``ListenerResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ListenerResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ListenerResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'defaultActions', this);
            cdk.requireProperty(properties, 'loadBalancerArn', this);
            cdk.requireProperty(properties, 'port', this);
            cdk.requireProperty(properties, 'protocol', this);
            this.listenerArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return listenerResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ListenerResource.resourceTypeName = "AWS::ElasticLoadBalancingV2::Listener";
    cloudformation.ListenerResource = ListenerResource;
    /**
     * Determine whether the given properties match those of a ``ActionProperty``
     *
     * @param properties the TypeScript properties of a ``ActionProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_ActionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticateCognitoConfig', ListenerResource_AuthenticateCognitoConfigPropertyValidator)(properties.authenticateCognitoConfig));
        errors.collect(cdk.propertyValidator('authenticateOidcConfig', ListenerResource_AuthenticateOidcConfigPropertyValidator)(properties.authenticateOidcConfig));
        errors.collect(cdk.propertyValidator('fixedResponseConfig', ListenerResource_FixedResponseConfigPropertyValidator)(properties.fixedResponseConfig));
        errors.collect(cdk.propertyValidator('order', cdk.validateNumber)(properties.order));
        errors.collect(cdk.propertyValidator('redirectConfig', ListenerResource_RedirectConfigPropertyValidator)(properties.redirectConfig));
        errors.collect(cdk.propertyValidator('targetGroupArn', cdk.validateString)(properties.targetGroupArn));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "ActionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.Action`` resource
     *
     * @param properties the TypeScript properties of a ``ActionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.Action`` resource.
     */
    function listenerResourceActionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_ActionPropertyValidator(properties).assertSuccess();
        return {
            AuthenticateCognitoConfig: listenerResourceAuthenticateCognitoConfigPropertyToCloudFormation(properties.authenticateCognitoConfig),
            AuthenticateOidcConfig: listenerResourceAuthenticateOidcConfigPropertyToCloudFormation(properties.authenticateOidcConfig),
            FixedResponseConfig: listenerResourceFixedResponseConfigPropertyToCloudFormation(properties.fixedResponseConfig),
            Order: cdk.numberToCloudFormation(properties.order),
            RedirectConfig: listenerResourceRedirectConfigPropertyToCloudFormation(properties.redirectConfig),
            TargetGroupArn: cdk.stringToCloudFormation(properties.targetGroupArn),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``AuthenticateCognitoConfigProperty``
     *
     * @param properties the TypeScript properties of a ``AuthenticateCognitoConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_AuthenticateCognitoConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticationRequestExtraParams', cdk.hashValidator(cdk.validateString))(properties.authenticationRequestExtraParams));
        errors.collect(cdk.propertyValidator('onUnauthenticatedRequest', cdk.validateString)(properties.onUnauthenticatedRequest));
        errors.collect(cdk.propertyValidator('scope', cdk.validateString)(properties.scope));
        errors.collect(cdk.propertyValidator('sessionCookieName', cdk.validateString)(properties.sessionCookieName));
        errors.collect(cdk.propertyValidator('sessionTimeout', cdk.validateNumber)(properties.sessionTimeout));
        errors.collect(cdk.propertyValidator('userPoolArn', cdk.requiredValidator)(properties.userPoolArn));
        errors.collect(cdk.propertyValidator('userPoolArn', cdk.validateString)(properties.userPoolArn));
        errors.collect(cdk.propertyValidator('userPoolClientId', cdk.requiredValidator)(properties.userPoolClientId));
        errors.collect(cdk.propertyValidator('userPoolClientId', cdk.validateString)(properties.userPoolClientId));
        errors.collect(cdk.propertyValidator('userPoolDomain', cdk.requiredValidator)(properties.userPoolDomain));
        errors.collect(cdk.propertyValidator('userPoolDomain', cdk.validateString)(properties.userPoolDomain));
        return errors.wrap('supplied properties not correct for "AuthenticateCognitoConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.AuthenticateCognitoConfig`` resource
     *
     * @param properties the TypeScript properties of a ``AuthenticateCognitoConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.AuthenticateCognitoConfig`` resource.
     */
    function listenerResourceAuthenticateCognitoConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_AuthenticateCognitoConfigPropertyValidator(properties).assertSuccess();
        return {
            AuthenticationRequestExtraParams: cdk.hashMapper(cdk.stringToCloudFormation)(properties.authenticationRequestExtraParams),
            OnUnauthenticatedRequest: cdk.stringToCloudFormation(properties.onUnauthenticatedRequest),
            Scope: cdk.stringToCloudFormation(properties.scope),
            SessionCookieName: cdk.stringToCloudFormation(properties.sessionCookieName),
            SessionTimeout: cdk.numberToCloudFormation(properties.sessionTimeout),
            UserPoolArn: cdk.stringToCloudFormation(properties.userPoolArn),
            UserPoolClientId: cdk.stringToCloudFormation(properties.userPoolClientId),
            UserPoolDomain: cdk.stringToCloudFormation(properties.userPoolDomain),
        };
    }
    /**
     * Determine whether the given properties match those of a ``AuthenticateOidcConfigProperty``
     *
     * @param properties the TypeScript properties of a ``AuthenticateOidcConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_AuthenticateOidcConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticationRequestExtraParams', cdk.hashValidator(cdk.validateString))(properties.authenticationRequestExtraParams));
        errors.collect(cdk.propertyValidator('authorizationEndpoint', cdk.requiredValidator)(properties.authorizationEndpoint));
        errors.collect(cdk.propertyValidator('authorizationEndpoint', cdk.validateString)(properties.authorizationEndpoint));
        errors.collect(cdk.propertyValidator('clientId', cdk.requiredValidator)(properties.clientId));
        errors.collect(cdk.propertyValidator('clientId', cdk.validateString)(properties.clientId));
        errors.collect(cdk.propertyValidator('clientSecret', cdk.requiredValidator)(properties.clientSecret));
        errors.collect(cdk.propertyValidator('clientSecret', cdk.validateString)(properties.clientSecret));
        errors.collect(cdk.propertyValidator('issuer', cdk.requiredValidator)(properties.issuer));
        errors.collect(cdk.propertyValidator('issuer', cdk.validateString)(properties.issuer));
        errors.collect(cdk.propertyValidator('onUnauthenticatedRequest', cdk.validateString)(properties.onUnauthenticatedRequest));
        errors.collect(cdk.propertyValidator('scope', cdk.validateString)(properties.scope));
        errors.collect(cdk.propertyValidator('sessionCookieName', cdk.validateString)(properties.sessionCookieName));
        errors.collect(cdk.propertyValidator('sessionTimeout', cdk.validateNumber)(properties.sessionTimeout));
        errors.collect(cdk.propertyValidator('tokenEndpoint', cdk.requiredValidator)(properties.tokenEndpoint));
        errors.collect(cdk.propertyValidator('tokenEndpoint', cdk.validateString)(properties.tokenEndpoint));
        errors.collect(cdk.propertyValidator('userInfoEndpoint', cdk.requiredValidator)(properties.userInfoEndpoint));
        errors.collect(cdk.propertyValidator('userInfoEndpoint', cdk.validateString)(properties.userInfoEndpoint));
        return errors.wrap('supplied properties not correct for "AuthenticateOidcConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.AuthenticateOidcConfig`` resource
     *
     * @param properties the TypeScript properties of a ``AuthenticateOidcConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.AuthenticateOidcConfig`` resource.
     */
    function listenerResourceAuthenticateOidcConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_AuthenticateOidcConfigPropertyValidator(properties).assertSuccess();
        return {
            AuthenticationRequestExtraParams: cdk.hashMapper(cdk.stringToCloudFormation)(properties.authenticationRequestExtraParams),
            AuthorizationEndpoint: cdk.stringToCloudFormation(properties.authorizationEndpoint),
            ClientId: cdk.stringToCloudFormation(properties.clientId),
            ClientSecret: cdk.stringToCloudFormation(properties.clientSecret),
            Issuer: cdk.stringToCloudFormation(properties.issuer),
            OnUnauthenticatedRequest: cdk.stringToCloudFormation(properties.onUnauthenticatedRequest),
            Scope: cdk.stringToCloudFormation(properties.scope),
            SessionCookieName: cdk.stringToCloudFormation(properties.sessionCookieName),
            SessionTimeout: cdk.numberToCloudFormation(properties.sessionTimeout),
            TokenEndpoint: cdk.stringToCloudFormation(properties.tokenEndpoint),
            UserInfoEndpoint: cdk.stringToCloudFormation(properties.userInfoEndpoint),
        };
    }
    /**
     * Determine whether the given properties match those of a ``CertificateProperty``
     *
     * @param properties the TypeScript properties of a ``CertificateProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_CertificatePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('certificateArn', cdk.validateString)(properties.certificateArn));
        return errors.wrap('supplied properties not correct for "CertificateProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.Certificate`` resource
     *
     * @param properties the TypeScript properties of a ``CertificateProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.Certificate`` resource.
     */
    function listenerResourceCertificatePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_CertificatePropertyValidator(properties).assertSuccess();
        return {
            CertificateArn: cdk.stringToCloudFormation(properties.certificateArn),
        };
    }
    /**
     * Determine whether the given properties match those of a ``FixedResponseConfigProperty``
     *
     * @param properties the TypeScript properties of a ``FixedResponseConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_FixedResponseConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('contentType', cdk.validateString)(properties.contentType));
        errors.collect(cdk.propertyValidator('messageBody', cdk.validateString)(properties.messageBody));
        errors.collect(cdk.propertyValidator('statusCode', cdk.requiredValidator)(properties.statusCode));
        errors.collect(cdk.propertyValidator('statusCode', cdk.validateString)(properties.statusCode));
        return errors.wrap('supplied properties not correct for "FixedResponseConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.FixedResponseConfig`` resource
     *
     * @param properties the TypeScript properties of a ``FixedResponseConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.FixedResponseConfig`` resource.
     */
    function listenerResourceFixedResponseConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_FixedResponseConfigPropertyValidator(properties).assertSuccess();
        return {
            ContentType: cdk.stringToCloudFormation(properties.contentType),
            MessageBody: cdk.stringToCloudFormation(properties.messageBody),
            StatusCode: cdk.stringToCloudFormation(properties.statusCode),
        };
    }
    /**
     * Determine whether the given properties match those of a ``RedirectConfigProperty``
     *
     * @param properties the TypeScript properties of a ``RedirectConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerResource_RedirectConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('host', cdk.validateString)(properties.host));
        errors.collect(cdk.propertyValidator('path', cdk.validateString)(properties.path));
        errors.collect(cdk.propertyValidator('port', cdk.validateString)(properties.port));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        errors.collect(cdk.propertyValidator('query', cdk.validateString)(properties.query));
        errors.collect(cdk.propertyValidator('statusCode', cdk.requiredValidator)(properties.statusCode));
        errors.collect(cdk.propertyValidator('statusCode', cdk.validateString)(properties.statusCode));
        return errors.wrap('supplied properties not correct for "RedirectConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.RedirectConfig`` resource
     *
     * @param properties the TypeScript properties of a ``RedirectConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::Listener.RedirectConfig`` resource.
     */
    function listenerResourceRedirectConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerResource_RedirectConfigPropertyValidator(properties).assertSuccess();
        return {
            Host: cdk.stringToCloudFormation(properties.host),
            Path: cdk.stringToCloudFormation(properties.path),
            Port: cdk.stringToCloudFormation(properties.port),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
            Query: cdk.stringToCloudFormation(properties.query),
            StatusCode: cdk.stringToCloudFormation(properties.statusCode),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ListenerCertificateResourceProps``
     *
     * @param properties the TypeScript properties of a ``ListenerCertificateResourceProps``
     *
     * @returns the result of the validation.
     */
    function ListenerCertificateResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('certificates', cdk.requiredValidator)(properties.certificates));
        errors.collect(cdk.propertyValidator('certificates', cdk.listValidator(ListenerCertificateResource_CertificatePropertyValidator))(properties.certificates));
        errors.collect(cdk.propertyValidator('listenerArn', cdk.requiredValidator)(properties.listenerArn));
        errors.collect(cdk.propertyValidator('listenerArn', cdk.validateString)(properties.listenerArn));
        return errors.wrap('supplied properties not correct for "ListenerCertificateResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerCertificate`` resource
     *
     * @param properties the TypeScript properties of a ``ListenerCertificateResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerCertificate`` resource.
     */
    function listenerCertificateResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerCertificateResourcePropsValidator(properties).assertSuccess();
        return {
            Certificates: cdk.listMapper(listenerCertificateResourceCertificatePropertyToCloudFormation)(properties.certificates),
            ListenerArn: cdk.stringToCloudFormation(properties.listenerArn),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listenercertificate.html
     */
    class ListenerCertificateResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancingV2::ListenerCertificate``.
         *
         * @param parent   the ``cdk.Construct`` this ``ListenerCertificateResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ListenerCertificateResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ListenerCertificateResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'certificates', this);
            cdk.requireProperty(properties, 'listenerArn', this);
            this.listenerCertificateArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return listenerCertificateResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ListenerCertificateResource.resourceTypeName = "AWS::ElasticLoadBalancingV2::ListenerCertificate";
    cloudformation.ListenerCertificateResource = ListenerCertificateResource;
    /**
     * Determine whether the given properties match those of a ``CertificateProperty``
     *
     * @param properties the TypeScript properties of a ``CertificateProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerCertificateResource_CertificatePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('certificateArn', cdk.validateString)(properties.certificateArn));
        return errors.wrap('supplied properties not correct for "CertificateProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerCertificate.Certificate`` resource
     *
     * @param properties the TypeScript properties of a ``CertificateProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerCertificate.Certificate`` resource.
     */
    function listenerCertificateResourceCertificatePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerCertificateResource_CertificatePropertyValidator(properties).assertSuccess();
        return {
            CertificateArn: cdk.stringToCloudFormation(properties.certificateArn),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ListenerRuleResourceProps``
     *
     * @param properties the TypeScript properties of a ``ListenerRuleResourceProps``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('actions', cdk.requiredValidator)(properties.actions));
        errors.collect(cdk.propertyValidator('actions', cdk.listValidator(ListenerRuleResource_ActionPropertyValidator))(properties.actions));
        errors.collect(cdk.propertyValidator('conditions', cdk.requiredValidator)(properties.conditions));
        errors.collect(cdk.propertyValidator('conditions', cdk.listValidator(ListenerRuleResource_RuleConditionPropertyValidator))(properties.conditions));
        errors.collect(cdk.propertyValidator('listenerArn', cdk.requiredValidator)(properties.listenerArn));
        errors.collect(cdk.propertyValidator('listenerArn', cdk.validateString)(properties.listenerArn));
        errors.collect(cdk.propertyValidator('priority', cdk.requiredValidator)(properties.priority));
        errors.collect(cdk.propertyValidator('priority', cdk.validateNumber)(properties.priority));
        return errors.wrap('supplied properties not correct for "ListenerRuleResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule`` resource
     *
     * @param properties the TypeScript properties of a ``ListenerRuleResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule`` resource.
     */
    function listenerRuleResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResourcePropsValidator(properties).assertSuccess();
        return {
            Actions: cdk.listMapper(listenerRuleResourceActionPropertyToCloudFormation)(properties.actions),
            Conditions: cdk.listMapper(listenerRuleResourceRuleConditionPropertyToCloudFormation)(properties.conditions),
            ListenerArn: cdk.stringToCloudFormation(properties.listenerArn),
            Priority: cdk.numberToCloudFormation(properties.priority),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listenerrule.html
     */
    class ListenerRuleResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancingV2::ListenerRule``.
         *
         * @param parent   the ``cdk.Construct`` this ``ListenerRuleResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ListenerRuleResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ListenerRuleResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'actions', this);
            cdk.requireProperty(properties, 'conditions', this);
            cdk.requireProperty(properties, 'listenerArn', this);
            cdk.requireProperty(properties, 'priority', this);
            this.listenerRuleArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return listenerRuleResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ListenerRuleResource.resourceTypeName = "AWS::ElasticLoadBalancingV2::ListenerRule";
    cloudformation.ListenerRuleResource = ListenerRuleResource;
    /**
     * Determine whether the given properties match those of a ``ActionProperty``
     *
     * @param properties the TypeScript properties of a ``ActionProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_ActionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticateCognitoConfig', ListenerRuleResource_AuthenticateCognitoConfigPropertyValidator)(properties.authenticateCognitoConfig));
        errors.collect(cdk.propertyValidator('authenticateOidcConfig', ListenerRuleResource_AuthenticateOidcConfigPropertyValidator)(properties.authenticateOidcConfig));
        errors.collect(cdk.propertyValidator('fixedResponseConfig', ListenerRuleResource_FixedResponseConfigPropertyValidator)(properties.fixedResponseConfig));
        errors.collect(cdk.propertyValidator('order', cdk.validateNumber)(properties.order));
        errors.collect(cdk.propertyValidator('redirectConfig', ListenerRuleResource_RedirectConfigPropertyValidator)(properties.redirectConfig));
        errors.collect(cdk.propertyValidator('targetGroupArn', cdk.validateString)(properties.targetGroupArn));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "ActionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.Action`` resource
     *
     * @param properties the TypeScript properties of a ``ActionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.Action`` resource.
     */
    function listenerRuleResourceActionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_ActionPropertyValidator(properties).assertSuccess();
        return {
            AuthenticateCognitoConfig: listenerRuleResourceAuthenticateCognitoConfigPropertyToCloudFormation(properties.authenticateCognitoConfig),
            AuthenticateOidcConfig: listenerRuleResourceAuthenticateOidcConfigPropertyToCloudFormation(properties.authenticateOidcConfig),
            FixedResponseConfig: listenerRuleResourceFixedResponseConfigPropertyToCloudFormation(properties.fixedResponseConfig),
            Order: cdk.numberToCloudFormation(properties.order),
            RedirectConfig: listenerRuleResourceRedirectConfigPropertyToCloudFormation(properties.redirectConfig),
            TargetGroupArn: cdk.stringToCloudFormation(properties.targetGroupArn),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``AuthenticateCognitoConfigProperty``
     *
     * @param properties the TypeScript properties of a ``AuthenticateCognitoConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_AuthenticateCognitoConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticationRequestExtraParams', cdk.hashValidator(cdk.validateString))(properties.authenticationRequestExtraParams));
        errors.collect(cdk.propertyValidator('onUnauthenticatedRequest', cdk.validateString)(properties.onUnauthenticatedRequest));
        errors.collect(cdk.propertyValidator('scope', cdk.validateString)(properties.scope));
        errors.collect(cdk.propertyValidator('sessionCookieName', cdk.validateString)(properties.sessionCookieName));
        errors.collect(cdk.propertyValidator('sessionTimeout', cdk.validateNumber)(properties.sessionTimeout));
        errors.collect(cdk.propertyValidator('userPoolArn', cdk.requiredValidator)(properties.userPoolArn));
        errors.collect(cdk.propertyValidator('userPoolArn', cdk.validateString)(properties.userPoolArn));
        errors.collect(cdk.propertyValidator('userPoolClientId', cdk.requiredValidator)(properties.userPoolClientId));
        errors.collect(cdk.propertyValidator('userPoolClientId', cdk.validateString)(properties.userPoolClientId));
        errors.collect(cdk.propertyValidator('userPoolDomain', cdk.requiredValidator)(properties.userPoolDomain));
        errors.collect(cdk.propertyValidator('userPoolDomain', cdk.validateString)(properties.userPoolDomain));
        return errors.wrap('supplied properties not correct for "AuthenticateCognitoConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.AuthenticateCognitoConfig`` resource
     *
     * @param properties the TypeScript properties of a ``AuthenticateCognitoConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.AuthenticateCognitoConfig`` resource.
     */
    function listenerRuleResourceAuthenticateCognitoConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_AuthenticateCognitoConfigPropertyValidator(properties).assertSuccess();
        return {
            AuthenticationRequestExtraParams: cdk.hashMapper(cdk.stringToCloudFormation)(properties.authenticationRequestExtraParams),
            OnUnauthenticatedRequest: cdk.stringToCloudFormation(properties.onUnauthenticatedRequest),
            Scope: cdk.stringToCloudFormation(properties.scope),
            SessionCookieName: cdk.stringToCloudFormation(properties.sessionCookieName),
            SessionTimeout: cdk.numberToCloudFormation(properties.sessionTimeout),
            UserPoolArn: cdk.stringToCloudFormation(properties.userPoolArn),
            UserPoolClientId: cdk.stringToCloudFormation(properties.userPoolClientId),
            UserPoolDomain: cdk.stringToCloudFormation(properties.userPoolDomain),
        };
    }
    /**
     * Determine whether the given properties match those of a ``AuthenticateOidcConfigProperty``
     *
     * @param properties the TypeScript properties of a ``AuthenticateOidcConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_AuthenticateOidcConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('authenticationRequestExtraParams', cdk.hashValidator(cdk.validateString))(properties.authenticationRequestExtraParams));
        errors.collect(cdk.propertyValidator('authorizationEndpoint', cdk.requiredValidator)(properties.authorizationEndpoint));
        errors.collect(cdk.propertyValidator('authorizationEndpoint', cdk.validateString)(properties.authorizationEndpoint));
        errors.collect(cdk.propertyValidator('clientId', cdk.requiredValidator)(properties.clientId));
        errors.collect(cdk.propertyValidator('clientId', cdk.validateString)(properties.clientId));
        errors.collect(cdk.propertyValidator('clientSecret', cdk.requiredValidator)(properties.clientSecret));
        errors.collect(cdk.propertyValidator('clientSecret', cdk.validateString)(properties.clientSecret));
        errors.collect(cdk.propertyValidator('issuer', cdk.requiredValidator)(properties.issuer));
        errors.collect(cdk.propertyValidator('issuer', cdk.validateString)(properties.issuer));
        errors.collect(cdk.propertyValidator('onUnauthenticatedRequest', cdk.validateString)(properties.onUnauthenticatedRequest));
        errors.collect(cdk.propertyValidator('scope', cdk.validateString)(properties.scope));
        errors.collect(cdk.propertyValidator('sessionCookieName', cdk.validateString)(properties.sessionCookieName));
        errors.collect(cdk.propertyValidator('sessionTimeout', cdk.validateNumber)(properties.sessionTimeout));
        errors.collect(cdk.propertyValidator('tokenEndpoint', cdk.requiredValidator)(properties.tokenEndpoint));
        errors.collect(cdk.propertyValidator('tokenEndpoint', cdk.validateString)(properties.tokenEndpoint));
        errors.collect(cdk.propertyValidator('userInfoEndpoint', cdk.requiredValidator)(properties.userInfoEndpoint));
        errors.collect(cdk.propertyValidator('userInfoEndpoint', cdk.validateString)(properties.userInfoEndpoint));
        return errors.wrap('supplied properties not correct for "AuthenticateOidcConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.AuthenticateOidcConfig`` resource
     *
     * @param properties the TypeScript properties of a ``AuthenticateOidcConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.AuthenticateOidcConfig`` resource.
     */
    function listenerRuleResourceAuthenticateOidcConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_AuthenticateOidcConfigPropertyValidator(properties).assertSuccess();
        return {
            AuthenticationRequestExtraParams: cdk.hashMapper(cdk.stringToCloudFormation)(properties.authenticationRequestExtraParams),
            AuthorizationEndpoint: cdk.stringToCloudFormation(properties.authorizationEndpoint),
            ClientId: cdk.stringToCloudFormation(properties.clientId),
            ClientSecret: cdk.stringToCloudFormation(properties.clientSecret),
            Issuer: cdk.stringToCloudFormation(properties.issuer),
            OnUnauthenticatedRequest: cdk.stringToCloudFormation(properties.onUnauthenticatedRequest),
            Scope: cdk.stringToCloudFormation(properties.scope),
            SessionCookieName: cdk.stringToCloudFormation(properties.sessionCookieName),
            SessionTimeout: cdk.numberToCloudFormation(properties.sessionTimeout),
            TokenEndpoint: cdk.stringToCloudFormation(properties.tokenEndpoint),
            UserInfoEndpoint: cdk.stringToCloudFormation(properties.userInfoEndpoint),
        };
    }
    /**
     * Determine whether the given properties match those of a ``FixedResponseConfigProperty``
     *
     * @param properties the TypeScript properties of a ``FixedResponseConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_FixedResponseConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('contentType', cdk.validateString)(properties.contentType));
        errors.collect(cdk.propertyValidator('messageBody', cdk.validateString)(properties.messageBody));
        errors.collect(cdk.propertyValidator('statusCode', cdk.requiredValidator)(properties.statusCode));
        errors.collect(cdk.propertyValidator('statusCode', cdk.validateString)(properties.statusCode));
        return errors.wrap('supplied properties not correct for "FixedResponseConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.FixedResponseConfig`` resource
     *
     * @param properties the TypeScript properties of a ``FixedResponseConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.FixedResponseConfig`` resource.
     */
    function listenerRuleResourceFixedResponseConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_FixedResponseConfigPropertyValidator(properties).assertSuccess();
        return {
            ContentType: cdk.stringToCloudFormation(properties.contentType),
            MessageBody: cdk.stringToCloudFormation(properties.messageBody),
            StatusCode: cdk.stringToCloudFormation(properties.statusCode),
        };
    }
    /**
     * Determine whether the given properties match those of a ``RedirectConfigProperty``
     *
     * @param properties the TypeScript properties of a ``RedirectConfigProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_RedirectConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('host', cdk.validateString)(properties.host));
        errors.collect(cdk.propertyValidator('path', cdk.validateString)(properties.path));
        errors.collect(cdk.propertyValidator('port', cdk.validateString)(properties.port));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        errors.collect(cdk.propertyValidator('query', cdk.validateString)(properties.query));
        errors.collect(cdk.propertyValidator('statusCode', cdk.requiredValidator)(properties.statusCode));
        errors.collect(cdk.propertyValidator('statusCode', cdk.validateString)(properties.statusCode));
        return errors.wrap('supplied properties not correct for "RedirectConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.RedirectConfig`` resource
     *
     * @param properties the TypeScript properties of a ``RedirectConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.RedirectConfig`` resource.
     */
    function listenerRuleResourceRedirectConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_RedirectConfigPropertyValidator(properties).assertSuccess();
        return {
            Host: cdk.stringToCloudFormation(properties.host),
            Path: cdk.stringToCloudFormation(properties.path),
            Port: cdk.stringToCloudFormation(properties.port),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
            Query: cdk.stringToCloudFormation(properties.query),
            StatusCode: cdk.stringToCloudFormation(properties.statusCode),
        };
    }
    /**
     * Determine whether the given properties match those of a ``RuleConditionProperty``
     *
     * @param properties the TypeScript properties of a ``RuleConditionProperty``
     *
     * @returns the result of the validation.
     */
    function ListenerRuleResource_RuleConditionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('field', cdk.validateString)(properties.field));
        errors.collect(cdk.propertyValidator('values', cdk.listValidator(cdk.validateString))(properties.values));
        return errors.wrap('supplied properties not correct for "RuleConditionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.RuleCondition`` resource
     *
     * @param properties the TypeScript properties of a ``RuleConditionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::ListenerRule.RuleCondition`` resource.
     */
    function listenerRuleResourceRuleConditionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ListenerRuleResource_RuleConditionPropertyValidator(properties).assertSuccess();
        return {
            Field: cdk.stringToCloudFormation(properties.field),
            Values: cdk.listMapper(cdk.stringToCloudFormation)(properties.values),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``LoadBalancerResourceProps``
     *
     * @param properties the TypeScript properties of a ``LoadBalancerResourceProps``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('ipAddressType', cdk.validateString)(properties.ipAddressType));
        errors.collect(cdk.propertyValidator('loadBalancerAttributes', cdk.listValidator(LoadBalancerResource_LoadBalancerAttributePropertyValidator))(properties.loadBalancerAttributes));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('scheme', cdk.validateString)(properties.scheme));
        errors.collect(cdk.propertyValidator('securityGroups', cdk.listValidator(cdk.validateString))(properties.securityGroups));
        errors.collect(cdk.propertyValidator('subnetMappings', cdk.listValidator(LoadBalancerResource_SubnetMappingPropertyValidator))(properties.subnetMappings));
        errors.collect(cdk.propertyValidator('subnets', cdk.listValidator(cdk.validateString))(properties.subnets));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateTag))(properties.tags));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "LoadBalancerResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer`` resource
     *
     * @param properties the TypeScript properties of a ``LoadBalancerResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer`` resource.
     */
    function loadBalancerResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResourcePropsValidator(properties).assertSuccess();
        return {
            IpAddressType: cdk.stringToCloudFormation(properties.ipAddressType),
            LoadBalancerAttributes: cdk.listMapper(loadBalancerResourceLoadBalancerAttributePropertyToCloudFormation)(properties.loadBalancerAttributes),
            Name: cdk.stringToCloudFormation(properties.name),
            Scheme: cdk.stringToCloudFormation(properties.scheme),
            SecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroups),
            SubnetMappings: cdk.listMapper(loadBalancerResourceSubnetMappingPropertyToCloudFormation)(properties.subnetMappings),
            Subnets: cdk.listMapper(cdk.stringToCloudFormation)(properties.subnets),
            Tags: cdk.listMapper(cdk.tagToCloudFormation)(properties.tags),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-loadbalancer.html
     */
    class LoadBalancerResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancingV2::LoadBalancer``.
         *
         * @param parent   the ``cdk.Construct`` this ``LoadBalancerResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``LoadBalancerResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: LoadBalancerResource.resourceTypeName, properties });
            this.loadBalancerCanonicalHostedZoneId = this.getAtt('CanonicalHostedZoneID').toString();
            this.loadBalancerDnsName = this.getAtt('DNSName').toString();
            this.loadBalancerFullName = this.getAtt('LoadBalancerFullName').toString();
            this.loadBalancerName = this.getAtt('LoadBalancerName').toString();
            this.loadBalancerSecurityGroups = new LoadBalancerSecurityGroups(this.getAtt('SecurityGroups'));
            this.loadBalancerArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return loadBalancerResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    LoadBalancerResource.resourceTypeName = "AWS::ElasticLoadBalancingV2::LoadBalancer";
    cloudformation.LoadBalancerResource = LoadBalancerResource;
    /**
     * Determine whether the given properties match those of a ``LoadBalancerAttributeProperty``
     *
     * @param properties the TypeScript properties of a ``LoadBalancerAttributeProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_LoadBalancerAttributePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('key', cdk.validateString)(properties.key));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "LoadBalancerAttributeProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer.LoadBalancerAttribute`` resource
     *
     * @param properties the TypeScript properties of a ``LoadBalancerAttributeProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer.LoadBalancerAttribute`` resource.
     */
    function loadBalancerResourceLoadBalancerAttributePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_LoadBalancerAttributePropertyValidator(properties).assertSuccess();
        return {
            Key: cdk.stringToCloudFormation(properties.key),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
    /**
     * Determine whether the given properties match those of a ``SubnetMappingProperty``
     *
     * @param properties the TypeScript properties of a ``SubnetMappingProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_SubnetMappingPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('allocationId', cdk.requiredValidator)(properties.allocationId));
        errors.collect(cdk.propertyValidator('allocationId', cdk.validateString)(properties.allocationId));
        errors.collect(cdk.propertyValidator('subnetId', cdk.requiredValidator)(properties.subnetId));
        errors.collect(cdk.propertyValidator('subnetId', cdk.validateString)(properties.subnetId));
        return errors.wrap('supplied properties not correct for "SubnetMappingProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer.SubnetMapping`` resource
     *
     * @param properties the TypeScript properties of a ``SubnetMappingProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::LoadBalancer.SubnetMapping`` resource.
     */
    function loadBalancerResourceSubnetMappingPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_SubnetMappingPropertyValidator(properties).assertSuccess();
        return {
            AllocationId: cdk.stringToCloudFormation(properties.allocationId),
            SubnetId: cdk.stringToCloudFormation(properties.subnetId),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
class LoadBalancerSecurityGroups extends cdk.Token {
    constructor() {
        super(...arguments);
        // @ts-ignore: private but unused on purpose.
        this.thisIsALoadBalancerSecurityGroups = true;
    }
}
exports.LoadBalancerSecurityGroups = LoadBalancerSecurityGroups;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``TargetGroupResourceProps``
     *
     * @param properties the TypeScript properties of a ``TargetGroupResourceProps``
     *
     * @returns the result of the validation.
     */
    function TargetGroupResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('healthCheckIntervalSeconds', cdk.validateNumber)(properties.healthCheckIntervalSeconds));
        errors.collect(cdk.propertyValidator('healthCheckPath', cdk.validateString)(properties.healthCheckPath));
        errors.collect(cdk.propertyValidator('healthCheckPort', cdk.validateString)(properties.healthCheckPort));
        errors.collect(cdk.propertyValidator('healthCheckProtocol', cdk.validateString)(properties.healthCheckProtocol));
        errors.collect(cdk.propertyValidator('healthCheckTimeoutSeconds', cdk.validateNumber)(properties.healthCheckTimeoutSeconds));
        errors.collect(cdk.propertyValidator('healthyThresholdCount', cdk.validateNumber)(properties.healthyThresholdCount));
        errors.collect(cdk.propertyValidator('matcher', TargetGroupResource_MatcherPropertyValidator)(properties.matcher));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('port', cdk.requiredValidator)(properties.port));
        errors.collect(cdk.propertyValidator('port', cdk.validateNumber)(properties.port));
        errors.collect(cdk.propertyValidator('protocol', cdk.requiredValidator)(properties.protocol));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateTag))(properties.tags));
        errors.collect(cdk.propertyValidator('targetGroupAttributes', cdk.listValidator(TargetGroupResource_TargetGroupAttributePropertyValidator))(properties.targetGroupAttributes));
        errors.collect(cdk.propertyValidator('targetType', cdk.validateString)(properties.targetType));
        errors.collect(cdk.propertyValidator('targets', cdk.listValidator(TargetGroupResource_TargetDescriptionPropertyValidator))(properties.targets));
        errors.collect(cdk.propertyValidator('unhealthyThresholdCount', cdk.validateNumber)(properties.unhealthyThresholdCount));
        errors.collect(cdk.propertyValidator('vpcId', cdk.requiredValidator)(properties.vpcId));
        errors.collect(cdk.propertyValidator('vpcId', cdk.validateString)(properties.vpcId));
        return errors.wrap('supplied properties not correct for "TargetGroupResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup`` resource
     *
     * @param properties the TypeScript properties of a ``TargetGroupResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup`` resource.
     */
    function targetGroupResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TargetGroupResourcePropsValidator(properties).assertSuccess();
        return {
            Port: cdk.numberToCloudFormation(properties.port),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
            VpcId: cdk.stringToCloudFormation(properties.vpcId),
            HealthCheckIntervalSeconds: cdk.numberToCloudFormation(properties.healthCheckIntervalSeconds),
            HealthCheckPath: cdk.stringToCloudFormation(properties.healthCheckPath),
            HealthCheckPort: cdk.stringToCloudFormation(properties.healthCheckPort),
            HealthCheckProtocol: cdk.stringToCloudFormation(properties.healthCheckProtocol),
            HealthCheckTimeoutSeconds: cdk.numberToCloudFormation(properties.healthCheckTimeoutSeconds),
            HealthyThresholdCount: cdk.numberToCloudFormation(properties.healthyThresholdCount),
            Matcher: targetGroupResourceMatcherPropertyToCloudFormation(properties.matcher),
            Name: cdk.stringToCloudFormation(properties.name),
            Tags: cdk.listMapper(cdk.tagToCloudFormation)(properties.tags),
            TargetGroupAttributes: cdk.listMapper(targetGroupResourceTargetGroupAttributePropertyToCloudFormation)(properties.targetGroupAttributes),
            Targets: cdk.listMapper(targetGroupResourceTargetDescriptionPropertyToCloudFormation)(properties.targets),
            TargetType: cdk.stringToCloudFormation(properties.targetType),
            UnhealthyThresholdCount: cdk.numberToCloudFormation(properties.unhealthyThresholdCount),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html
     */
    class TargetGroupResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancingV2::TargetGroup``.
         *
         * @param parent   the ``cdk.Construct`` this ``TargetGroupResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``TargetGroupResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: TargetGroupResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'port', this);
            cdk.requireProperty(properties, 'protocol', this);
            cdk.requireProperty(properties, 'vpcId', this);
            this.targetGroupLoadBalancerArns = new TargetGroupLoadBalancerArns(this.getAtt('LoadBalancerArns'));
            this.targetGroupFullName = this.getAtt('TargetGroupFullName').toString();
            this.targetGroupName = this.getAtt('TargetGroupName').toString();
            this.targetGroupArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return targetGroupResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    TargetGroupResource.resourceTypeName = "AWS::ElasticLoadBalancingV2::TargetGroup";
    cloudformation.TargetGroupResource = TargetGroupResource;
    /**
     * Determine whether the given properties match those of a ``MatcherProperty``
     *
     * @param properties the TypeScript properties of a ``MatcherProperty``
     *
     * @returns the result of the validation.
     */
    function TargetGroupResource_MatcherPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('httpCode', cdk.requiredValidator)(properties.httpCode));
        errors.collect(cdk.propertyValidator('httpCode', cdk.validateString)(properties.httpCode));
        return errors.wrap('supplied properties not correct for "MatcherProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.Matcher`` resource
     *
     * @param properties the TypeScript properties of a ``MatcherProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.Matcher`` resource.
     */
    function targetGroupResourceMatcherPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TargetGroupResource_MatcherPropertyValidator(properties).assertSuccess();
        return {
            HttpCode: cdk.stringToCloudFormation(properties.httpCode),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TargetDescriptionProperty``
     *
     * @param properties the TypeScript properties of a ``TargetDescriptionProperty``
     *
     * @returns the result of the validation.
     */
    function TargetGroupResource_TargetDescriptionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('availabilityZone', cdk.validateString)(properties.availabilityZone));
        errors.collect(cdk.propertyValidator('id', cdk.requiredValidator)(properties.id));
        errors.collect(cdk.propertyValidator('id', cdk.validateString)(properties.id));
        errors.collect(cdk.propertyValidator('port', cdk.validateNumber)(properties.port));
        return errors.wrap('supplied properties not correct for "TargetDescriptionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.TargetDescription`` resource
     *
     * @param properties the TypeScript properties of a ``TargetDescriptionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.TargetDescription`` resource.
     */
    function targetGroupResourceTargetDescriptionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TargetGroupResource_TargetDescriptionPropertyValidator(properties).assertSuccess();
        return {
            AvailabilityZone: cdk.stringToCloudFormation(properties.availabilityZone),
            Id: cdk.stringToCloudFormation(properties.id),
            Port: cdk.numberToCloudFormation(properties.port),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TargetGroupAttributeProperty``
     *
     * @param properties the TypeScript properties of a ``TargetGroupAttributeProperty``
     *
     * @returns the result of the validation.
     */
    function TargetGroupResource_TargetGroupAttributePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('key', cdk.validateString)(properties.key));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "TargetGroupAttributeProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.TargetGroupAttribute`` resource
     *
     * @param properties the TypeScript properties of a ``TargetGroupAttributeProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancingV2::TargetGroup.TargetGroupAttribute`` resource.
     */
    function targetGroupResourceTargetGroupAttributePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TargetGroupResource_TargetGroupAttributePropertyValidator(properties).assertSuccess();
        return {
            Key: cdk.stringToCloudFormation(properties.key),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
class TargetGroupLoadBalancerArns extends cdk.Token {
    constructor() {
        super(...arguments);
        // @ts-ignore: private but unused on purpose.
        this.thisIsATargetGroupLoadBalancerArns = true;
    }
}
exports.TargetGroupLoadBalancerArns = TargetGroupLoadBalancerArns;
