// @aws-cdk/aws-stepfunctions/lib/states/wait.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chain_1 = require("../chain");
const state_1 = require("./state");
/**
 * Define a Wait state in the state machine
 *
 * A Wait state can be used to delay execution of the state machine for a while.
 */
class Wait extends state_1.State {
    constructor(parent, id, props) {
        super(parent, id, props);
        this.seconds = props.seconds;
        this.timestamp = props.timestamp;
        this.secondsPath = props.secondsPath;
        this.timestampPath = props.timestampPath;
        this.endStates = [this];
    }
    /**
     * Continue normal execution with the given state
     */
    next(next) {
        super.makeNext(next.startState);
        return chain_1.Chain.sequence(this, next);
    }
    /**
     * Return the Amazon States Language object for this state
     */
    toStateJson() {
        return Object.assign({ Type: state_1.StateType.Wait, Comment: this.comment, Seconds: this.seconds, Timestamp: this.timestamp, SecondsPath: this.secondsPath, TimestampPath: this.timestampPath }, this.renderNextEnd());
    }
}
exports.Wait = Wait;
