// @aws-cdk/aws-ecs/lib/base/scalable-task-count.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const appscaling = require("@aws-cdk/aws-applicationautoscaling");
/**
 * Scalable attribute representing task count
 */
class ScalableTaskCount extends appscaling.BaseScalableAttribute {
    /**
     * Scale out or in based on time
     */
    scaleOnSchedule(id, props) {
        return super.doScaleOnSchedule(id, props);
    }
    /**
     * Scale out or in based on a metric value
     */
    scaleOnMetric(id, props) {
        return super.doScaleOnMetric(id, props);
    }
    /**
     * Scale out or in to achieve a target CPU utilization
     */
    scaleOnCpuUtilization(id, props) {
        return super.doScaleToTrackMetric(id, {
            predefinedMetric: appscaling.PredefinedMetric.ECSServiceAverageCPUUtilization,
            policyName: props.policyName,
            disableScaleIn: props.disableScaleIn,
            targetValue: props.targetUtilizationPercent,
            scaleInCooldownSec: props.scaleInCooldownSec,
            scaleOutCooldownSec: props.scaleOutCooldownSec,
        });
    }
    /**
     * Scale out or in to achieve a target memory utilization utilization
     */
    scaleOnMemoryUtilization(id, props) {
        return super.doScaleToTrackMetric(id, {
            predefinedMetric: appscaling.PredefinedMetric.ECSServiceAverageMemoryUtilization,
            targetValue: props.targetUtilizationPercent,
            policyName: props.policyName,
            disableScaleIn: props.disableScaleIn,
            scaleInCooldownSec: props.scaleInCooldownSec,
            scaleOutCooldownSec: props.scaleOutCooldownSec,
        });
    }
    /**
     * Scale out or in to track a custom metric
     */
    scaleToTrackCustomMetric(id, props) {
        return super.doScaleToTrackMetric(id, {
            customMetric: props.metric,
            targetValue: props.targetValue,
            policyName: props.policyName,
            disableScaleIn: props.disableScaleIn,
            scaleInCooldownSec: props.scaleInCooldownSec,
            scaleOutCooldownSec: props.scaleOutCooldownSec,
        });
    }
}
exports.ScalableTaskCount = ScalableTaskCount;
