// @aws-cdk/aws-autoscaling/test/test.lifecyclehooks.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/cdk");
const autoscaling = require("../lib");
class FakeNotificationTarget {
    asLifecycleHookTarget(lifecycleHook) {
        lifecycleHook.role.addToPolicy(new iam.PolicyStatement()
            .addAction('action:Work')
            .addAllResources());
        return { notificationTargetArn: 'target:arn', };
    }
}
module.exports = {
    'we can add a lifecycle hook to an ASG'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const vpc = new ec2.VpcNetwork(stack, 'VPC');
        const asg = new autoscaling.AutoScalingGroup(stack, 'ASG', {
            vpc,
            instanceType: new ec2.InstanceTypePair(ec2.InstanceClass.M4, ec2.InstanceSize.Micro),
            machineImage: new ec2.AmazonLinuxImage(),
        });
        // WHEN
        asg.onLifecycleTransition('Transition', {
            notificationTarget: new FakeNotificationTarget(),
            lifecycleTransition: autoscaling.LifecycleTransition.InstanceLaunching,
            defaultResult: autoscaling.DefaultResult.Abandon,
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::LifecycleHook', {
            LifecycleTransition: "autoscaling:EC2_INSTANCE_LAUNCHING",
            DefaultResult: "ABANDON",
            NotificationTargetARN: "target:arn",
        }));
        // Lifecycle Hook has a dependency on the policy object
        assert_1.expect(stack).to(assert_1.haveResource('AWS::AutoScaling::LifecycleHook', {
            DependsOn: [
                "ASGLifecycleHookTransitionRole3AAA6BB7",
                "ASGLifecycleHookTransitionRoleDefaultPolicy2E50C7DB"
            ]
        }, assert_1.ResourcePart.CompleteDefinition));
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Version: '2012-10-17',
                Statement: [
                    {
                        Action: "sts:AssumeRole",
                        Effect: "Allow",
                        Principal: { Service: "autoscaling.amazonaws.com" }
                    }
                ],
            }
        }));
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
            PolicyDocument: {
                Version: '2012-10-17',
                Statement: [
                    {
                        Action: "action:Work",
                        Effect: "Allow",
                        Resource: "*"
                    }
                ],
            }
        }));
        test.done();
    }
};
