// @aws-cdk/aws-route53/test/test.txt-record.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk_1 = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
class TestApp {
    constructor() {
        this.app = new cdk_1.App();
        const account = '123456789012';
        const region = 'bermuda-triangle';
        this.app.setContext(`availability-zones:${account}:${region}`, [`${region}-1a`]);
        this.stack = new cdk_1.Stack(this.app, 'MyStack', { env: { account, region } });
    }
    synthesizeTemplate() {
        return this.app.synthesizeStack(this.stack.name);
    }
}
module.exports = {
    'TXT records': {
        TXT(test) {
            const app = new TestApp();
            const zone = new lib_1.PublicHostedZone(app.stack, 'HostedZone', { zoneName: 'test.public' });
            new lib_1.TXTRecord(zone, 'TXT', { recordName: '_foo', recordValue: 'Bar!' });
            assert_1.expect(app.synthesizeTemplate()).to(assert_1.exactlyMatchTemplate({
                Resources: {
                    HostedZoneDB99F866: {
                        Type: 'AWS::Route53::HostedZone',
                        Properties: {
                            Name: 'test.public.'
                        }
                    },
                    HostedZoneTXT69C29760: {
                        Type: 'AWS::Route53::RecordSet',
                        Properties: {
                            HostedZoneId: {
                                Ref: 'HostedZoneDB99F866'
                            },
                            Name: '_foo.test.public.',
                            ResourceRecords: ['"Bar!"'],
                            Type: 'TXT',
                            TTL: '1800'
                        }
                    }
                }
            }));
            test.done();
        }
    }
};
