// @aws-cdk/aws-route53/test/test.hosted-zone-provider.js
"use strict";
const cdk = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
module.exports = {
    'Hosted Zone Provider': {
        'HostedZoneProvider will return context values if availble'(test) {
            // GIVEN
            const stack = new cdk.Stack(undefined, 'TestStack', { env: { account: '12345', region: 'us-east-1' } });
            const filter = { domainName: 'test.com' };
            new lib_1.HostedZoneProvider(stack, filter).findHostedZone();
            const key = Object.keys(stack.missingContext)[0];
            const fakeZone = {
                Id: "/hostedzone/11111111111111",
                Name: "example.com.",
                CallerReference: "TestLates-PublicZo-OESZPDFV7G6A",
                Config: {
                    Comment: "CDK created",
                    PrivateZone: false
                },
                ResourceRecordSetCount: 3
            };
            stack.setContext(key, fakeZone);
            const cdkZoneProps = {
                hostedZoneId: fakeZone.Id,
                zoneName: 'example.com',
            };
            const cdkZone = lib_1.HostedZoneRef.import(stack, 'MyZone', cdkZoneProps);
            // WHEN
            const provider = new lib_1.HostedZoneProvider(stack, filter);
            const zoneProps = cdk.resolve(provider.findHostedZone());
            const zoneRef = provider.findAndImport(stack, 'MyZoneProvider');
            // THEN
            test.deepEqual(zoneProps, cdkZoneProps);
            test.deepEqual(zoneRef.hostedZoneId, cdkZone.hostedZoneId);
            test.done();
        },
    }
};
