// @aws-cdk/aws-route53/test/test.alias-record.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk_1 = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
module.exports = {
    'test alias record'(test) {
        // GIVEN
        const stack = new cdk_1.Stack();
        const zone = new lib_1.PublicHostedZone(stack, 'HostedZone', { zoneName: 'test.public' });
        const target = {
            asAliasRecordTarget: () => {
                return {
                    hostedZoneId: 'Z2P70J7EXAMPLE',
                    dnsName: 'foo.example.com'
                };
            }
        };
        // WHEN
        new lib_1.AliasRecord(zone, 'Alias', {
            recordName: '_foo',
            target
        });
        // THEN - stack contains a record set
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Route53::RecordSet', {
            Name: '_foo.test.public.',
            HostedZoneId: {
                Ref: 'HostedZoneDB99F866'
            },
            Type: 'A',
            AliasTarget: {
                HostedZoneId: 'Z2P70J7EXAMPLE',
                DNSName: 'foo.example.com',
            }
        }));
        test.done();
    }
};
