// @aws-cdk/aws-elasticloadbalancing/lib/elasticloadbalancing.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:03:58.454Z","fingerprint":"VgghZ/956JoMAJfD9aYwnZym2iGMEWWWOULLrHJdiww="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``LoadBalancerResourceProps``
     *
     * @param properties the TypeScript properties of a ``LoadBalancerResourceProps``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('accessLoggingPolicy', LoadBalancerResource_AccessLoggingPolicyPropertyValidator)(properties.accessLoggingPolicy));
        errors.collect(cdk.propertyValidator('appCookieStickinessPolicy', cdk.listValidator(LoadBalancerResource_AppCookieStickinessPolicyPropertyValidator))(properties.appCookieStickinessPolicy));
        errors.collect(cdk.propertyValidator('availabilityZones', cdk.listValidator(cdk.validateString))(properties.availabilityZones));
        errors.collect(cdk.propertyValidator('connectionDrainingPolicy', LoadBalancerResource_ConnectionDrainingPolicyPropertyValidator)(properties.connectionDrainingPolicy));
        errors.collect(cdk.propertyValidator('connectionSettings', LoadBalancerResource_ConnectionSettingsPropertyValidator)(properties.connectionSettings));
        errors.collect(cdk.propertyValidator('crossZone', cdk.validateBoolean)(properties.crossZone));
        errors.collect(cdk.propertyValidator('healthCheck', LoadBalancerResource_HealthCheckPropertyValidator)(properties.healthCheck));
        errors.collect(cdk.propertyValidator('instances', cdk.listValidator(cdk.validateString))(properties.instances));
        errors.collect(cdk.propertyValidator('lbCookieStickinessPolicy', cdk.listValidator(LoadBalancerResource_LBCookieStickinessPolicyPropertyValidator))(properties.lbCookieStickinessPolicy));
        errors.collect(cdk.propertyValidator('listeners', cdk.requiredValidator)(properties.listeners));
        errors.collect(cdk.propertyValidator('listeners', cdk.listValidator(LoadBalancerResource_ListenersPropertyValidator))(properties.listeners));
        errors.collect(cdk.propertyValidator('loadBalancerName', cdk.validateString)(properties.loadBalancerName));
        errors.collect(cdk.propertyValidator('policies', cdk.listValidator(LoadBalancerResource_PoliciesPropertyValidator))(properties.policies));
        errors.collect(cdk.propertyValidator('scheme', cdk.validateString)(properties.scheme));
        errors.collect(cdk.propertyValidator('securityGroups', cdk.listValidator(cdk.validateString))(properties.securityGroups));
        errors.collect(cdk.propertyValidator('subnets', cdk.listValidator(cdk.validateString))(properties.subnets));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateTag))(properties.tags));
        return errors.wrap('supplied properties not correct for "LoadBalancerResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer`` resource
     *
     * @param properties the TypeScript properties of a ``LoadBalancerResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer`` resource.
     */
    function loadBalancerResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResourcePropsValidator(properties).assertSuccess();
        return {
            Listeners: cdk.listMapper(loadBalancerResourceListenersPropertyToCloudFormation)(properties.listeners),
            AccessLoggingPolicy: loadBalancerResourceAccessLoggingPolicyPropertyToCloudFormation(properties.accessLoggingPolicy),
            AppCookieStickinessPolicy: cdk.listMapper(loadBalancerResourceAppCookieStickinessPolicyPropertyToCloudFormation)(properties.appCookieStickinessPolicy),
            AvailabilityZones: cdk.listMapper(cdk.stringToCloudFormation)(properties.availabilityZones),
            ConnectionDrainingPolicy: loadBalancerResourceConnectionDrainingPolicyPropertyToCloudFormation(properties.connectionDrainingPolicy),
            ConnectionSettings: loadBalancerResourceConnectionSettingsPropertyToCloudFormation(properties.connectionSettings),
            CrossZone: cdk.booleanToCloudFormation(properties.crossZone),
            HealthCheck: loadBalancerResourceHealthCheckPropertyToCloudFormation(properties.healthCheck),
            Instances: cdk.listMapper(cdk.stringToCloudFormation)(properties.instances),
            LBCookieStickinessPolicy: cdk.listMapper(loadBalancerResourceLBCookieStickinessPolicyPropertyToCloudFormation)(properties.lbCookieStickinessPolicy),
            LoadBalancerName: cdk.stringToCloudFormation(properties.loadBalancerName),
            Policies: cdk.listMapper(loadBalancerResourcePoliciesPropertyToCloudFormation)(properties.policies),
            Scheme: cdk.stringToCloudFormation(properties.scheme),
            SecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroups),
            Subnets: cdk.listMapper(cdk.stringToCloudFormation)(properties.subnets),
            Tags: cdk.listMapper(cdk.tagToCloudFormation)(properties.tags),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-elb.html
     */
    class LoadBalancerResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ElasticLoadBalancing::LoadBalancer``.
         *
         * @param parent   the ``cdk.Construct`` this ``LoadBalancerResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``LoadBalancerResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: LoadBalancerResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'listeners', this);
            this.loadBalancerCanonicalHostedZoneName = this.getAtt('CanonicalHostedZoneName').toString();
            this.loadBalancerCanonicalHostedZoneNameId = this.getAtt('CanonicalHostedZoneNameID').toString();
            this.loadBalancerDnsName = this.getAtt('DNSName').toString();
            this.loadBalancerSourceSecurityGroupGroupName = this.getAtt('SourceSecurityGroup.GroupName').toString();
            this.loadBalancerSourceSecurityGroupOwnerAlias = this.getAtt('SourceSecurityGroup.OwnerAlias').toString();
            this.loadBalancerName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return loadBalancerResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    LoadBalancerResource.resourceTypeName = "AWS::ElasticLoadBalancing::LoadBalancer";
    cloudformation.LoadBalancerResource = LoadBalancerResource;
    /**
     * Determine whether the given properties match those of a ``AccessLoggingPolicyProperty``
     *
     * @param properties the TypeScript properties of a ``AccessLoggingPolicyProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_AccessLoggingPolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('emitInterval', cdk.validateNumber)(properties.emitInterval));
        errors.collect(cdk.propertyValidator('enabled', cdk.requiredValidator)(properties.enabled));
        errors.collect(cdk.propertyValidator('enabled', cdk.validateBoolean)(properties.enabled));
        errors.collect(cdk.propertyValidator('s3BucketName', cdk.requiredValidator)(properties.s3BucketName));
        errors.collect(cdk.propertyValidator('s3BucketName', cdk.validateString)(properties.s3BucketName));
        errors.collect(cdk.propertyValidator('s3BucketPrefix', cdk.validateString)(properties.s3BucketPrefix));
        return errors.wrap('supplied properties not correct for "AccessLoggingPolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.AccessLoggingPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``AccessLoggingPolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.AccessLoggingPolicy`` resource.
     */
    function loadBalancerResourceAccessLoggingPolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_AccessLoggingPolicyPropertyValidator(properties).assertSuccess();
        return {
            EmitInterval: cdk.numberToCloudFormation(properties.emitInterval),
            Enabled: cdk.booleanToCloudFormation(properties.enabled),
            S3BucketName: cdk.stringToCloudFormation(properties.s3BucketName),
            S3BucketPrefix: cdk.stringToCloudFormation(properties.s3BucketPrefix),
        };
    }
    /**
     * Determine whether the given properties match those of a ``AppCookieStickinessPolicyProperty``
     *
     * @param properties the TypeScript properties of a ``AppCookieStickinessPolicyProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_AppCookieStickinessPolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('cookieName', cdk.requiredValidator)(properties.cookieName));
        errors.collect(cdk.propertyValidator('cookieName', cdk.validateString)(properties.cookieName));
        errors.collect(cdk.propertyValidator('policyName', cdk.requiredValidator)(properties.policyName));
        errors.collect(cdk.propertyValidator('policyName', cdk.validateString)(properties.policyName));
        return errors.wrap('supplied properties not correct for "AppCookieStickinessPolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.AppCookieStickinessPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``AppCookieStickinessPolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.AppCookieStickinessPolicy`` resource.
     */
    function loadBalancerResourceAppCookieStickinessPolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_AppCookieStickinessPolicyPropertyValidator(properties).assertSuccess();
        return {
            CookieName: cdk.stringToCloudFormation(properties.cookieName),
            PolicyName: cdk.stringToCloudFormation(properties.policyName),
        };
    }
    /**
     * Determine whether the given properties match those of a ``ConnectionDrainingPolicyProperty``
     *
     * @param properties the TypeScript properties of a ``ConnectionDrainingPolicyProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_ConnectionDrainingPolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('enabled', cdk.requiredValidator)(properties.enabled));
        errors.collect(cdk.propertyValidator('enabled', cdk.validateBoolean)(properties.enabled));
        errors.collect(cdk.propertyValidator('timeout', cdk.validateNumber)(properties.timeout));
        return errors.wrap('supplied properties not correct for "ConnectionDrainingPolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.ConnectionDrainingPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``ConnectionDrainingPolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.ConnectionDrainingPolicy`` resource.
     */
    function loadBalancerResourceConnectionDrainingPolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_ConnectionDrainingPolicyPropertyValidator(properties).assertSuccess();
        return {
            Enabled: cdk.booleanToCloudFormation(properties.enabled),
            Timeout: cdk.numberToCloudFormation(properties.timeout),
        };
    }
    /**
     * Determine whether the given properties match those of a ``ConnectionSettingsProperty``
     *
     * @param properties the TypeScript properties of a ``ConnectionSettingsProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_ConnectionSettingsPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('idleTimeout', cdk.requiredValidator)(properties.idleTimeout));
        errors.collect(cdk.propertyValidator('idleTimeout', cdk.validateNumber)(properties.idleTimeout));
        return errors.wrap('supplied properties not correct for "ConnectionSettingsProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.ConnectionSettings`` resource
     *
     * @param properties the TypeScript properties of a ``ConnectionSettingsProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.ConnectionSettings`` resource.
     */
    function loadBalancerResourceConnectionSettingsPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_ConnectionSettingsPropertyValidator(properties).assertSuccess();
        return {
            IdleTimeout: cdk.numberToCloudFormation(properties.idleTimeout),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HealthCheckProperty``
     *
     * @param properties the TypeScript properties of a ``HealthCheckProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_HealthCheckPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('healthyThreshold', cdk.requiredValidator)(properties.healthyThreshold));
        errors.collect(cdk.propertyValidator('healthyThreshold', cdk.validateString)(properties.healthyThreshold));
        errors.collect(cdk.propertyValidator('interval', cdk.requiredValidator)(properties.interval));
        errors.collect(cdk.propertyValidator('interval', cdk.validateString)(properties.interval));
        errors.collect(cdk.propertyValidator('target', cdk.requiredValidator)(properties.target));
        errors.collect(cdk.propertyValidator('target', cdk.validateString)(properties.target));
        errors.collect(cdk.propertyValidator('timeout', cdk.requiredValidator)(properties.timeout));
        errors.collect(cdk.propertyValidator('timeout', cdk.validateString)(properties.timeout));
        errors.collect(cdk.propertyValidator('unhealthyThreshold', cdk.requiredValidator)(properties.unhealthyThreshold));
        errors.collect(cdk.propertyValidator('unhealthyThreshold', cdk.validateString)(properties.unhealthyThreshold));
        return errors.wrap('supplied properties not correct for "HealthCheckProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.HealthCheck`` resource
     *
     * @param properties the TypeScript properties of a ``HealthCheckProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.HealthCheck`` resource.
     */
    function loadBalancerResourceHealthCheckPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_HealthCheckPropertyValidator(properties).assertSuccess();
        return {
            HealthyThreshold: cdk.stringToCloudFormation(properties.healthyThreshold),
            Interval: cdk.stringToCloudFormation(properties.interval),
            Target: cdk.stringToCloudFormation(properties.target),
            Timeout: cdk.stringToCloudFormation(properties.timeout),
            UnhealthyThreshold: cdk.stringToCloudFormation(properties.unhealthyThreshold),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LBCookieStickinessPolicyProperty``
     *
     * @param properties the TypeScript properties of a ``LBCookieStickinessPolicyProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_LBCookieStickinessPolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('cookieExpirationPeriod', cdk.validateString)(properties.cookieExpirationPeriod));
        errors.collect(cdk.propertyValidator('policyName', cdk.validateString)(properties.policyName));
        return errors.wrap('supplied properties not correct for "LBCookieStickinessPolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.LBCookieStickinessPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``LBCookieStickinessPolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.LBCookieStickinessPolicy`` resource.
     */
    function loadBalancerResourceLBCookieStickinessPolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_LBCookieStickinessPolicyPropertyValidator(properties).assertSuccess();
        return {
            CookieExpirationPeriod: cdk.stringToCloudFormation(properties.cookieExpirationPeriod),
            PolicyName: cdk.stringToCloudFormation(properties.policyName),
        };
    }
    /**
     * Determine whether the given properties match those of a ``ListenersProperty``
     *
     * @param properties the TypeScript properties of a ``ListenersProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_ListenersPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('instancePort', cdk.requiredValidator)(properties.instancePort));
        errors.collect(cdk.propertyValidator('instancePort', cdk.validateString)(properties.instancePort));
        errors.collect(cdk.propertyValidator('instanceProtocol', cdk.validateString)(properties.instanceProtocol));
        errors.collect(cdk.propertyValidator('loadBalancerPort', cdk.requiredValidator)(properties.loadBalancerPort));
        errors.collect(cdk.propertyValidator('loadBalancerPort', cdk.validateString)(properties.loadBalancerPort));
        errors.collect(cdk.propertyValidator('policyNames', cdk.listValidator(cdk.validateString))(properties.policyNames));
        errors.collect(cdk.propertyValidator('protocol', cdk.requiredValidator)(properties.protocol));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        errors.collect(cdk.propertyValidator('sslCertificateId', cdk.validateString)(properties.sslCertificateId));
        return errors.wrap('supplied properties not correct for "ListenersProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.Listeners`` resource
     *
     * @param properties the TypeScript properties of a ``ListenersProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.Listeners`` resource.
     */
    function loadBalancerResourceListenersPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_ListenersPropertyValidator(properties).assertSuccess();
        return {
            InstancePort: cdk.stringToCloudFormation(properties.instancePort),
            InstanceProtocol: cdk.stringToCloudFormation(properties.instanceProtocol),
            LoadBalancerPort: cdk.stringToCloudFormation(properties.loadBalancerPort),
            PolicyNames: cdk.listMapper(cdk.stringToCloudFormation)(properties.policyNames),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
            SSLCertificateId: cdk.stringToCloudFormation(properties.sslCertificateId),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PoliciesProperty``
     *
     * @param properties the TypeScript properties of a ``PoliciesProperty``
     *
     * @returns the result of the validation.
     */
    function LoadBalancerResource_PoliciesPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('attributes', cdk.requiredValidator)(properties.attributes));
        errors.collect(cdk.propertyValidator('attributes', cdk.listValidator(cdk.validateObject))(properties.attributes));
        errors.collect(cdk.propertyValidator('instancePorts', cdk.listValidator(cdk.validateString))(properties.instancePorts));
        errors.collect(cdk.propertyValidator('loadBalancerPorts', cdk.listValidator(cdk.validateString))(properties.loadBalancerPorts));
        errors.collect(cdk.propertyValidator('policyName', cdk.requiredValidator)(properties.policyName));
        errors.collect(cdk.propertyValidator('policyName', cdk.validateString)(properties.policyName));
        errors.collect(cdk.propertyValidator('policyType', cdk.requiredValidator)(properties.policyType));
        errors.collect(cdk.propertyValidator('policyType', cdk.validateString)(properties.policyType));
        return errors.wrap('supplied properties not correct for "PoliciesProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.Policies`` resource
     *
     * @param properties the TypeScript properties of a ``PoliciesProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ElasticLoadBalancing::LoadBalancer.Policies`` resource.
     */
    function loadBalancerResourcePoliciesPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LoadBalancerResource_PoliciesPropertyValidator(properties).assertSuccess();
        return {
            Attributes: cdk.listMapper(cdk.objectToCloudFormation)(properties.attributes),
            InstancePorts: cdk.listMapper(cdk.stringToCloudFormation)(properties.instancePorts),
            LoadBalancerPorts: cdk.listMapper(cdk.stringToCloudFormation)(properties.loadBalancerPorts),
            PolicyName: cdk.stringToCloudFormation(properties.policyName),
            PolicyType: cdk.stringToCloudFormation(properties.policyType),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
