// @aws-cdk/aws-ecs/lib/drain-hook/instance-drain-hook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const autoscaling = require("@aws-cdk/aws-autoscaling");
const iam = require("@aws-cdk/aws-iam");
const lambda = require("@aws-cdk/aws-lambda");
const sns = require("@aws-cdk/aws-sns");
const cdk = require("@aws-cdk/cdk");
const fs = require("fs");
const path = require("path");
/**
 * A hook to drain instances from ECS traffic before they're terminated
 */
class InstanceDrainHook extends cdk.Construct {
    constructor(parent, id, props) {
        super(parent, id);
        const drainTimeSeconds = props.drainTimeSec !== undefined ? props.drainTimeSec : 300;
        if (drainTimeSeconds < 0 || drainTimeSeconds > 900) {
            throw new Error(`Drain time must be between 0 and 900 seconds, got: ${drainTimeSeconds}`);
        }
        // Invoke Lambda via SNS Topic
        const topic = new sns.Topic(this, 'Topic');
        const fn = new lambda.Function(this, 'Function', {
            code: lambda.Code.inline(fs.readFileSync(path.join(__dirname, 'lambda-source', 'index.py'), { encoding: 'utf-8' })),
            handler: 'index.lambda_handler',
            runtime: lambda.Runtime.Python36,
            // Timeout: some extra margin for additional API calls made by the Lambda,
            // up to a maximum of 15 minutes.
            timeout: Math.min(drainTimeSeconds + 10, 900),
            environment: {
                CLUSTER: props.cluster.clusterName
            }
        });
        // Hook everything up: ASG -> Topic, Topic -> Lambda
        props.autoScalingGroup.onLifecycleTransition('DrainHook', {
            lifecycleTransition: autoscaling.LifecycleTransition.InstanceTerminating,
            defaultResult: autoscaling.DefaultResult.Continue,
            notificationTarget: topic,
            heartbeatTimeoutSec: drainTimeSeconds,
        });
        topic.subscribeLambda(fn);
        // FIXME: These should probably be restricted usefully in some way, but I don't exactly
        // know how.
        fn.addToRolePolicy(new iam.PolicyStatement()
            .addActions('autoscaling:CompleteLifecycleAction', 'ec2:DescribeInstances', 'ec2:DescribeInstanceAttribute', 'ec2:DescribeInstanceStatus', 'ec2:DescribeHosts')
            .addAllResources());
        // FIXME: These should be restricted to the ECS cluster probably, but I don't exactly
        // know how.
        fn.addToRolePolicy(new iam.PolicyStatement()
            .addActions('ecs:ListContainerInstances', 'ecs:SubmitContainerStateChange', 'ecs:SubmitTaskStateChange', 'ecs:DescribeContainerInstances', 'ecs:UpdateContainerInstancesState', 'ecs:ListTasks', 'ecs:DescribeTasks')
            .addAllResources());
    }
}
exports.InstanceDrainHook = InstanceDrainHook;
