// @aws-cdk/cdk/lib/cloudformation/dynamic-reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const construct_1 = require("../core/construct");
const tokens_1 = require("../core/tokens");
/**
 * References a dynamically retrieved value
 *
 * This is a Construct so that subclasses will (eventually) be able to attach
 * metadata to themselves without having to change call signatures.
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html
 */
class DynamicReference extends construct_1.Construct {
    constructor(parent, id, props) {
        super(parent, id);
        this._value = this.makeResolveValue(props.service, props.referenceKey);
    }
    /**
     * The value of this dynamic reference
     */
    get value() {
        return this._value;
    }
    /**
     * Make a dynamic reference Token value
     *
     * This is a value (similar to CDK Tokens) that will be substituted by
     * CloudFormation before executing the changeset.
     */
    makeResolveValue(service, referenceKey) {
        const resolveString = '{{resolve:' + service + ':' + referenceKey + '}}';
        // We don't strictly need to Tokenize a string here, but we do it anyway to be perfectly
        // clear that DynamicReference.value is unparseable in CDK apps.
        return new tokens_1.Token(resolveString).toString();
    }
}
exports.DynamicReference = DynamicReference;
/**
 * The service to retrieve the dynamic reference from
 */
var DynamicReferenceService;
(function (DynamicReferenceService) {
    /**
     * Plaintext value stored in AWS Systems Manager Parameter Store
     */
    DynamicReferenceService["Ssm"] = "ssm";
    /**
     * Secure string stored in AWS Systems Manager Parameter Store
     */
    DynamicReferenceService["SsmSecure"] = "ssm-secure";
    /**
     * Secret stored in AWS Secrets Manager
     */
    DynamicReferenceService["SecretsManager"] = "secretsmanager";
})(DynamicReferenceService = exports.DynamicReferenceService || (exports.DynamicReferenceService = {}));
