// @aws-cdk/aws-cloudformation/lib/cloudformation.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:06:23.737Z","fingerprint":"T+a06FzClZYjMljjkCpX2NEHkB16xbtWjBRnnRMAC6o="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``CustomResourceProps``
     *
     * @param properties the TypeScript properties of a ``CustomResourceProps``
     *
     * @returns the result of the validation.
     */
    function CustomResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('serviceToken', cdk.requiredValidator)(properties.serviceToken));
        errors.collect(cdk.propertyValidator('serviceToken', cdk.validateString)(properties.serviceToken));
        return errors.wrap('supplied properties not correct for "CustomResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CloudFormation::CustomResource`` resource
     *
     * @param properties the TypeScript properties of a ``CustomResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CloudFormation::CustomResource`` resource.
     */
    function customResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        CustomResourcePropsValidator(properties).assertSuccess();
        return {
            ServiceToken: cdk.stringToCloudFormation(properties.serviceToken),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cfn-customresource.html
     */
    class CustomResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CloudFormation::CustomResource``.
         *
         * @param parent   the ``cdk.Construct`` this ``CustomResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``CustomResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: CustomResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'serviceToken', this);
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return customResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    CustomResource.resourceTypeName = "AWS::CloudFormation::CustomResource";
    cloudformation.CustomResource = CustomResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``MacroResourceProps``
     *
     * @param properties the TypeScript properties of a ``MacroResourceProps``
     *
     * @returns the result of the validation.
     */
    function MacroResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
        errors.collect(cdk.propertyValidator('functionName', cdk.requiredValidator)(properties.functionName));
        errors.collect(cdk.propertyValidator('functionName', cdk.validateString)(properties.functionName));
        errors.collect(cdk.propertyValidator('logGroupName', cdk.validateString)(properties.logGroupName));
        errors.collect(cdk.propertyValidator('logRoleArn', cdk.validateString)(properties.logRoleArn));
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        return errors.wrap('supplied properties not correct for "MacroResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CloudFormation::Macro`` resource
     *
     * @param properties the TypeScript properties of a ``MacroResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CloudFormation::Macro`` resource.
     */
    function macroResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        MacroResourcePropsValidator(properties).assertSuccess();
        return {
            FunctionName: cdk.stringToCloudFormation(properties.functionName),
            Name: cdk.stringToCloudFormation(properties.name),
            Description: cdk.stringToCloudFormation(properties.description),
            LogGroupName: cdk.stringToCloudFormation(properties.logGroupName),
            LogRoleARN: cdk.stringToCloudFormation(properties.logRoleArn),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-macro.html
     */
    class MacroResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CloudFormation::Macro``.
         *
         * @param parent   the ``cdk.Construct`` this ``MacroResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``MacroResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: MacroResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'functionName', this);
            cdk.requireProperty(properties, 'name', this);
            this.macroName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return macroResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    MacroResource.resourceTypeName = "AWS::CloudFormation::Macro";
    cloudformation.MacroResource = MacroResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``StackResourceProps``
     *
     * @param properties the TypeScript properties of a ``StackResourceProps``
     *
     * @returns the result of the validation.
     */
    function StackResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('notificationArns', cdk.listValidator(cdk.validateString))(properties.notificationArns));
        errors.collect(cdk.propertyValidator('parameters', cdk.hashValidator(cdk.validateString))(properties.parameters));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateTag))(properties.tags));
        errors.collect(cdk.propertyValidator('templateUrl', cdk.requiredValidator)(properties.templateUrl));
        errors.collect(cdk.propertyValidator('templateUrl', cdk.validateString)(properties.templateUrl));
        errors.collect(cdk.propertyValidator('timeoutInMinutes', cdk.validateNumber)(properties.timeoutInMinutes));
        return errors.wrap('supplied properties not correct for "StackResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CloudFormation::Stack`` resource
     *
     * @param properties the TypeScript properties of a ``StackResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CloudFormation::Stack`` resource.
     */
    function stackResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        StackResourcePropsValidator(properties).assertSuccess();
        return {
            TemplateURL: cdk.stringToCloudFormation(properties.templateUrl),
            NotificationARNs: cdk.listMapper(cdk.stringToCloudFormation)(properties.notificationArns),
            Parameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.parameters),
            Tags: cdk.listMapper(cdk.tagToCloudFormation)(properties.tags),
            TimeoutInMinutes: cdk.numberToCloudFormation(properties.timeoutInMinutes),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-stack.html
     */
    class StackResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CloudFormation::Stack``.
         *
         * @param parent   the ``cdk.Construct`` this ``StackResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``StackResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: StackResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'templateUrl', this);
            this.stackId = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return stackResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    StackResource.resourceTypeName = "AWS::CloudFormation::Stack";
    cloudformation.StackResource = StackResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``WaitConditionResourceProps``
     *
     * @param properties the TypeScript properties of a ``WaitConditionResourceProps``
     *
     * @returns the result of the validation.
     */
    function WaitConditionResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('count', cdk.validateNumber)(properties.count));
        errors.collect(cdk.propertyValidator('handle', cdk.validateString)(properties.handle));
        errors.collect(cdk.propertyValidator('timeout', cdk.validateString)(properties.timeout));
        return errors.wrap('supplied properties not correct for "WaitConditionResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CloudFormation::WaitCondition`` resource
     *
     * @param properties the TypeScript properties of a ``WaitConditionResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CloudFormation::WaitCondition`` resource.
     */
    function waitConditionResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        WaitConditionResourcePropsValidator(properties).assertSuccess();
        return {
            Count: cdk.numberToCloudFormation(properties.count),
            Handle: cdk.stringToCloudFormation(properties.handle),
            Timeout: cdk.stringToCloudFormation(properties.timeout),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-waitcondition.html
     */
    class WaitConditionResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CloudFormation::WaitCondition``.
         *
         * @param parent   the ``cdk.Construct`` this ``WaitConditionResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``WaitConditionResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: WaitConditionResource.resourceTypeName, properties });
            this.waitConditionData = new WaitConditionData(this.getAtt('Data'));
            this.waitConditionName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return waitConditionResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    WaitConditionResource.resourceTypeName = "AWS::CloudFormation::WaitCondition";
    cloudformation.WaitConditionResource = WaitConditionResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
class WaitConditionData extends cdk.Token {
    constructor() {
        super(...arguments);
        // @ts-ignore: private but unused on purpose.
        this.thisIsAWaitConditionData = true;
    }
}
exports.WaitConditionData = WaitConditionData;
(function (cloudformation) {
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-waitconditionhandle.html
     */
    class WaitConditionHandleResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CloudFormation::WaitConditionHandle``.
         *
         * @param parent   the ``cdk.Construct`` this ``WaitConditionHandleResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``WaitConditionHandleResource``
         */
        constructor(parent, name) {
            super(parent, name, { type: WaitConditionHandleResource.resourceTypeName });
            this.waitConditionHandleUrl = this.ref.toString();
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    WaitConditionHandleResource.resourceTypeName = "AWS::CloudFormation::WaitConditionHandle";
    cloudformation.WaitConditionHandleResource = WaitConditionHandleResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
