// @aws-cdk/aws-ecs/lib/fargate/fargate-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_service_1 = require("../base/base-service");
const util_1 = require("../util");
/**
 * Start a service on an ECS cluster
 */
class FargateService extends base_service_1.BaseService {
    constructor(parent, name, props) {
        if (!util_1.isFargateCompatible(props.taskDefinition.compatibility)) {
            throw new Error('Supplied TaskDefinition is not configured for compatibility with Fargate');
        }
        super(parent, name, Object.assign({}, props, { desiredCount: props.desiredCount !== undefined ? props.desiredCount : 1 }), {
            cluster: props.cluster.clusterName,
            taskDefinition: props.taskDefinition.taskDefinitionArn,
            launchType: 'FARGATE',
            platformVersion: props.platformVersion,
        }, props.cluster.clusterName, props.taskDefinition);
        this.configureAwsVpcNetworking(props.cluster.vpc, props.assignPublicIp, props.vpcPlacement, props.securityGroup);
        if (!props.taskDefinition.defaultContainer) {
            throw new Error('A TaskDefinition must have at least one essential container');
        }
    }
}
exports.FargateService = FargateService;
/**
 * Fargate platform version
 *
 * @see https://docs.aws.amazon.com/AmazonECS/latest/developerguide/platform_versions.html
 */
var FargatePlatformVersion;
(function (FargatePlatformVersion) {
    /**
     * The latest, recommended platform version
     */
    FargatePlatformVersion["Latest"] = "LATEST";
    /**
     * Version 1.2
     *
     * Supports private registries.
     */
    FargatePlatformVersion["Version1_2"] = "1.2.0";
    /**
     * Version 1.1.0
     *
     * Supports task metadata, health checks, service discovery.
     */
    FargatePlatformVersion["Version1_1"] = "1.1.0";
    /**
     * Initial release
     *
     * Based on Amazon Linux 2017.09.
     */
    FargatePlatformVersion["Version1_0"] = "1.0.0";
})(FargatePlatformVersion = exports.FargatePlatformVersion || (exports.FargatePlatformVersion = {}));
