// @aws-cdk/aws-applicationautoscaling/lib/applicationautoscaling.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:03:58.430Z","fingerprint":"Uf2i85k3zU4Vd4cQh+yNUr0gJQK8Rp0DyQghHnPjJnQ="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ScalableTargetResourceProps``
     *
     * @param properties the TypeScript properties of a ``ScalableTargetResourceProps``
     *
     * @returns the result of the validation.
     */
    function ScalableTargetResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('maxCapacity', cdk.requiredValidator)(properties.maxCapacity));
        errors.collect(cdk.propertyValidator('maxCapacity', cdk.validateNumber)(properties.maxCapacity));
        errors.collect(cdk.propertyValidator('minCapacity', cdk.requiredValidator)(properties.minCapacity));
        errors.collect(cdk.propertyValidator('minCapacity', cdk.validateNumber)(properties.minCapacity));
        errors.collect(cdk.propertyValidator('resourceId', cdk.requiredValidator)(properties.resourceId));
        errors.collect(cdk.propertyValidator('resourceId', cdk.validateString)(properties.resourceId));
        errors.collect(cdk.propertyValidator('roleArn', cdk.requiredValidator)(properties.roleArn));
        errors.collect(cdk.propertyValidator('roleArn', cdk.validateString)(properties.roleArn));
        errors.collect(cdk.propertyValidator('scalableDimension', cdk.requiredValidator)(properties.scalableDimension));
        errors.collect(cdk.propertyValidator('scalableDimension', cdk.validateString)(properties.scalableDimension));
        errors.collect(cdk.propertyValidator('scheduledActions', cdk.listValidator(ScalableTargetResource_ScheduledActionPropertyValidator))(properties.scheduledActions));
        errors.collect(cdk.propertyValidator('serviceNamespace', cdk.requiredValidator)(properties.serviceNamespace));
        errors.collect(cdk.propertyValidator('serviceNamespace', cdk.validateString)(properties.serviceNamespace));
        return errors.wrap('supplied properties not correct for "ScalableTargetResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget`` resource
     *
     * @param properties the TypeScript properties of a ``ScalableTargetResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget`` resource.
     */
    function scalableTargetResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalableTargetResourcePropsValidator(properties).assertSuccess();
        return {
            MaxCapacity: cdk.numberToCloudFormation(properties.maxCapacity),
            MinCapacity: cdk.numberToCloudFormation(properties.minCapacity),
            ResourceId: cdk.stringToCloudFormation(properties.resourceId),
            RoleARN: cdk.stringToCloudFormation(properties.roleArn),
            ScalableDimension: cdk.stringToCloudFormation(properties.scalableDimension),
            ServiceNamespace: cdk.stringToCloudFormation(properties.serviceNamespace),
            ScheduledActions: cdk.listMapper(scalableTargetResourceScheduledActionPropertyToCloudFormation)(properties.scheduledActions),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-applicationautoscaling-scalabletarget.html
     */
    class ScalableTargetResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ApplicationAutoScaling::ScalableTarget``.
         *
         * @param parent   the ``cdk.Construct`` this ``ScalableTargetResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ScalableTargetResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ScalableTargetResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'maxCapacity', this);
            cdk.requireProperty(properties, 'minCapacity', this);
            cdk.requireProperty(properties, 'resourceId', this);
            cdk.requireProperty(properties, 'roleArn', this);
            cdk.requireProperty(properties, 'scalableDimension', this);
            cdk.requireProperty(properties, 'serviceNamespace', this);
            this.scalableTargetId = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return scalableTargetResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ScalableTargetResource.resourceTypeName = "AWS::ApplicationAutoScaling::ScalableTarget";
    cloudformation.ScalableTargetResource = ScalableTargetResource;
    /**
     * Determine whether the given properties match those of a ``ScalableTargetActionProperty``
     *
     * @param properties the TypeScript properties of a ``ScalableTargetActionProperty``
     *
     * @returns the result of the validation.
     */
    function ScalableTargetResource_ScalableTargetActionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('maxCapacity', cdk.validateNumber)(properties.maxCapacity));
        errors.collect(cdk.propertyValidator('minCapacity', cdk.validateNumber)(properties.minCapacity));
        return errors.wrap('supplied properties not correct for "ScalableTargetActionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget.ScalableTargetAction`` resource
     *
     * @param properties the TypeScript properties of a ``ScalableTargetActionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget.ScalableTargetAction`` resource.
     */
    function scalableTargetResourceScalableTargetActionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalableTargetResource_ScalableTargetActionPropertyValidator(properties).assertSuccess();
        return {
            MaxCapacity: cdk.numberToCloudFormation(properties.maxCapacity),
            MinCapacity: cdk.numberToCloudFormation(properties.minCapacity),
        };
    }
    /**
     * Determine whether the given properties match those of a ``ScheduledActionProperty``
     *
     * @param properties the TypeScript properties of a ``ScheduledActionProperty``
     *
     * @returns the result of the validation.
     */
    function ScalableTargetResource_ScheduledActionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('endTime', cdk.validateDate)(properties.endTime));
        errors.collect(cdk.propertyValidator('scalableTargetAction', ScalableTargetResource_ScalableTargetActionPropertyValidator)(properties.scalableTargetAction));
        errors.collect(cdk.propertyValidator('schedule', cdk.requiredValidator)(properties.schedule));
        errors.collect(cdk.propertyValidator('schedule', cdk.validateString)(properties.schedule));
        errors.collect(cdk.propertyValidator('scheduledActionName', cdk.requiredValidator)(properties.scheduledActionName));
        errors.collect(cdk.propertyValidator('scheduledActionName', cdk.validateString)(properties.scheduledActionName));
        errors.collect(cdk.propertyValidator('startTime', cdk.validateDate)(properties.startTime));
        return errors.wrap('supplied properties not correct for "ScheduledActionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget.ScheduledAction`` resource
     *
     * @param properties the TypeScript properties of a ``ScheduledActionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalableTarget.ScheduledAction`` resource.
     */
    function scalableTargetResourceScheduledActionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalableTargetResource_ScheduledActionPropertyValidator(properties).assertSuccess();
        return {
            EndTime: cdk.dateToCloudFormation(properties.endTime),
            ScalableTargetAction: scalableTargetResourceScalableTargetActionPropertyToCloudFormation(properties.scalableTargetAction),
            Schedule: cdk.stringToCloudFormation(properties.schedule),
            ScheduledActionName: cdk.stringToCloudFormation(properties.scheduledActionName),
            StartTime: cdk.dateToCloudFormation(properties.startTime),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ScalingPolicyResourceProps``
     *
     * @param properties the TypeScript properties of a ``ScalingPolicyResourceProps``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('policyName', cdk.requiredValidator)(properties.policyName));
        errors.collect(cdk.propertyValidator('policyName', cdk.validateString)(properties.policyName));
        errors.collect(cdk.propertyValidator('policyType', cdk.requiredValidator)(properties.policyType));
        errors.collect(cdk.propertyValidator('policyType', cdk.validateString)(properties.policyType));
        errors.collect(cdk.propertyValidator('resourceId', cdk.validateString)(properties.resourceId));
        errors.collect(cdk.propertyValidator('scalableDimension', cdk.validateString)(properties.scalableDimension));
        errors.collect(cdk.propertyValidator('scalingTargetId', cdk.validateString)(properties.scalingTargetId));
        errors.collect(cdk.propertyValidator('serviceNamespace', cdk.validateString)(properties.serviceNamespace));
        errors.collect(cdk.propertyValidator('stepScalingPolicyConfiguration', ScalingPolicyResource_StepScalingPolicyConfigurationPropertyValidator)(properties.stepScalingPolicyConfiguration));
        errors.collect(cdk.propertyValidator('targetTrackingScalingPolicyConfiguration', ScalingPolicyResource_TargetTrackingScalingPolicyConfigurationPropertyValidator)(properties.targetTrackingScalingPolicyConfiguration));
        return errors.wrap('supplied properties not correct for "ScalingPolicyResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``ScalingPolicyResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy`` resource.
     */
    function scalingPolicyResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResourcePropsValidator(properties).assertSuccess();
        return {
            PolicyName: cdk.stringToCloudFormation(properties.policyName),
            PolicyType: cdk.stringToCloudFormation(properties.policyType),
            ResourceId: cdk.stringToCloudFormation(properties.resourceId),
            ScalableDimension: cdk.stringToCloudFormation(properties.scalableDimension),
            ScalingTargetId: cdk.stringToCloudFormation(properties.scalingTargetId),
            ServiceNamespace: cdk.stringToCloudFormation(properties.serviceNamespace),
            StepScalingPolicyConfiguration: scalingPolicyResourceStepScalingPolicyConfigurationPropertyToCloudFormation(properties.stepScalingPolicyConfiguration),
            TargetTrackingScalingPolicyConfiguration: scalingPolicyResourceTargetTrackingScalingPolicyConfigurationPropertyToCloudFormation(properties.targetTrackingScalingPolicyConfiguration),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-applicationautoscaling-scalingpolicy.html
     */
    class ScalingPolicyResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ApplicationAutoScaling::ScalingPolicy``.
         *
         * @param parent   the ``cdk.Construct`` this ``ScalingPolicyResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ScalingPolicyResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ScalingPolicyResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'policyName', this);
            cdk.requireProperty(properties, 'policyType', this);
            this.scalingPolicyArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return scalingPolicyResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ScalingPolicyResource.resourceTypeName = "AWS::ApplicationAutoScaling::ScalingPolicy";
    cloudformation.ScalingPolicyResource = ScalingPolicyResource;
    /**
     * Determine whether the given properties match those of a ``CustomizedMetricSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``CustomizedMetricSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('dimensions', cdk.listValidator(ScalingPolicyResource_MetricDimensionPropertyValidator))(properties.dimensions));
        errors.collect(cdk.propertyValidator('metricName', cdk.requiredValidator)(properties.metricName));
        errors.collect(cdk.propertyValidator('metricName', cdk.validateString)(properties.metricName));
        errors.collect(cdk.propertyValidator('namespace', cdk.requiredValidator)(properties.namespace));
        errors.collect(cdk.propertyValidator('namespace', cdk.validateString)(properties.namespace));
        errors.collect(cdk.propertyValidator('statistic', cdk.requiredValidator)(properties.statistic));
        errors.collect(cdk.propertyValidator('statistic', cdk.validateString)(properties.statistic));
        errors.collect(cdk.propertyValidator('unit', cdk.validateString)(properties.unit));
        return errors.wrap('supplied properties not correct for "CustomizedMetricSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.CustomizedMetricSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``CustomizedMetricSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.CustomizedMetricSpecification`` resource.
     */
    function scalingPolicyResourceCustomizedMetricSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator(properties).assertSuccess();
        return {
            Dimensions: cdk.listMapper(scalingPolicyResourceMetricDimensionPropertyToCloudFormation)(properties.dimensions),
            MetricName: cdk.stringToCloudFormation(properties.metricName),
            Namespace: cdk.stringToCloudFormation(properties.namespace),
            Statistic: cdk.stringToCloudFormation(properties.statistic),
            Unit: cdk.stringToCloudFormation(properties.unit),
        };
    }
    /**
     * Determine whether the given properties match those of a ``MetricDimensionProperty``
     *
     * @param properties the TypeScript properties of a ``MetricDimensionProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_MetricDimensionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "MetricDimensionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.MetricDimension`` resource
     *
     * @param properties the TypeScript properties of a ``MetricDimensionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.MetricDimension`` resource.
     */
    function scalingPolicyResourceMetricDimensionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_MetricDimensionPropertyValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PredefinedMetricSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``PredefinedMetricSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('predefinedMetricType', cdk.requiredValidator)(properties.predefinedMetricType));
        errors.collect(cdk.propertyValidator('predefinedMetricType', cdk.validateString)(properties.predefinedMetricType));
        errors.collect(cdk.propertyValidator('resourceLabel', cdk.validateString)(properties.resourceLabel));
        return errors.wrap('supplied properties not correct for "PredefinedMetricSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.PredefinedMetricSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``PredefinedMetricSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.PredefinedMetricSpecification`` resource.
     */
    function scalingPolicyResourcePredefinedMetricSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator(properties).assertSuccess();
        return {
            PredefinedMetricType: cdk.stringToCloudFormation(properties.predefinedMetricType),
            ResourceLabel: cdk.stringToCloudFormation(properties.resourceLabel),
        };
    }
    /**
     * Determine whether the given properties match those of a ``StepAdjustmentProperty``
     *
     * @param properties the TypeScript properties of a ``StepAdjustmentProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_StepAdjustmentPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('metricIntervalLowerBound', cdk.validateNumber)(properties.metricIntervalLowerBound));
        errors.collect(cdk.propertyValidator('metricIntervalUpperBound', cdk.validateNumber)(properties.metricIntervalUpperBound));
        errors.collect(cdk.propertyValidator('scalingAdjustment', cdk.requiredValidator)(properties.scalingAdjustment));
        errors.collect(cdk.propertyValidator('scalingAdjustment', cdk.validateNumber)(properties.scalingAdjustment));
        return errors.wrap('supplied properties not correct for "StepAdjustmentProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.StepAdjustment`` resource
     *
     * @param properties the TypeScript properties of a ``StepAdjustmentProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.StepAdjustment`` resource.
     */
    function scalingPolicyResourceStepAdjustmentPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_StepAdjustmentPropertyValidator(properties).assertSuccess();
        return {
            MetricIntervalLowerBound: cdk.numberToCloudFormation(properties.metricIntervalLowerBound),
            MetricIntervalUpperBound: cdk.numberToCloudFormation(properties.metricIntervalUpperBound),
            ScalingAdjustment: cdk.numberToCloudFormation(properties.scalingAdjustment),
        };
    }
    /**
     * Determine whether the given properties match those of a ``StepScalingPolicyConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``StepScalingPolicyConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_StepScalingPolicyConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('adjustmentType', cdk.validateString)(properties.adjustmentType));
        errors.collect(cdk.propertyValidator('cooldown', cdk.validateNumber)(properties.cooldown));
        errors.collect(cdk.propertyValidator('metricAggregationType', cdk.validateString)(properties.metricAggregationType));
        errors.collect(cdk.propertyValidator('minAdjustmentMagnitude', cdk.validateNumber)(properties.minAdjustmentMagnitude));
        errors.collect(cdk.propertyValidator('stepAdjustments', cdk.listValidator(ScalingPolicyResource_StepAdjustmentPropertyValidator))(properties.stepAdjustments));
        return errors.wrap('supplied properties not correct for "StepScalingPolicyConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.StepScalingPolicyConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``StepScalingPolicyConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.StepScalingPolicyConfiguration`` resource.
     */
    function scalingPolicyResourceStepScalingPolicyConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_StepScalingPolicyConfigurationPropertyValidator(properties).assertSuccess();
        return {
            AdjustmentType: cdk.stringToCloudFormation(properties.adjustmentType),
            Cooldown: cdk.numberToCloudFormation(properties.cooldown),
            MetricAggregationType: cdk.stringToCloudFormation(properties.metricAggregationType),
            MinAdjustmentMagnitude: cdk.numberToCloudFormation(properties.minAdjustmentMagnitude),
            StepAdjustments: cdk.listMapper(scalingPolicyResourceStepAdjustmentPropertyToCloudFormation)(properties.stepAdjustments),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TargetTrackingScalingPolicyConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``TargetTrackingScalingPolicyConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_TargetTrackingScalingPolicyConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('customizedMetricSpecification', ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator)(properties.customizedMetricSpecification));
        errors.collect(cdk.propertyValidator('disableScaleIn', cdk.validateBoolean)(properties.disableScaleIn));
        errors.collect(cdk.propertyValidator('predefinedMetricSpecification', ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator)(properties.predefinedMetricSpecification));
        errors.collect(cdk.propertyValidator('scaleInCooldown', cdk.validateNumber)(properties.scaleInCooldown));
        errors.collect(cdk.propertyValidator('scaleOutCooldown', cdk.validateNumber)(properties.scaleOutCooldown));
        errors.collect(cdk.propertyValidator('targetValue', cdk.requiredValidator)(properties.targetValue));
        errors.collect(cdk.propertyValidator('targetValue', cdk.validateNumber)(properties.targetValue));
        return errors.wrap('supplied properties not correct for "TargetTrackingScalingPolicyConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.TargetTrackingScalingPolicyConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``TargetTrackingScalingPolicyConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ApplicationAutoScaling::ScalingPolicy.TargetTrackingScalingPolicyConfiguration`` resource.
     */
    function scalingPolicyResourceTargetTrackingScalingPolicyConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_TargetTrackingScalingPolicyConfigurationPropertyValidator(properties).assertSuccess();
        return {
            CustomizedMetricSpecification: scalingPolicyResourceCustomizedMetricSpecificationPropertyToCloudFormation(properties.customizedMetricSpecification),
            DisableScaleIn: cdk.booleanToCloudFormation(properties.disableScaleIn),
            PredefinedMetricSpecification: scalingPolicyResourcePredefinedMetricSpecificationPropertyToCloudFormation(properties.predefinedMetricSpecification),
            ScaleInCooldown: cdk.numberToCloudFormation(properties.scaleInCooldown),
            ScaleOutCooldown: cdk.numberToCloudFormation(properties.scaleOutCooldown),
            TargetValue: cdk.numberToCloudFormation(properties.targetValue),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
