// @aws-cdk/aws-certificatemanager/lib/certificatemanager.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:03:40.854Z","fingerprint":"yz+Ec6mawF6ucegPv9WpmCZ7NLTBmDPXnnKuWZuXr/8="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``CertificateResourceProps``
     *
     * @param properties the TypeScript properties of a ``CertificateResourceProps``
     *
     * @returns the result of the validation.
     */
    function CertificateResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('domainName', cdk.requiredValidator)(properties.domainName));
        errors.collect(cdk.propertyValidator('domainName', cdk.validateString)(properties.domainName));
        errors.collect(cdk.propertyValidator('domainValidationOptions', cdk.listValidator(CertificateResource_DomainValidationOptionPropertyValidator))(properties.domainValidationOptions));
        errors.collect(cdk.propertyValidator('subjectAlternativeNames', cdk.listValidator(cdk.validateString))(properties.subjectAlternativeNames));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(cdk.validateTag))(properties.tags));
        errors.collect(cdk.propertyValidator('validationMethod', cdk.validateString)(properties.validationMethod));
        return errors.wrap('supplied properties not correct for "CertificateResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CertificateManager::Certificate`` resource
     *
     * @param properties the TypeScript properties of a ``CertificateResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CertificateManager::Certificate`` resource.
     */
    function certificateResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        CertificateResourcePropsValidator(properties).assertSuccess();
        return {
            DomainName: cdk.stringToCloudFormation(properties.domainName),
            DomainValidationOptions: cdk.listMapper(certificateResourceDomainValidationOptionPropertyToCloudFormation)(properties.domainValidationOptions),
            SubjectAlternativeNames: cdk.listMapper(cdk.stringToCloudFormation)(properties.subjectAlternativeNames),
            Tags: cdk.listMapper(cdk.tagToCloudFormation)(properties.tags),
            ValidationMethod: cdk.stringToCloudFormation(properties.validationMethod),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-certificatemanager-certificate.html
     */
    class CertificateResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::CertificateManager::Certificate``.
         *
         * @param parent   the ``cdk.Construct`` this ``CertificateResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``CertificateResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: CertificateResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'domainName', this);
            this.certificateArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return certificateResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    CertificateResource.resourceTypeName = "AWS::CertificateManager::Certificate";
    cloudformation.CertificateResource = CertificateResource;
    /**
     * Determine whether the given properties match those of a ``DomainValidationOptionProperty``
     *
     * @param properties the TypeScript properties of a ``DomainValidationOptionProperty``
     *
     * @returns the result of the validation.
     */
    function CertificateResource_DomainValidationOptionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('domainName', cdk.requiredValidator)(properties.domainName));
        errors.collect(cdk.propertyValidator('domainName', cdk.validateString)(properties.domainName));
        errors.collect(cdk.propertyValidator('validationDomain', cdk.requiredValidator)(properties.validationDomain));
        errors.collect(cdk.propertyValidator('validationDomain', cdk.validateString)(properties.validationDomain));
        return errors.wrap('supplied properties not correct for "DomainValidationOptionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::CertificateManager::Certificate.DomainValidationOption`` resource
     *
     * @param properties the TypeScript properties of a ``DomainValidationOptionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::CertificateManager::Certificate.DomainValidationOption`` resource.
     */
    function certificateResourceDomainValidationOptionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        CertificateResource_DomainValidationOptionPropertyValidator(properties).assertSuccess();
        return {
            DomainName: cdk.stringToCloudFormation(properties.domainName),
            ValidationDomain: cdk.stringToCloudFormation(properties.validationDomain),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
