// @aws-cdk/aws-ecs/lib/ecs.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:08:59.357Z","fingerprint":"39CKOrv7pBeqnbg5NQJq4CuiIpRgD2T2eoPcnSzTT5k="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ClusterResourceProps``
     *
     * @param properties the TypeScript properties of a ``ClusterResourceProps``
     *
     * @returns the result of the validation.
     */
    function ClusterResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('clusterName', cdk.validateString)(properties.clusterName));
        return errors.wrap('supplied properties not correct for "ClusterResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Cluster`` resource
     *
     * @param properties the TypeScript properties of a ``ClusterResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Cluster`` resource.
     */
    function clusterResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ClusterResourcePropsValidator(properties).assertSuccess();
        return {
            ClusterName: cdk.stringToCloudFormation(properties.clusterName),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-cluster.html
     */
    class ClusterResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ECS::Cluster``.
         *
         * @param parent   the ``cdk.Construct`` this ``ClusterResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ClusterResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ClusterResource.resourceTypeName, properties });
            this.clusterArn = this.getAtt('Arn').toString();
            this.clusterName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return clusterResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ClusterResource.resourceTypeName = "AWS::ECS::Cluster";
    cloudformation.ClusterResource = ClusterResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ServiceResourceProps``
     *
     * @param properties the TypeScript properties of a ``ServiceResourceProps``
     *
     * @returns the result of the validation.
     */
    function ServiceResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('cluster', cdk.validateString)(properties.cluster));
        errors.collect(cdk.propertyValidator('deploymentConfiguration', ServiceResource_DeploymentConfigurationPropertyValidator)(properties.deploymentConfiguration));
        errors.collect(cdk.propertyValidator('desiredCount', cdk.validateNumber)(properties.desiredCount));
        errors.collect(cdk.propertyValidator('healthCheckGracePeriodSeconds', cdk.validateNumber)(properties.healthCheckGracePeriodSeconds));
        errors.collect(cdk.propertyValidator('launchType', cdk.validateString)(properties.launchType));
        errors.collect(cdk.propertyValidator('loadBalancers', cdk.listValidator(ServiceResource_LoadBalancerPropertyValidator))(properties.loadBalancers));
        errors.collect(cdk.propertyValidator('networkConfiguration', ServiceResource_NetworkConfigurationPropertyValidator)(properties.networkConfiguration));
        errors.collect(cdk.propertyValidator('placementConstraints', cdk.listValidator(ServiceResource_PlacementConstraintPropertyValidator))(properties.placementConstraints));
        errors.collect(cdk.propertyValidator('placementStrategies', cdk.listValidator(ServiceResource_PlacementStrategyPropertyValidator))(properties.placementStrategies));
        errors.collect(cdk.propertyValidator('platformVersion', cdk.validateString)(properties.platformVersion));
        errors.collect(cdk.propertyValidator('role', cdk.validateString)(properties.role));
        errors.collect(cdk.propertyValidator('schedulingStrategy', cdk.validateString)(properties.schedulingStrategy));
        errors.collect(cdk.propertyValidator('serviceName', cdk.validateString)(properties.serviceName));
        errors.collect(cdk.propertyValidator('serviceRegistries', cdk.listValidator(ServiceResource_ServiceRegistryPropertyValidator))(properties.serviceRegistries));
        errors.collect(cdk.propertyValidator('taskDefinition', cdk.requiredValidator)(properties.taskDefinition));
        errors.collect(cdk.propertyValidator('taskDefinition', cdk.validateString)(properties.taskDefinition));
        return errors.wrap('supplied properties not correct for "ServiceResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service`` resource
     *
     * @param properties the TypeScript properties of a ``ServiceResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service`` resource.
     */
    function serviceResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResourcePropsValidator(properties).assertSuccess();
        return {
            TaskDefinition: cdk.stringToCloudFormation(properties.taskDefinition),
            Cluster: cdk.stringToCloudFormation(properties.cluster),
            DeploymentConfiguration: serviceResourceDeploymentConfigurationPropertyToCloudFormation(properties.deploymentConfiguration),
            DesiredCount: cdk.numberToCloudFormation(properties.desiredCount),
            HealthCheckGracePeriodSeconds: cdk.numberToCloudFormation(properties.healthCheckGracePeriodSeconds),
            LaunchType: cdk.stringToCloudFormation(properties.launchType),
            LoadBalancers: cdk.listMapper(serviceResourceLoadBalancerPropertyToCloudFormation)(properties.loadBalancers),
            NetworkConfiguration: serviceResourceNetworkConfigurationPropertyToCloudFormation(properties.networkConfiguration),
            PlacementConstraints: cdk.listMapper(serviceResourcePlacementConstraintPropertyToCloudFormation)(properties.placementConstraints),
            PlacementStrategies: cdk.listMapper(serviceResourcePlacementStrategyPropertyToCloudFormation)(properties.placementStrategies),
            PlatformVersion: cdk.stringToCloudFormation(properties.platformVersion),
            Role: cdk.stringToCloudFormation(properties.role),
            SchedulingStrategy: cdk.stringToCloudFormation(properties.schedulingStrategy),
            ServiceName: cdk.stringToCloudFormation(properties.serviceName),
            ServiceRegistries: cdk.listMapper(serviceResourceServiceRegistryPropertyToCloudFormation)(properties.serviceRegistries),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-service.html
     */
    class ServiceResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ECS::Service``.
         *
         * @param parent   the ``cdk.Construct`` this ``ServiceResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ServiceResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ServiceResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'taskDefinition', this);
            this.serviceName = this.getAtt('Name').toString();
            this.serviceArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return serviceResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ServiceResource.resourceTypeName = "AWS::ECS::Service";
    cloudformation.ServiceResource = ServiceResource;
    /**
     * Determine whether the given properties match those of a ``AwsVpcConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``AwsVpcConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_AwsVpcConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('assignPublicIp', cdk.validateString)(properties.assignPublicIp));
        errors.collect(cdk.propertyValidator('securityGroups', cdk.listValidator(cdk.validateString))(properties.securityGroups));
        errors.collect(cdk.propertyValidator('subnets', cdk.requiredValidator)(properties.subnets));
        errors.collect(cdk.propertyValidator('subnets', cdk.listValidator(cdk.validateString))(properties.subnets));
        return errors.wrap('supplied properties not correct for "AwsVpcConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.AwsVpcConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``AwsVpcConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.AwsVpcConfiguration`` resource.
     */
    function serviceResourceAwsVpcConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_AwsVpcConfigurationPropertyValidator(properties).assertSuccess();
        return {
            AssignPublicIp: cdk.stringToCloudFormation(properties.assignPublicIp),
            SecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroups),
            Subnets: cdk.listMapper(cdk.stringToCloudFormation)(properties.subnets),
        };
    }
    /**
     * Determine whether the given properties match those of a ``DeploymentConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``DeploymentConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_DeploymentConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('maximumPercent', cdk.validateNumber)(properties.maximumPercent));
        errors.collect(cdk.propertyValidator('minimumHealthyPercent', cdk.validateNumber)(properties.minimumHealthyPercent));
        return errors.wrap('supplied properties not correct for "DeploymentConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.DeploymentConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``DeploymentConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.DeploymentConfiguration`` resource.
     */
    function serviceResourceDeploymentConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_DeploymentConfigurationPropertyValidator(properties).assertSuccess();
        return {
            MaximumPercent: cdk.numberToCloudFormation(properties.maximumPercent),
            MinimumHealthyPercent: cdk.numberToCloudFormation(properties.minimumHealthyPercent),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LoadBalancerProperty``
     *
     * @param properties the TypeScript properties of a ``LoadBalancerProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_LoadBalancerPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerName', cdk.validateString)(properties.containerName));
        errors.collect(cdk.propertyValidator('containerPort', cdk.requiredValidator)(properties.containerPort));
        errors.collect(cdk.propertyValidator('containerPort', cdk.validateNumber)(properties.containerPort));
        errors.collect(cdk.propertyValidator('loadBalancerName', cdk.validateString)(properties.loadBalancerName));
        errors.collect(cdk.propertyValidator('targetGroupArn', cdk.validateString)(properties.targetGroupArn));
        return errors.wrap('supplied properties not correct for "LoadBalancerProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.LoadBalancer`` resource
     *
     * @param properties the TypeScript properties of a ``LoadBalancerProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.LoadBalancer`` resource.
     */
    function serviceResourceLoadBalancerPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_LoadBalancerPropertyValidator(properties).assertSuccess();
        return {
            ContainerName: cdk.stringToCloudFormation(properties.containerName),
            ContainerPort: cdk.numberToCloudFormation(properties.containerPort),
            LoadBalancerName: cdk.stringToCloudFormation(properties.loadBalancerName),
            TargetGroupArn: cdk.stringToCloudFormation(properties.targetGroupArn),
        };
    }
    /**
     * Determine whether the given properties match those of a ``NetworkConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``NetworkConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_NetworkConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('awsvpcConfiguration', ServiceResource_AwsVpcConfigurationPropertyValidator)(properties.awsvpcConfiguration));
        return errors.wrap('supplied properties not correct for "NetworkConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.NetworkConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``NetworkConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.NetworkConfiguration`` resource.
     */
    function serviceResourceNetworkConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_NetworkConfigurationPropertyValidator(properties).assertSuccess();
        return {
            AwsvpcConfiguration: serviceResourceAwsVpcConfigurationPropertyToCloudFormation(properties.awsvpcConfiguration),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PlacementConstraintProperty``
     *
     * @param properties the TypeScript properties of a ``PlacementConstraintProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_PlacementConstraintPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('expression', cdk.validateString)(properties.expression));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "PlacementConstraintProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.PlacementConstraint`` resource
     *
     * @param properties the TypeScript properties of a ``PlacementConstraintProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.PlacementConstraint`` resource.
     */
    function serviceResourcePlacementConstraintPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_PlacementConstraintPropertyValidator(properties).assertSuccess();
        return {
            Expression: cdk.stringToCloudFormation(properties.expression),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PlacementStrategyProperty``
     *
     * @param properties the TypeScript properties of a ``PlacementStrategyProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_PlacementStrategyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('field', cdk.validateString)(properties.field));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "PlacementStrategyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.PlacementStrategy`` resource
     *
     * @param properties the TypeScript properties of a ``PlacementStrategyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.PlacementStrategy`` resource.
     */
    function serviceResourcePlacementStrategyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_PlacementStrategyPropertyValidator(properties).assertSuccess();
        return {
            Field: cdk.stringToCloudFormation(properties.field),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``ServiceRegistryProperty``
     *
     * @param properties the TypeScript properties of a ``ServiceRegistryProperty``
     *
     * @returns the result of the validation.
     */
    function ServiceResource_ServiceRegistryPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerName', cdk.validateString)(properties.containerName));
        errors.collect(cdk.propertyValidator('containerPort', cdk.validateNumber)(properties.containerPort));
        errors.collect(cdk.propertyValidator('port', cdk.validateNumber)(properties.port));
        errors.collect(cdk.propertyValidator('registryArn', cdk.validateString)(properties.registryArn));
        return errors.wrap('supplied properties not correct for "ServiceRegistryProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::Service.ServiceRegistry`` resource
     *
     * @param properties the TypeScript properties of a ``ServiceRegistryProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::Service.ServiceRegistry`` resource.
     */
    function serviceResourceServiceRegistryPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ServiceResource_ServiceRegistryPropertyValidator(properties).assertSuccess();
        return {
            ContainerName: cdk.stringToCloudFormation(properties.containerName),
            ContainerPort: cdk.numberToCloudFormation(properties.containerPort),
            Port: cdk.numberToCloudFormation(properties.port),
            RegistryArn: cdk.stringToCloudFormation(properties.registryArn),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``TaskDefinitionResourceProps``
     *
     * @param properties the TypeScript properties of a ``TaskDefinitionResourceProps``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerDefinitions', cdk.listValidator(TaskDefinitionResource_ContainerDefinitionPropertyValidator))(properties.containerDefinitions));
        errors.collect(cdk.propertyValidator('cpu', cdk.validateString)(properties.cpu));
        errors.collect(cdk.propertyValidator('executionRoleArn', cdk.validateString)(properties.executionRoleArn));
        errors.collect(cdk.propertyValidator('family', cdk.validateString)(properties.family));
        errors.collect(cdk.propertyValidator('memory', cdk.validateString)(properties.memory));
        errors.collect(cdk.propertyValidator('networkMode', cdk.validateString)(properties.networkMode));
        errors.collect(cdk.propertyValidator('placementConstraints', cdk.listValidator(TaskDefinitionResource_TaskDefinitionPlacementConstraintPropertyValidator))(properties.placementConstraints));
        errors.collect(cdk.propertyValidator('requiresCompatibilities', cdk.listValidator(cdk.validateString))(properties.requiresCompatibilities));
        errors.collect(cdk.propertyValidator('taskRoleArn', cdk.validateString)(properties.taskRoleArn));
        errors.collect(cdk.propertyValidator('volumes', cdk.listValidator(TaskDefinitionResource_VolumePropertyValidator))(properties.volumes));
        return errors.wrap('supplied properties not correct for "TaskDefinitionResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition`` resource
     *
     * @param properties the TypeScript properties of a ``TaskDefinitionResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition`` resource.
     */
    function taskDefinitionResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResourcePropsValidator(properties).assertSuccess();
        return {
            ContainerDefinitions: cdk.listMapper(taskDefinitionResourceContainerDefinitionPropertyToCloudFormation)(properties.containerDefinitions),
            Cpu: cdk.stringToCloudFormation(properties.cpu),
            ExecutionRoleArn: cdk.stringToCloudFormation(properties.executionRoleArn),
            Family: cdk.stringToCloudFormation(properties.family),
            Memory: cdk.stringToCloudFormation(properties.memory),
            NetworkMode: cdk.stringToCloudFormation(properties.networkMode),
            PlacementConstraints: cdk.listMapper(taskDefinitionResourceTaskDefinitionPlacementConstraintPropertyToCloudFormation)(properties.placementConstraints),
            RequiresCompatibilities: cdk.listMapper(cdk.stringToCloudFormation)(properties.requiresCompatibilities),
            TaskRoleArn: cdk.stringToCloudFormation(properties.taskRoleArn),
            Volumes: cdk.listMapper(taskDefinitionResourceVolumePropertyToCloudFormation)(properties.volumes),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskdefinition.html
     */
    class TaskDefinitionResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ECS::TaskDefinition``.
         *
         * @param parent   the ``cdk.Construct`` this ``TaskDefinitionResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``TaskDefinitionResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: TaskDefinitionResource.resourceTypeName, properties });
            this.taskDefinitionArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return taskDefinitionResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    TaskDefinitionResource.resourceTypeName = "AWS::ECS::TaskDefinition";
    cloudformation.TaskDefinitionResource = TaskDefinitionResource;
    /**
     * Determine whether the given properties match those of a ``ContainerDefinitionProperty``
     *
     * @param properties the TypeScript properties of a ``ContainerDefinitionProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_ContainerDefinitionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('command', cdk.listValidator(cdk.validateString))(properties.command));
        errors.collect(cdk.propertyValidator('cpu', cdk.validateNumber)(properties.cpu));
        errors.collect(cdk.propertyValidator('disableNetworking', cdk.validateBoolean)(properties.disableNetworking));
        errors.collect(cdk.propertyValidator('dnsSearchDomains', cdk.listValidator(cdk.validateString))(properties.dnsSearchDomains));
        errors.collect(cdk.propertyValidator('dnsServers', cdk.listValidator(cdk.validateString))(properties.dnsServers));
        errors.collect(cdk.propertyValidator('dockerLabels', cdk.hashValidator(cdk.validateString))(properties.dockerLabels));
        errors.collect(cdk.propertyValidator('dockerSecurityOptions', cdk.listValidator(cdk.validateString))(properties.dockerSecurityOptions));
        errors.collect(cdk.propertyValidator('entryPoint', cdk.listValidator(cdk.validateString))(properties.entryPoint));
        errors.collect(cdk.propertyValidator('environment', cdk.listValidator(TaskDefinitionResource_KeyValuePairPropertyValidator))(properties.environment));
        errors.collect(cdk.propertyValidator('essential', cdk.validateBoolean)(properties.essential));
        errors.collect(cdk.propertyValidator('extraHosts', cdk.listValidator(TaskDefinitionResource_HostEntryPropertyValidator))(properties.extraHosts));
        errors.collect(cdk.propertyValidator('healthCheck', TaskDefinitionResource_HealthCheckPropertyValidator)(properties.healthCheck));
        errors.collect(cdk.propertyValidator('hostname', cdk.validateString)(properties.hostname));
        errors.collect(cdk.propertyValidator('image', cdk.validateString)(properties.image));
        errors.collect(cdk.propertyValidator('links', cdk.listValidator(cdk.validateString))(properties.links));
        errors.collect(cdk.propertyValidator('linuxParameters', TaskDefinitionResource_LinuxParametersPropertyValidator)(properties.linuxParameters));
        errors.collect(cdk.propertyValidator('logConfiguration', TaskDefinitionResource_LogConfigurationPropertyValidator)(properties.logConfiguration));
        errors.collect(cdk.propertyValidator('memory', cdk.validateNumber)(properties.memory));
        errors.collect(cdk.propertyValidator('memoryReservation', cdk.validateNumber)(properties.memoryReservation));
        errors.collect(cdk.propertyValidator('mountPoints', cdk.listValidator(TaskDefinitionResource_MountPointPropertyValidator))(properties.mountPoints));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('portMappings', cdk.listValidator(TaskDefinitionResource_PortMappingPropertyValidator))(properties.portMappings));
        errors.collect(cdk.propertyValidator('privileged', cdk.validateBoolean)(properties.privileged));
        errors.collect(cdk.propertyValidator('readonlyRootFilesystem', cdk.validateBoolean)(properties.readonlyRootFilesystem));
        errors.collect(cdk.propertyValidator('repositoryCredentials', TaskDefinitionResource_RepositoryCredentialsPropertyValidator)(properties.repositoryCredentials));
        errors.collect(cdk.propertyValidator('ulimits', cdk.listValidator(TaskDefinitionResource_UlimitPropertyValidator))(properties.ulimits));
        errors.collect(cdk.propertyValidator('user', cdk.validateString)(properties.user));
        errors.collect(cdk.propertyValidator('volumesFrom', cdk.listValidator(TaskDefinitionResource_VolumeFromPropertyValidator))(properties.volumesFrom));
        errors.collect(cdk.propertyValidator('workingDirectory', cdk.validateString)(properties.workingDirectory));
        return errors.wrap('supplied properties not correct for "ContainerDefinitionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.ContainerDefinition`` resource
     *
     * @param properties the TypeScript properties of a ``ContainerDefinitionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.ContainerDefinition`` resource.
     */
    function taskDefinitionResourceContainerDefinitionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_ContainerDefinitionPropertyValidator(properties).assertSuccess();
        return {
            Command: cdk.listMapper(cdk.stringToCloudFormation)(properties.command),
            Cpu: cdk.numberToCloudFormation(properties.cpu),
            DisableNetworking: cdk.booleanToCloudFormation(properties.disableNetworking),
            DnsSearchDomains: cdk.listMapper(cdk.stringToCloudFormation)(properties.dnsSearchDomains),
            DnsServers: cdk.listMapper(cdk.stringToCloudFormation)(properties.dnsServers),
            DockerLabels: cdk.hashMapper(cdk.stringToCloudFormation)(properties.dockerLabels),
            DockerSecurityOptions: cdk.listMapper(cdk.stringToCloudFormation)(properties.dockerSecurityOptions),
            EntryPoint: cdk.listMapper(cdk.stringToCloudFormation)(properties.entryPoint),
            Environment: cdk.listMapper(taskDefinitionResourceKeyValuePairPropertyToCloudFormation)(properties.environment),
            Essential: cdk.booleanToCloudFormation(properties.essential),
            ExtraHosts: cdk.listMapper(taskDefinitionResourceHostEntryPropertyToCloudFormation)(properties.extraHosts),
            HealthCheck: taskDefinitionResourceHealthCheckPropertyToCloudFormation(properties.healthCheck),
            Hostname: cdk.stringToCloudFormation(properties.hostname),
            Image: cdk.stringToCloudFormation(properties.image),
            Links: cdk.listMapper(cdk.stringToCloudFormation)(properties.links),
            LinuxParameters: taskDefinitionResourceLinuxParametersPropertyToCloudFormation(properties.linuxParameters),
            LogConfiguration: taskDefinitionResourceLogConfigurationPropertyToCloudFormation(properties.logConfiguration),
            Memory: cdk.numberToCloudFormation(properties.memory),
            MemoryReservation: cdk.numberToCloudFormation(properties.memoryReservation),
            MountPoints: cdk.listMapper(taskDefinitionResourceMountPointPropertyToCloudFormation)(properties.mountPoints),
            Name: cdk.stringToCloudFormation(properties.name),
            PortMappings: cdk.listMapper(taskDefinitionResourcePortMappingPropertyToCloudFormation)(properties.portMappings),
            Privileged: cdk.booleanToCloudFormation(properties.privileged),
            ReadonlyRootFilesystem: cdk.booleanToCloudFormation(properties.readonlyRootFilesystem),
            RepositoryCredentials: taskDefinitionResourceRepositoryCredentialsPropertyToCloudFormation(properties.repositoryCredentials),
            Ulimits: cdk.listMapper(taskDefinitionResourceUlimitPropertyToCloudFormation)(properties.ulimits),
            User: cdk.stringToCloudFormation(properties.user),
            VolumesFrom: cdk.listMapper(taskDefinitionResourceVolumeFromPropertyToCloudFormation)(properties.volumesFrom),
            WorkingDirectory: cdk.stringToCloudFormation(properties.workingDirectory),
        };
    }
    /**
     * Determine whether the given properties match those of a ``DeviceProperty``
     *
     * @param properties the TypeScript properties of a ``DeviceProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_DevicePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerPath', cdk.validateString)(properties.containerPath));
        errors.collect(cdk.propertyValidator('hostPath', cdk.requiredValidator)(properties.hostPath));
        errors.collect(cdk.propertyValidator('hostPath', cdk.validateString)(properties.hostPath));
        errors.collect(cdk.propertyValidator('permissions', cdk.listValidator(cdk.validateString))(properties.permissions));
        return errors.wrap('supplied properties not correct for "DeviceProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Device`` resource
     *
     * @param properties the TypeScript properties of a ``DeviceProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Device`` resource.
     */
    function taskDefinitionResourceDevicePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_DevicePropertyValidator(properties).assertSuccess();
        return {
            ContainerPath: cdk.stringToCloudFormation(properties.containerPath),
            HostPath: cdk.stringToCloudFormation(properties.hostPath),
            Permissions: cdk.listMapper(cdk.stringToCloudFormation)(properties.permissions),
        };
    }
    /**
     * Determine whether the given properties match those of a ``DockerVolumeConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``DockerVolumeConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_DockerVolumeConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('autoprovision', cdk.validateBoolean)(properties.autoprovision));
        errors.collect(cdk.propertyValidator('driver', cdk.validateString)(properties.driver));
        errors.collect(cdk.propertyValidator('driverOpts', cdk.hashValidator(cdk.validateString))(properties.driverOpts));
        errors.collect(cdk.propertyValidator('labels', cdk.hashValidator(cdk.validateString))(properties.labels));
        errors.collect(cdk.propertyValidator('scope', cdk.validateString)(properties.scope));
        return errors.wrap('supplied properties not correct for "DockerVolumeConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.DockerVolumeConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``DockerVolumeConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.DockerVolumeConfiguration`` resource.
     */
    function taskDefinitionResourceDockerVolumeConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_DockerVolumeConfigurationPropertyValidator(properties).assertSuccess();
        return {
            Autoprovision: cdk.booleanToCloudFormation(properties.autoprovision),
            Driver: cdk.stringToCloudFormation(properties.driver),
            DriverOpts: cdk.hashMapper(cdk.stringToCloudFormation)(properties.driverOpts),
            Labels: cdk.hashMapper(cdk.stringToCloudFormation)(properties.labels),
            Scope: cdk.stringToCloudFormation(properties.scope),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HealthCheckProperty``
     *
     * @param properties the TypeScript properties of a ``HealthCheckProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_HealthCheckPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('command', cdk.requiredValidator)(properties.command));
        errors.collect(cdk.propertyValidator('command', cdk.listValidator(cdk.validateString))(properties.command));
        errors.collect(cdk.propertyValidator('interval', cdk.validateNumber)(properties.interval));
        errors.collect(cdk.propertyValidator('retries', cdk.validateNumber)(properties.retries));
        errors.collect(cdk.propertyValidator('startPeriod', cdk.validateNumber)(properties.startPeriod));
        errors.collect(cdk.propertyValidator('timeout', cdk.validateNumber)(properties.timeout));
        return errors.wrap('supplied properties not correct for "HealthCheckProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HealthCheck`` resource
     *
     * @param properties the TypeScript properties of a ``HealthCheckProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HealthCheck`` resource.
     */
    function taskDefinitionResourceHealthCheckPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_HealthCheckPropertyValidator(properties).assertSuccess();
        return {
            Command: cdk.listMapper(cdk.stringToCloudFormation)(properties.command),
            Interval: cdk.numberToCloudFormation(properties.interval),
            Retries: cdk.numberToCloudFormation(properties.retries),
            StartPeriod: cdk.numberToCloudFormation(properties.startPeriod),
            Timeout: cdk.numberToCloudFormation(properties.timeout),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HostEntryProperty``
     *
     * @param properties the TypeScript properties of a ``HostEntryProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_HostEntryPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('hostname', cdk.requiredValidator)(properties.hostname));
        errors.collect(cdk.propertyValidator('hostname', cdk.validateString)(properties.hostname));
        errors.collect(cdk.propertyValidator('ipAddress', cdk.requiredValidator)(properties.ipAddress));
        errors.collect(cdk.propertyValidator('ipAddress', cdk.validateString)(properties.ipAddress));
        return errors.wrap('supplied properties not correct for "HostEntryProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HostEntry`` resource
     *
     * @param properties the TypeScript properties of a ``HostEntryProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HostEntry`` resource.
     */
    function taskDefinitionResourceHostEntryPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_HostEntryPropertyValidator(properties).assertSuccess();
        return {
            Hostname: cdk.stringToCloudFormation(properties.hostname),
            IpAddress: cdk.stringToCloudFormation(properties.ipAddress),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HostVolumePropertiesProperty``
     *
     * @param properties the TypeScript properties of a ``HostVolumePropertiesProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_HostVolumePropertiesPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('sourcePath', cdk.validateString)(properties.sourcePath));
        return errors.wrap('supplied properties not correct for "HostVolumePropertiesProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HostVolumeProperties`` resource
     *
     * @param properties the TypeScript properties of a ``HostVolumePropertiesProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.HostVolumeProperties`` resource.
     */
    function taskDefinitionResourceHostVolumePropertiesPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_HostVolumePropertiesPropertyValidator(properties).assertSuccess();
        return {
            SourcePath: cdk.stringToCloudFormation(properties.sourcePath),
        };
    }
    /**
     * Determine whether the given properties match those of a ``KernelCapabilitiesProperty``
     *
     * @param properties the TypeScript properties of a ``KernelCapabilitiesProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_KernelCapabilitiesPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('add', cdk.listValidator(cdk.validateString))(properties.add));
        errors.collect(cdk.propertyValidator('drop', cdk.listValidator(cdk.validateString))(properties.drop));
        return errors.wrap('supplied properties not correct for "KernelCapabilitiesProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.KernelCapabilities`` resource
     *
     * @param properties the TypeScript properties of a ``KernelCapabilitiesProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.KernelCapabilities`` resource.
     */
    function taskDefinitionResourceKernelCapabilitiesPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_KernelCapabilitiesPropertyValidator(properties).assertSuccess();
        return {
            Add: cdk.listMapper(cdk.stringToCloudFormation)(properties.add),
            Drop: cdk.listMapper(cdk.stringToCloudFormation)(properties.drop),
        };
    }
    /**
     * Determine whether the given properties match those of a ``KeyValuePairProperty``
     *
     * @param properties the TypeScript properties of a ``KeyValuePairProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_KeyValuePairPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "KeyValuePairProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.KeyValuePair`` resource
     *
     * @param properties the TypeScript properties of a ``KeyValuePairProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.KeyValuePair`` resource.
     */
    function taskDefinitionResourceKeyValuePairPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_KeyValuePairPropertyValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LinuxParametersProperty``
     *
     * @param properties the TypeScript properties of a ``LinuxParametersProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_LinuxParametersPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('capabilities', TaskDefinitionResource_KernelCapabilitiesPropertyValidator)(properties.capabilities));
        errors.collect(cdk.propertyValidator('devices', cdk.listValidator(TaskDefinitionResource_DevicePropertyValidator))(properties.devices));
        errors.collect(cdk.propertyValidator('initProcessEnabled', cdk.validateBoolean)(properties.initProcessEnabled));
        errors.collect(cdk.propertyValidator('sharedMemorySize', cdk.validateNumber)(properties.sharedMemorySize));
        errors.collect(cdk.propertyValidator('tmpfs', cdk.listValidator(TaskDefinitionResource_TmpfsPropertyValidator))(properties.tmpfs));
        return errors.wrap('supplied properties not correct for "LinuxParametersProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.LinuxParameters`` resource
     *
     * @param properties the TypeScript properties of a ``LinuxParametersProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.LinuxParameters`` resource.
     */
    function taskDefinitionResourceLinuxParametersPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_LinuxParametersPropertyValidator(properties).assertSuccess();
        return {
            Capabilities: taskDefinitionResourceKernelCapabilitiesPropertyToCloudFormation(properties.capabilities),
            Devices: cdk.listMapper(taskDefinitionResourceDevicePropertyToCloudFormation)(properties.devices),
            InitProcessEnabled: cdk.booleanToCloudFormation(properties.initProcessEnabled),
            SharedMemorySize: cdk.numberToCloudFormation(properties.sharedMemorySize),
            Tmpfs: cdk.listMapper(taskDefinitionResourceTmpfsPropertyToCloudFormation)(properties.tmpfs),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LogConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``LogConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_LogConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('logDriver', cdk.requiredValidator)(properties.logDriver));
        errors.collect(cdk.propertyValidator('logDriver', cdk.validateString)(properties.logDriver));
        errors.collect(cdk.propertyValidator('options', cdk.hashValidator(cdk.validateString))(properties.options));
        return errors.wrap('supplied properties not correct for "LogConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.LogConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``LogConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.LogConfiguration`` resource.
     */
    function taskDefinitionResourceLogConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_LogConfigurationPropertyValidator(properties).assertSuccess();
        return {
            LogDriver: cdk.stringToCloudFormation(properties.logDriver),
            Options: cdk.hashMapper(cdk.stringToCloudFormation)(properties.options),
        };
    }
    /**
     * Determine whether the given properties match those of a ``MountPointProperty``
     *
     * @param properties the TypeScript properties of a ``MountPointProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_MountPointPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerPath', cdk.validateString)(properties.containerPath));
        errors.collect(cdk.propertyValidator('readOnly', cdk.validateBoolean)(properties.readOnly));
        errors.collect(cdk.propertyValidator('sourceVolume', cdk.validateString)(properties.sourceVolume));
        return errors.wrap('supplied properties not correct for "MountPointProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.MountPoint`` resource
     *
     * @param properties the TypeScript properties of a ``MountPointProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.MountPoint`` resource.
     */
    function taskDefinitionResourceMountPointPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_MountPointPropertyValidator(properties).assertSuccess();
        return {
            ContainerPath: cdk.stringToCloudFormation(properties.containerPath),
            ReadOnly: cdk.booleanToCloudFormation(properties.readOnly),
            SourceVolume: cdk.stringToCloudFormation(properties.sourceVolume),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PortMappingProperty``
     *
     * @param properties the TypeScript properties of a ``PortMappingProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_PortMappingPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerPort', cdk.validateNumber)(properties.containerPort));
        errors.collect(cdk.propertyValidator('hostPort', cdk.validateNumber)(properties.hostPort));
        errors.collect(cdk.propertyValidator('protocol', cdk.validateString)(properties.protocol));
        return errors.wrap('supplied properties not correct for "PortMappingProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.PortMapping`` resource
     *
     * @param properties the TypeScript properties of a ``PortMappingProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.PortMapping`` resource.
     */
    function taskDefinitionResourcePortMappingPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_PortMappingPropertyValidator(properties).assertSuccess();
        return {
            ContainerPort: cdk.numberToCloudFormation(properties.containerPort),
            HostPort: cdk.numberToCloudFormation(properties.hostPort),
            Protocol: cdk.stringToCloudFormation(properties.protocol),
        };
    }
    /**
     * Determine whether the given properties match those of a ``RepositoryCredentialsProperty``
     *
     * @param properties the TypeScript properties of a ``RepositoryCredentialsProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_RepositoryCredentialsPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('credentialsParameter', cdk.validateString)(properties.credentialsParameter));
        return errors.wrap('supplied properties not correct for "RepositoryCredentialsProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.RepositoryCredentials`` resource
     *
     * @param properties the TypeScript properties of a ``RepositoryCredentialsProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.RepositoryCredentials`` resource.
     */
    function taskDefinitionResourceRepositoryCredentialsPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_RepositoryCredentialsPropertyValidator(properties).assertSuccess();
        return {
            CredentialsParameter: cdk.stringToCloudFormation(properties.credentialsParameter),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TaskDefinitionPlacementConstraintProperty``
     *
     * @param properties the TypeScript properties of a ``TaskDefinitionPlacementConstraintProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_TaskDefinitionPlacementConstraintPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('expression', cdk.validateString)(properties.expression));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "TaskDefinitionPlacementConstraintProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.TaskDefinitionPlacementConstraint`` resource
     *
     * @param properties the TypeScript properties of a ``TaskDefinitionPlacementConstraintProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.TaskDefinitionPlacementConstraint`` resource.
     */
    function taskDefinitionResourceTaskDefinitionPlacementConstraintPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_TaskDefinitionPlacementConstraintPropertyValidator(properties).assertSuccess();
        return {
            Expression: cdk.stringToCloudFormation(properties.expression),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TmpfsProperty``
     *
     * @param properties the TypeScript properties of a ``TmpfsProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_TmpfsPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('containerPath', cdk.validateString)(properties.containerPath));
        errors.collect(cdk.propertyValidator('mountOptions', cdk.listValidator(cdk.validateString))(properties.mountOptions));
        errors.collect(cdk.propertyValidator('size', cdk.validateNumber)(properties.size));
        return errors.wrap('supplied properties not correct for "TmpfsProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Tmpfs`` resource
     *
     * @param properties the TypeScript properties of a ``TmpfsProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Tmpfs`` resource.
     */
    function taskDefinitionResourceTmpfsPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_TmpfsPropertyValidator(properties).assertSuccess();
        return {
            ContainerPath: cdk.stringToCloudFormation(properties.containerPath),
            MountOptions: cdk.listMapper(cdk.stringToCloudFormation)(properties.mountOptions),
            Size: cdk.numberToCloudFormation(properties.size),
        };
    }
    /**
     * Determine whether the given properties match those of a ``UlimitProperty``
     *
     * @param properties the TypeScript properties of a ``UlimitProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_UlimitPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('hardLimit', cdk.requiredValidator)(properties.hardLimit));
        errors.collect(cdk.propertyValidator('hardLimit', cdk.validateNumber)(properties.hardLimit));
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('softLimit', cdk.requiredValidator)(properties.softLimit));
        errors.collect(cdk.propertyValidator('softLimit', cdk.validateNumber)(properties.softLimit));
        return errors.wrap('supplied properties not correct for "UlimitProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Ulimit`` resource
     *
     * @param properties the TypeScript properties of a ``UlimitProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Ulimit`` resource.
     */
    function taskDefinitionResourceUlimitPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_UlimitPropertyValidator(properties).assertSuccess();
        return {
            HardLimit: cdk.numberToCloudFormation(properties.hardLimit),
            Name: cdk.stringToCloudFormation(properties.name),
            SoftLimit: cdk.numberToCloudFormation(properties.softLimit),
        };
    }
    /**
     * Determine whether the given properties match those of a ``VolumeProperty``
     *
     * @param properties the TypeScript properties of a ``VolumeProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_VolumePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('dockerVolumeConfiguration', TaskDefinitionResource_DockerVolumeConfigurationPropertyValidator)(properties.dockerVolumeConfiguration));
        errors.collect(cdk.propertyValidator('host', TaskDefinitionResource_HostVolumePropertiesPropertyValidator)(properties.host));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        return errors.wrap('supplied properties not correct for "VolumeProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Volume`` resource
     *
     * @param properties the TypeScript properties of a ``VolumeProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.Volume`` resource.
     */
    function taskDefinitionResourceVolumePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_VolumePropertyValidator(properties).assertSuccess();
        return {
            DockerVolumeConfiguration: taskDefinitionResourceDockerVolumeConfigurationPropertyToCloudFormation(properties.dockerVolumeConfiguration),
            Host: taskDefinitionResourceHostVolumePropertiesPropertyToCloudFormation(properties.host),
            Name: cdk.stringToCloudFormation(properties.name),
        };
    }
    /**
     * Determine whether the given properties match those of a ``VolumeFromProperty``
     *
     * @param properties the TypeScript properties of a ``VolumeFromProperty``
     *
     * @returns the result of the validation.
     */
    function TaskDefinitionResource_VolumeFromPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('readOnly', cdk.validateBoolean)(properties.readOnly));
        errors.collect(cdk.propertyValidator('sourceContainer', cdk.validateString)(properties.sourceContainer));
        return errors.wrap('supplied properties not correct for "VolumeFromProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.VolumeFrom`` resource
     *
     * @param properties the TypeScript properties of a ``VolumeFromProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECS::TaskDefinition.VolumeFrom`` resource.
     */
    function taskDefinitionResourceVolumeFromPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        TaskDefinitionResource_VolumeFromPropertyValidator(properties).assertSuccess();
        return {
            ReadOnly: cdk.booleanToCloudFormation(properties.readOnly),
            SourceContainer: cdk.stringToCloudFormation(properties.sourceContainer),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
