// @aws-cdk/aws-ecs/lib/load-balanced-ecs-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const elbv2 = require("@aws-cdk/aws-elasticloadbalancingv2");
const cdk = require("@aws-cdk/cdk");
const ec2_service_1 = require("./ec2/ec2-service");
const ec2_task_definition_1 = require("./ec2/ec2-task-definition");
/**
 * A single task running on an ECS cluster fronted by a load balancer
 */
class LoadBalancedEc2Service extends cdk.Construct {
    constructor(parent, id, props) {
        super(parent, id);
        const taskDefinition = new ec2_task_definition_1.Ec2TaskDefinition(this, 'TaskDef', {});
        const container = taskDefinition.addContainer('web', {
            image: props.image,
            memoryLimitMiB: props.memoryLimitMiB,
        });
        container.addPortMappings({
            containerPort: props.containerPort || 80,
        });
        const service = new ec2_service_1.Ec2Service(this, "Service", {
            cluster: props.cluster,
            desiredCount: props.desiredCount || 1,
            taskDefinition,
        });
        const internetFacing = props.publicLoadBalancer !== undefined ? props.publicLoadBalancer : true;
        const lb = new elbv2.ApplicationLoadBalancer(this, 'LB', {
            vpc: props.cluster.vpc,
            internetFacing
        });
        this.loadBalancer = lb;
        const listener = lb.addListener('PublicListener', { port: 80, open: true });
        listener.addTargets('ECS', {
            port: 80,
            targets: [service]
        });
        new cdk.Output(this, 'LoadBalancerDNS', { value: lb.dnsName });
    }
}
exports.LoadBalancedEc2Service = LoadBalancedEc2Service;
