// @aws-cdk/aws-applicationautoscaling/test/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const scalingcommon = require("@aws-cdk/aws-autoscaling-common");
const fc = require("fast-check");
const appscaling = require("../lib");
const lib_1 = require("../lib");
function createScalableTarget(parent) {
    return new appscaling.ScalableTarget(parent, 'Target', {
        serviceNamespace: lib_1.ServiceNamespace.DynamoDb,
        scalableDimension: 'test:TestCount',
        resourceId: 'test:this/test',
        minCapacity: 1,
        maxCapacity: 20,
    });
}
exports.createScalableTarget = createScalableTarget;
class ArbitraryInputIntervals extends fc.Arbitrary {
    generate(mrng) {
        const ret = scalingcommon.generateArbitraryIntervals(mrng);
        return new fc.Shrinkable(ret.intervals);
    }
}
exports.ArbitraryInputIntervals = ArbitraryInputIntervals;
function arbitrary_input_intervals() {
    return new ArbitraryInputIntervals();
}
exports.arbitrary_input_intervals = arbitrary_input_intervals;
