// @aws-cdk/aws-ecs/lib/log-drivers/aws-log-driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logs = require("@aws-cdk/aws-logs");
const cdk = require("@aws-cdk/cdk");
const log_driver_1 = require("./log-driver");
/**
 * A log driver that will log to an AWS Log Group
 */
class AwsLogDriver extends log_driver_1.LogDriver {
    constructor(parent, id, props) {
        super(parent, id);
        this.props = props;
        this.logGroup = props.logGroup || new logs.LogGroup(this, 'LogGroup', {
            retentionDays: 365,
        });
    }
    /**
     * Called when the log driver is configured on a container
     */
    bind(containerDefinition) {
        this.logGroup.grantWrite(containerDefinition.taskDefinition.obtainExecutionRole());
    }
    /**
     * Return the log driver CloudFormation JSON
     */
    renderLogDriver() {
        return {
            logDriver: 'awslogs',
            options: removeEmpty({
                'awslogs-group': this.logGroup.logGroupName,
                'awslogs-stream-prefix': this.props.streamPrefix,
                'awslogs-region': `${new cdk.AwsRegion()}`,
                'awslogs-datetime-format': this.props.datetimeFormat,
                'awslogs-multiline-pattern': this.props.multilinePattern,
            }),
        };
    }
}
exports.AwsLogDriver = AwsLogDriver;
/**
 * Remove undefined values from a dictionary
 */
function removeEmpty(x) {
    for (const key of Object.keys(x)) {
        if (!x[key]) {
            delete x[key];
        }
    }
    return x;
}
