// @aws-cdk/aws-autoscaling/lib/autoscaling.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:06:23.757Z","fingerprint":"tUr1ohcEA7NsuDv9E9cfxP1Q2hA725eTlL/l+095ICc="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``AutoScalingGroupResourceProps``
     *
     * @param properties the TypeScript properties of a ``AutoScalingGroupResourceProps``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.validateString)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('availabilityZones', cdk.listValidator(cdk.validateString))(properties.availabilityZones));
        errors.collect(cdk.propertyValidator('cooldown', cdk.validateString)(properties.cooldown));
        errors.collect(cdk.propertyValidator('desiredCapacity', cdk.validateString)(properties.desiredCapacity));
        errors.collect(cdk.propertyValidator('healthCheckGracePeriod', cdk.validateNumber)(properties.healthCheckGracePeriod));
        errors.collect(cdk.propertyValidator('healthCheckType', cdk.validateString)(properties.healthCheckType));
        errors.collect(cdk.propertyValidator('instanceId', cdk.validateString)(properties.instanceId));
        errors.collect(cdk.propertyValidator('launchConfigurationName', cdk.validateString)(properties.launchConfigurationName));
        errors.collect(cdk.propertyValidator('launchTemplate', AutoScalingGroupResource_LaunchTemplateSpecificationPropertyValidator)(properties.launchTemplate));
        errors.collect(cdk.propertyValidator('lifecycleHookSpecificationList', cdk.listValidator(AutoScalingGroupResource_LifecycleHookSpecificationPropertyValidator))(properties.lifecycleHookSpecificationList));
        errors.collect(cdk.propertyValidator('loadBalancerNames', cdk.listValidator(cdk.validateString))(properties.loadBalancerNames));
        errors.collect(cdk.propertyValidator('maxSize', cdk.requiredValidator)(properties.maxSize));
        errors.collect(cdk.propertyValidator('maxSize', cdk.validateString)(properties.maxSize));
        errors.collect(cdk.propertyValidator('metricsCollection', cdk.listValidator(AutoScalingGroupResource_MetricsCollectionPropertyValidator))(properties.metricsCollection));
        errors.collect(cdk.propertyValidator('minSize', cdk.requiredValidator)(properties.minSize));
        errors.collect(cdk.propertyValidator('minSize', cdk.validateString)(properties.minSize));
        errors.collect(cdk.propertyValidator('mixedInstancesPolicy', AutoScalingGroupResource_MixedInstancesPolicyPropertyValidator)(properties.mixedInstancesPolicy));
        errors.collect(cdk.propertyValidator('notificationConfigurations', cdk.listValidator(AutoScalingGroupResource_NotificationConfigurationPropertyValidator))(properties.notificationConfigurations));
        errors.collect(cdk.propertyValidator('placementGroup', cdk.validateString)(properties.placementGroup));
        errors.collect(cdk.propertyValidator('serviceLinkedRoleArn', cdk.validateString)(properties.serviceLinkedRoleArn));
        errors.collect(cdk.propertyValidator('tags', cdk.listValidator(AutoScalingGroupResource_TagPropertyPropertyValidator))(properties.tags));
        errors.collect(cdk.propertyValidator('targetGroupArns', cdk.listValidator(cdk.validateString))(properties.targetGroupArns));
        errors.collect(cdk.propertyValidator('terminationPolicies', cdk.listValidator(cdk.validateString))(properties.terminationPolicies));
        errors.collect(cdk.propertyValidator('vpcZoneIdentifier', cdk.listValidator(cdk.validateString))(properties.vpcZoneIdentifier));
        return errors.wrap('supplied properties not correct for "AutoScalingGroupResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup`` resource
     *
     * @param properties the TypeScript properties of a ``AutoScalingGroupResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup`` resource.
     */
    function autoScalingGroupResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResourcePropsValidator(properties).assertSuccess();
        return {
            MaxSize: cdk.stringToCloudFormation(properties.maxSize),
            MinSize: cdk.stringToCloudFormation(properties.minSize),
            AutoScalingGroupName: cdk.stringToCloudFormation(properties.autoScalingGroupName),
            AvailabilityZones: cdk.listMapper(cdk.stringToCloudFormation)(properties.availabilityZones),
            Cooldown: cdk.stringToCloudFormation(properties.cooldown),
            DesiredCapacity: cdk.stringToCloudFormation(properties.desiredCapacity),
            HealthCheckGracePeriod: cdk.numberToCloudFormation(properties.healthCheckGracePeriod),
            HealthCheckType: cdk.stringToCloudFormation(properties.healthCheckType),
            InstanceId: cdk.stringToCloudFormation(properties.instanceId),
            LaunchConfigurationName: cdk.stringToCloudFormation(properties.launchConfigurationName),
            LaunchTemplate: autoScalingGroupResourceLaunchTemplateSpecificationPropertyToCloudFormation(properties.launchTemplate),
            LifecycleHookSpecificationList: cdk.listMapper(autoScalingGroupResourceLifecycleHookSpecificationPropertyToCloudFormation)(properties.lifecycleHookSpecificationList),
            LoadBalancerNames: cdk.listMapper(cdk.stringToCloudFormation)(properties.loadBalancerNames),
            MetricsCollection: cdk.listMapper(autoScalingGroupResourceMetricsCollectionPropertyToCloudFormation)(properties.metricsCollection),
            MixedInstancesPolicy: autoScalingGroupResourceMixedInstancesPolicyPropertyToCloudFormation(properties.mixedInstancesPolicy),
            NotificationConfigurations: cdk.listMapper(autoScalingGroupResourceNotificationConfigurationPropertyToCloudFormation)(properties.notificationConfigurations),
            PlacementGroup: cdk.stringToCloudFormation(properties.placementGroup),
            ServiceLinkedRoleARN: cdk.stringToCloudFormation(properties.serviceLinkedRoleArn),
            Tags: cdk.listMapper(autoScalingGroupResourceTagPropertyPropertyToCloudFormation)(properties.tags),
            TargetGroupARNs: cdk.listMapper(cdk.stringToCloudFormation)(properties.targetGroupArns),
            TerminationPolicies: cdk.listMapper(cdk.stringToCloudFormation)(properties.terminationPolicies),
            VPCZoneIdentifier: cdk.listMapper(cdk.stringToCloudFormation)(properties.vpcZoneIdentifier),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html
     */
    class AutoScalingGroupResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::AutoScaling::AutoScalingGroup``.
         *
         * @param parent   the ``cdk.Construct`` this ``AutoScalingGroupResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``AutoScalingGroupResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: AutoScalingGroupResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'maxSize', this);
            cdk.requireProperty(properties, 'minSize', this);
            this.autoScalingGroupName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return autoScalingGroupResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    AutoScalingGroupResource.resourceTypeName = "AWS::AutoScaling::AutoScalingGroup";
    cloudformation.AutoScalingGroupResource = AutoScalingGroupResource;
    /**
     * Determine whether the given properties match those of a ``InstancesDistributionProperty``
     *
     * @param properties the TypeScript properties of a ``InstancesDistributionProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_InstancesDistributionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('onDemandAllocationStrategy', cdk.validateString)(properties.onDemandAllocationStrategy));
        errors.collect(cdk.propertyValidator('onDemandBaseCapacity', cdk.validateNumber)(properties.onDemandBaseCapacity));
        errors.collect(cdk.propertyValidator('onDemandPercentageAboveBaseCapacity', cdk.validateNumber)(properties.onDemandPercentageAboveBaseCapacity));
        errors.collect(cdk.propertyValidator('spotAllocationStrategy', cdk.validateString)(properties.spotAllocationStrategy));
        errors.collect(cdk.propertyValidator('spotInstancePools', cdk.validateNumber)(properties.spotInstancePools));
        errors.collect(cdk.propertyValidator('spotMaxPrice', cdk.validateString)(properties.spotMaxPrice));
        return errors.wrap('supplied properties not correct for "InstancesDistributionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.InstancesDistribution`` resource
     *
     * @param properties the TypeScript properties of a ``InstancesDistributionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.InstancesDistribution`` resource.
     */
    function autoScalingGroupResourceInstancesDistributionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_InstancesDistributionPropertyValidator(properties).assertSuccess();
        return {
            OnDemandAllocationStrategy: cdk.stringToCloudFormation(properties.onDemandAllocationStrategy),
            OnDemandBaseCapacity: cdk.numberToCloudFormation(properties.onDemandBaseCapacity),
            OnDemandPercentageAboveBaseCapacity: cdk.numberToCloudFormation(properties.onDemandPercentageAboveBaseCapacity),
            SpotAllocationStrategy: cdk.stringToCloudFormation(properties.spotAllocationStrategy),
            SpotInstancePools: cdk.numberToCloudFormation(properties.spotInstancePools),
            SpotMaxPrice: cdk.stringToCloudFormation(properties.spotMaxPrice),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LaunchTemplateProperty``
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_LaunchTemplatePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('launchTemplateSpecification', cdk.requiredValidator)(properties.launchTemplateSpecification));
        errors.collect(cdk.propertyValidator('launchTemplateSpecification', AutoScalingGroupResource_LaunchTemplateSpecificationPropertyValidator)(properties.launchTemplateSpecification));
        errors.collect(cdk.propertyValidator('overrides', cdk.listValidator(AutoScalingGroupResource_LaunchTemplateOverridesPropertyValidator))(properties.overrides));
        return errors.wrap('supplied properties not correct for "LaunchTemplateProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplate`` resource
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplate`` resource.
     */
    function autoScalingGroupResourceLaunchTemplatePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_LaunchTemplatePropertyValidator(properties).assertSuccess();
        return {
            LaunchTemplateSpecification: autoScalingGroupResourceLaunchTemplateSpecificationPropertyToCloudFormation(properties.launchTemplateSpecification),
            Overrides: cdk.listMapper(autoScalingGroupResourceLaunchTemplateOverridesPropertyToCloudFormation)(properties.overrides),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LaunchTemplateOverridesProperty``
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateOverridesProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_LaunchTemplateOverridesPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('instanceType', cdk.validateString)(properties.instanceType));
        return errors.wrap('supplied properties not correct for "LaunchTemplateOverridesProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplateOverrides`` resource
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateOverridesProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplateOverrides`` resource.
     */
    function autoScalingGroupResourceLaunchTemplateOverridesPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_LaunchTemplateOverridesPropertyValidator(properties).assertSuccess();
        return {
            InstanceType: cdk.stringToCloudFormation(properties.instanceType),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LaunchTemplateSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_LaunchTemplateSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('launchTemplateId', cdk.validateString)(properties.launchTemplateId));
        errors.collect(cdk.propertyValidator('launchTemplateName', cdk.validateString)(properties.launchTemplateName));
        errors.collect(cdk.propertyValidator('version', cdk.requiredValidator)(properties.version));
        errors.collect(cdk.propertyValidator('version', cdk.validateString)(properties.version));
        return errors.wrap('supplied properties not correct for "LaunchTemplateSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplateSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``LaunchTemplateSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LaunchTemplateSpecification`` resource.
     */
    function autoScalingGroupResourceLaunchTemplateSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_LaunchTemplateSpecificationPropertyValidator(properties).assertSuccess();
        return {
            LaunchTemplateId: cdk.stringToCloudFormation(properties.launchTemplateId),
            LaunchTemplateName: cdk.stringToCloudFormation(properties.launchTemplateName),
            Version: cdk.stringToCloudFormation(properties.version),
        };
    }
    /**
     * Determine whether the given properties match those of a ``LifecycleHookSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``LifecycleHookSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_LifecycleHookSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('defaultResult', cdk.validateString)(properties.defaultResult));
        errors.collect(cdk.propertyValidator('heartbeatTimeout', cdk.validateNumber)(properties.heartbeatTimeout));
        errors.collect(cdk.propertyValidator('lifecycleHookName', cdk.requiredValidator)(properties.lifecycleHookName));
        errors.collect(cdk.propertyValidator('lifecycleHookName', cdk.validateString)(properties.lifecycleHookName));
        errors.collect(cdk.propertyValidator('lifecycleTransition', cdk.requiredValidator)(properties.lifecycleTransition));
        errors.collect(cdk.propertyValidator('lifecycleTransition', cdk.validateString)(properties.lifecycleTransition));
        errors.collect(cdk.propertyValidator('notificationMetadata', cdk.validateString)(properties.notificationMetadata));
        errors.collect(cdk.propertyValidator('notificationTargetArn', cdk.validateString)(properties.notificationTargetArn));
        errors.collect(cdk.propertyValidator('roleArn', cdk.validateString)(properties.roleArn));
        return errors.wrap('supplied properties not correct for "LifecycleHookSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LifecycleHookSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``LifecycleHookSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.LifecycleHookSpecification`` resource.
     */
    function autoScalingGroupResourceLifecycleHookSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_LifecycleHookSpecificationPropertyValidator(properties).assertSuccess();
        return {
            DefaultResult: cdk.stringToCloudFormation(properties.defaultResult),
            HeartbeatTimeout: cdk.numberToCloudFormation(properties.heartbeatTimeout),
            LifecycleHookName: cdk.stringToCloudFormation(properties.lifecycleHookName),
            LifecycleTransition: cdk.stringToCloudFormation(properties.lifecycleTransition),
            NotificationMetadata: cdk.stringToCloudFormation(properties.notificationMetadata),
            NotificationTargetARN: cdk.stringToCloudFormation(properties.notificationTargetArn),
            RoleARN: cdk.stringToCloudFormation(properties.roleArn),
        };
    }
    /**
     * Determine whether the given properties match those of a ``MetricsCollectionProperty``
     *
     * @param properties the TypeScript properties of a ``MetricsCollectionProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_MetricsCollectionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('granularity', cdk.requiredValidator)(properties.granularity));
        errors.collect(cdk.propertyValidator('granularity', cdk.validateString)(properties.granularity));
        errors.collect(cdk.propertyValidator('metrics', cdk.listValidator(cdk.validateString))(properties.metrics));
        return errors.wrap('supplied properties not correct for "MetricsCollectionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.MetricsCollection`` resource
     *
     * @param properties the TypeScript properties of a ``MetricsCollectionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.MetricsCollection`` resource.
     */
    function autoScalingGroupResourceMetricsCollectionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_MetricsCollectionPropertyValidator(properties).assertSuccess();
        return {
            Granularity: cdk.stringToCloudFormation(properties.granularity),
            Metrics: cdk.listMapper(cdk.stringToCloudFormation)(properties.metrics),
        };
    }
    /**
     * Determine whether the given properties match those of a ``MixedInstancesPolicyProperty``
     *
     * @param properties the TypeScript properties of a ``MixedInstancesPolicyProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_MixedInstancesPolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('instancesDistribution', AutoScalingGroupResource_InstancesDistributionPropertyValidator)(properties.instancesDistribution));
        errors.collect(cdk.propertyValidator('launchTemplate', cdk.requiredValidator)(properties.launchTemplate));
        errors.collect(cdk.propertyValidator('launchTemplate', AutoScalingGroupResource_LaunchTemplatePropertyValidator)(properties.launchTemplate));
        return errors.wrap('supplied properties not correct for "MixedInstancesPolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.MixedInstancesPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``MixedInstancesPolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.MixedInstancesPolicy`` resource.
     */
    function autoScalingGroupResourceMixedInstancesPolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_MixedInstancesPolicyPropertyValidator(properties).assertSuccess();
        return {
            InstancesDistribution: autoScalingGroupResourceInstancesDistributionPropertyToCloudFormation(properties.instancesDistribution),
            LaunchTemplate: autoScalingGroupResourceLaunchTemplatePropertyToCloudFormation(properties.launchTemplate),
        };
    }
    /**
     * Determine whether the given properties match those of a ``NotificationConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``NotificationConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_NotificationConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('notificationTypes', cdk.listValidator(cdk.validateString))(properties.notificationTypes));
        errors.collect(cdk.propertyValidator('topicArn', cdk.requiredValidator)(properties.topicArn));
        errors.collect(cdk.propertyValidator('topicArn', cdk.validateString)(properties.topicArn));
        return errors.wrap('supplied properties not correct for "NotificationConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.NotificationConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``NotificationConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.NotificationConfiguration`` resource.
     */
    function autoScalingGroupResourceNotificationConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_NotificationConfigurationPropertyValidator(properties).assertSuccess();
        return {
            NotificationTypes: cdk.listMapper(cdk.stringToCloudFormation)(properties.notificationTypes),
            TopicARN: cdk.stringToCloudFormation(properties.topicArn),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TagPropertyProperty``
     *
     * @param properties the TypeScript properties of a ``TagPropertyProperty``
     *
     * @returns the result of the validation.
     */
    function AutoScalingGroupResource_TagPropertyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('key', cdk.requiredValidator)(properties.key));
        errors.collect(cdk.propertyValidator('key', cdk.validateString)(properties.key));
        errors.collect(cdk.propertyValidator('propagateAtLaunch', cdk.requiredValidator)(properties.propagateAtLaunch));
        errors.collect(cdk.propertyValidator('propagateAtLaunch', cdk.validateBoolean)(properties.propagateAtLaunch));
        errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "TagPropertyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.TagProperty`` resource
     *
     * @param properties the TypeScript properties of a ``TagPropertyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::AutoScalingGroup.TagProperty`` resource.
     */
    function autoScalingGroupResourceTagPropertyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        AutoScalingGroupResource_TagPropertyPropertyValidator(properties).assertSuccess();
        return {
            Key: cdk.stringToCloudFormation(properties.key),
            PropagateAtLaunch: cdk.booleanToCloudFormation(properties.propagateAtLaunch),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``LaunchConfigurationResourceProps``
     *
     * @param properties the TypeScript properties of a ``LaunchConfigurationResourceProps``
     *
     * @returns the result of the validation.
     */
    function LaunchConfigurationResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('associatePublicIpAddress', cdk.validateBoolean)(properties.associatePublicIpAddress));
        errors.collect(cdk.propertyValidator('blockDeviceMappings', cdk.listValidator(LaunchConfigurationResource_BlockDeviceMappingPropertyValidator))(properties.blockDeviceMappings));
        errors.collect(cdk.propertyValidator('classicLinkVpcId', cdk.validateString)(properties.classicLinkVpcId));
        errors.collect(cdk.propertyValidator('classicLinkVpcSecurityGroups', cdk.listValidator(cdk.validateString))(properties.classicLinkVpcSecurityGroups));
        errors.collect(cdk.propertyValidator('ebsOptimized', cdk.validateBoolean)(properties.ebsOptimized));
        errors.collect(cdk.propertyValidator('iamInstanceProfile', cdk.validateString)(properties.iamInstanceProfile));
        errors.collect(cdk.propertyValidator('imageId', cdk.requiredValidator)(properties.imageId));
        errors.collect(cdk.propertyValidator('imageId', cdk.validateString)(properties.imageId));
        errors.collect(cdk.propertyValidator('instanceId', cdk.validateString)(properties.instanceId));
        errors.collect(cdk.propertyValidator('instanceMonitoring', cdk.validateBoolean)(properties.instanceMonitoring));
        errors.collect(cdk.propertyValidator('instanceType', cdk.requiredValidator)(properties.instanceType));
        errors.collect(cdk.propertyValidator('instanceType', cdk.validateString)(properties.instanceType));
        errors.collect(cdk.propertyValidator('kernelId', cdk.validateString)(properties.kernelId));
        errors.collect(cdk.propertyValidator('keyName', cdk.validateString)(properties.keyName));
        errors.collect(cdk.propertyValidator('launchConfigurationName', cdk.validateString)(properties.launchConfigurationName));
        errors.collect(cdk.propertyValidator('placementTenancy', cdk.validateString)(properties.placementTenancy));
        errors.collect(cdk.propertyValidator('ramDiskId', cdk.validateString)(properties.ramDiskId));
        errors.collect(cdk.propertyValidator('securityGroups', cdk.listValidator(cdk.validateString))(properties.securityGroups));
        errors.collect(cdk.propertyValidator('spotPrice', cdk.validateString)(properties.spotPrice));
        errors.collect(cdk.propertyValidator('userData', cdk.validateString)(properties.userData));
        return errors.wrap('supplied properties not correct for "LaunchConfigurationResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``LaunchConfigurationResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration`` resource.
     */
    function launchConfigurationResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LaunchConfigurationResourcePropsValidator(properties).assertSuccess();
        return {
            ImageId: cdk.stringToCloudFormation(properties.imageId),
            InstanceType: cdk.stringToCloudFormation(properties.instanceType),
            AssociatePublicIpAddress: cdk.booleanToCloudFormation(properties.associatePublicIpAddress),
            BlockDeviceMappings: cdk.listMapper(launchConfigurationResourceBlockDeviceMappingPropertyToCloudFormation)(properties.blockDeviceMappings),
            ClassicLinkVPCId: cdk.stringToCloudFormation(properties.classicLinkVpcId),
            ClassicLinkVPCSecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.classicLinkVpcSecurityGroups),
            EbsOptimized: cdk.booleanToCloudFormation(properties.ebsOptimized),
            IamInstanceProfile: cdk.stringToCloudFormation(properties.iamInstanceProfile),
            InstanceId: cdk.stringToCloudFormation(properties.instanceId),
            InstanceMonitoring: cdk.booleanToCloudFormation(properties.instanceMonitoring),
            KernelId: cdk.stringToCloudFormation(properties.kernelId),
            KeyName: cdk.stringToCloudFormation(properties.keyName),
            LaunchConfigurationName: cdk.stringToCloudFormation(properties.launchConfigurationName),
            PlacementTenancy: cdk.stringToCloudFormation(properties.placementTenancy),
            RamDiskId: cdk.stringToCloudFormation(properties.ramDiskId),
            SecurityGroups: cdk.listMapper(cdk.stringToCloudFormation)(properties.securityGroups),
            SpotPrice: cdk.stringToCloudFormation(properties.spotPrice),
            UserData: cdk.stringToCloudFormation(properties.userData),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-launchconfig.html
     */
    class LaunchConfigurationResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::AutoScaling::LaunchConfiguration``.
         *
         * @param parent   the ``cdk.Construct`` this ``LaunchConfigurationResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``LaunchConfigurationResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: LaunchConfigurationResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'imageId', this);
            cdk.requireProperty(properties, 'instanceType', this);
            this.launchConfigurationName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return launchConfigurationResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    LaunchConfigurationResource.resourceTypeName = "AWS::AutoScaling::LaunchConfiguration";
    cloudformation.LaunchConfigurationResource = LaunchConfigurationResource;
    /**
     * Determine whether the given properties match those of a ``BlockDeviceProperty``
     *
     * @param properties the TypeScript properties of a ``BlockDeviceProperty``
     *
     * @returns the result of the validation.
     */
    function LaunchConfigurationResource_BlockDevicePropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('deleteOnTermination', cdk.validateBoolean)(properties.deleteOnTermination));
        errors.collect(cdk.propertyValidator('encrypted', cdk.validateBoolean)(properties.encrypted));
        errors.collect(cdk.propertyValidator('iops', cdk.validateNumber)(properties.iops));
        errors.collect(cdk.propertyValidator('snapshotId', cdk.validateString)(properties.snapshotId));
        errors.collect(cdk.propertyValidator('volumeSize', cdk.validateNumber)(properties.volumeSize));
        errors.collect(cdk.propertyValidator('volumeType', cdk.validateString)(properties.volumeType));
        return errors.wrap('supplied properties not correct for "BlockDeviceProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration.BlockDevice`` resource
     *
     * @param properties the TypeScript properties of a ``BlockDeviceProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration.BlockDevice`` resource.
     */
    function launchConfigurationResourceBlockDevicePropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LaunchConfigurationResource_BlockDevicePropertyValidator(properties).assertSuccess();
        return {
            DeleteOnTermination: cdk.booleanToCloudFormation(properties.deleteOnTermination),
            Encrypted: cdk.booleanToCloudFormation(properties.encrypted),
            Iops: cdk.numberToCloudFormation(properties.iops),
            SnapshotId: cdk.stringToCloudFormation(properties.snapshotId),
            VolumeSize: cdk.numberToCloudFormation(properties.volumeSize),
            VolumeType: cdk.stringToCloudFormation(properties.volumeType),
        };
    }
    /**
     * Determine whether the given properties match those of a ``BlockDeviceMappingProperty``
     *
     * @param properties the TypeScript properties of a ``BlockDeviceMappingProperty``
     *
     * @returns the result of the validation.
     */
    function LaunchConfigurationResource_BlockDeviceMappingPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('deviceName', cdk.requiredValidator)(properties.deviceName));
        errors.collect(cdk.propertyValidator('deviceName', cdk.validateString)(properties.deviceName));
        errors.collect(cdk.propertyValidator('ebs', LaunchConfigurationResource_BlockDevicePropertyValidator)(properties.ebs));
        errors.collect(cdk.propertyValidator('noDevice', cdk.validateBoolean)(properties.noDevice));
        errors.collect(cdk.propertyValidator('virtualName', cdk.validateString)(properties.virtualName));
        return errors.wrap('supplied properties not correct for "BlockDeviceMappingProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration.BlockDeviceMapping`` resource
     *
     * @param properties the TypeScript properties of a ``BlockDeviceMappingProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::LaunchConfiguration.BlockDeviceMapping`` resource.
     */
    function launchConfigurationResourceBlockDeviceMappingPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LaunchConfigurationResource_BlockDeviceMappingPropertyValidator(properties).assertSuccess();
        return {
            DeviceName: cdk.stringToCloudFormation(properties.deviceName),
            Ebs: launchConfigurationResourceBlockDevicePropertyToCloudFormation(properties.ebs),
            NoDevice: cdk.booleanToCloudFormation(properties.noDevice),
            VirtualName: cdk.stringToCloudFormation(properties.virtualName),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``LifecycleHookResourceProps``
     *
     * @param properties the TypeScript properties of a ``LifecycleHookResourceProps``
     *
     * @returns the result of the validation.
     */
    function LifecycleHookResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.requiredValidator)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.validateString)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('defaultResult', cdk.validateString)(properties.defaultResult));
        errors.collect(cdk.propertyValidator('heartbeatTimeout', cdk.validateNumber)(properties.heartbeatTimeout));
        errors.collect(cdk.propertyValidator('lifecycleHookName', cdk.validateString)(properties.lifecycleHookName));
        errors.collect(cdk.propertyValidator('lifecycleTransition', cdk.requiredValidator)(properties.lifecycleTransition));
        errors.collect(cdk.propertyValidator('lifecycleTransition', cdk.validateString)(properties.lifecycleTransition));
        errors.collect(cdk.propertyValidator('notificationMetadata', cdk.validateString)(properties.notificationMetadata));
        errors.collect(cdk.propertyValidator('notificationTargetArn', cdk.validateString)(properties.notificationTargetArn));
        errors.collect(cdk.propertyValidator('roleArn', cdk.validateString)(properties.roleArn));
        return errors.wrap('supplied properties not correct for "LifecycleHookResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::LifecycleHook`` resource
     *
     * @param properties the TypeScript properties of a ``LifecycleHookResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::LifecycleHook`` resource.
     */
    function lifecycleHookResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        LifecycleHookResourcePropsValidator(properties).assertSuccess();
        return {
            AutoScalingGroupName: cdk.stringToCloudFormation(properties.autoScalingGroupName),
            LifecycleTransition: cdk.stringToCloudFormation(properties.lifecycleTransition),
            DefaultResult: cdk.stringToCloudFormation(properties.defaultResult),
            HeartbeatTimeout: cdk.numberToCloudFormation(properties.heartbeatTimeout),
            LifecycleHookName: cdk.stringToCloudFormation(properties.lifecycleHookName),
            NotificationMetadata: cdk.stringToCloudFormation(properties.notificationMetadata),
            NotificationTargetARN: cdk.stringToCloudFormation(properties.notificationTargetArn),
            RoleARN: cdk.stringToCloudFormation(properties.roleArn),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-as-lifecyclehook.html
     */
    class LifecycleHookResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::AutoScaling::LifecycleHook``.
         *
         * @param parent   the ``cdk.Construct`` this ``LifecycleHookResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``LifecycleHookResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: LifecycleHookResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'autoScalingGroupName', this);
            cdk.requireProperty(properties, 'lifecycleTransition', this);
            this.lifecycleHookName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return lifecycleHookResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    LifecycleHookResource.resourceTypeName = "AWS::AutoScaling::LifecycleHook";
    cloudformation.LifecycleHookResource = LifecycleHookResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ScalingPolicyResourceProps``
     *
     * @param properties the TypeScript properties of a ``ScalingPolicyResourceProps``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('adjustmentType', cdk.validateString)(properties.adjustmentType));
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.requiredValidator)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.validateString)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('cooldown', cdk.validateString)(properties.cooldown));
        errors.collect(cdk.propertyValidator('estimatedInstanceWarmup', cdk.validateNumber)(properties.estimatedInstanceWarmup));
        errors.collect(cdk.propertyValidator('metricAggregationType', cdk.validateString)(properties.metricAggregationType));
        errors.collect(cdk.propertyValidator('minAdjustmentMagnitude', cdk.validateNumber)(properties.minAdjustmentMagnitude));
        errors.collect(cdk.propertyValidator('policyType', cdk.validateString)(properties.policyType));
        errors.collect(cdk.propertyValidator('scalingAdjustment', cdk.validateNumber)(properties.scalingAdjustment));
        errors.collect(cdk.propertyValidator('stepAdjustments', cdk.listValidator(ScalingPolicyResource_StepAdjustmentPropertyValidator))(properties.stepAdjustments));
        errors.collect(cdk.propertyValidator('targetTrackingConfiguration', ScalingPolicyResource_TargetTrackingConfigurationPropertyValidator)(properties.targetTrackingConfiguration));
        return errors.wrap('supplied properties not correct for "ScalingPolicyResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy`` resource
     *
     * @param properties the TypeScript properties of a ``ScalingPolicyResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy`` resource.
     */
    function scalingPolicyResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResourcePropsValidator(properties).assertSuccess();
        return {
            AutoScalingGroupName: cdk.stringToCloudFormation(properties.autoScalingGroupName),
            AdjustmentType: cdk.stringToCloudFormation(properties.adjustmentType),
            Cooldown: cdk.stringToCloudFormation(properties.cooldown),
            EstimatedInstanceWarmup: cdk.numberToCloudFormation(properties.estimatedInstanceWarmup),
            MetricAggregationType: cdk.stringToCloudFormation(properties.metricAggregationType),
            MinAdjustmentMagnitude: cdk.numberToCloudFormation(properties.minAdjustmentMagnitude),
            PolicyType: cdk.stringToCloudFormation(properties.policyType),
            ScalingAdjustment: cdk.numberToCloudFormation(properties.scalingAdjustment),
            StepAdjustments: cdk.listMapper(scalingPolicyResourceStepAdjustmentPropertyToCloudFormation)(properties.stepAdjustments),
            TargetTrackingConfiguration: scalingPolicyResourceTargetTrackingConfigurationPropertyToCloudFormation(properties.targetTrackingConfiguration),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-policy.html
     */
    class ScalingPolicyResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::AutoScaling::ScalingPolicy``.
         *
         * @param parent   the ``cdk.Construct`` this ``ScalingPolicyResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ScalingPolicyResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ScalingPolicyResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'autoScalingGroupName', this);
            this.scalingPolicyArn = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return scalingPolicyResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ScalingPolicyResource.resourceTypeName = "AWS::AutoScaling::ScalingPolicy";
    cloudformation.ScalingPolicyResource = ScalingPolicyResource;
    /**
     * Determine whether the given properties match those of a ``CustomizedMetricSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``CustomizedMetricSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('dimensions', cdk.listValidator(ScalingPolicyResource_MetricDimensionPropertyValidator))(properties.dimensions));
        errors.collect(cdk.propertyValidator('metricName', cdk.requiredValidator)(properties.metricName));
        errors.collect(cdk.propertyValidator('metricName', cdk.validateString)(properties.metricName));
        errors.collect(cdk.propertyValidator('namespace', cdk.requiredValidator)(properties.namespace));
        errors.collect(cdk.propertyValidator('namespace', cdk.validateString)(properties.namespace));
        errors.collect(cdk.propertyValidator('statistic', cdk.requiredValidator)(properties.statistic));
        errors.collect(cdk.propertyValidator('statistic', cdk.validateString)(properties.statistic));
        errors.collect(cdk.propertyValidator('unit', cdk.validateString)(properties.unit));
        return errors.wrap('supplied properties not correct for "CustomizedMetricSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.CustomizedMetricSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``CustomizedMetricSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.CustomizedMetricSpecification`` resource.
     */
    function scalingPolicyResourceCustomizedMetricSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator(properties).assertSuccess();
        return {
            Dimensions: cdk.listMapper(scalingPolicyResourceMetricDimensionPropertyToCloudFormation)(properties.dimensions),
            MetricName: cdk.stringToCloudFormation(properties.metricName),
            Namespace: cdk.stringToCloudFormation(properties.namespace),
            Statistic: cdk.stringToCloudFormation(properties.statistic),
            Unit: cdk.stringToCloudFormation(properties.unit),
        };
    }
    /**
     * Determine whether the given properties match those of a ``MetricDimensionProperty``
     *
     * @param properties the TypeScript properties of a ``MetricDimensionProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_MetricDimensionPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "MetricDimensionProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.MetricDimension`` resource
     *
     * @param properties the TypeScript properties of a ``MetricDimensionProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.MetricDimension`` resource.
     */
    function scalingPolicyResourceMetricDimensionPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_MetricDimensionPropertyValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
    /**
     * Determine whether the given properties match those of a ``PredefinedMetricSpecificationProperty``
     *
     * @param properties the TypeScript properties of a ``PredefinedMetricSpecificationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('predefinedMetricType', cdk.requiredValidator)(properties.predefinedMetricType));
        errors.collect(cdk.propertyValidator('predefinedMetricType', cdk.validateString)(properties.predefinedMetricType));
        errors.collect(cdk.propertyValidator('resourceLabel', cdk.validateString)(properties.resourceLabel));
        return errors.wrap('supplied properties not correct for "PredefinedMetricSpecificationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.PredefinedMetricSpecification`` resource
     *
     * @param properties the TypeScript properties of a ``PredefinedMetricSpecificationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.PredefinedMetricSpecification`` resource.
     */
    function scalingPolicyResourcePredefinedMetricSpecificationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator(properties).assertSuccess();
        return {
            PredefinedMetricType: cdk.stringToCloudFormation(properties.predefinedMetricType),
            ResourceLabel: cdk.stringToCloudFormation(properties.resourceLabel),
        };
    }
    /**
     * Determine whether the given properties match those of a ``StepAdjustmentProperty``
     *
     * @param properties the TypeScript properties of a ``StepAdjustmentProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_StepAdjustmentPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('metricIntervalLowerBound', cdk.validateNumber)(properties.metricIntervalLowerBound));
        errors.collect(cdk.propertyValidator('metricIntervalUpperBound', cdk.validateNumber)(properties.metricIntervalUpperBound));
        errors.collect(cdk.propertyValidator('scalingAdjustment', cdk.requiredValidator)(properties.scalingAdjustment));
        errors.collect(cdk.propertyValidator('scalingAdjustment', cdk.validateNumber)(properties.scalingAdjustment));
        return errors.wrap('supplied properties not correct for "StepAdjustmentProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.StepAdjustment`` resource
     *
     * @param properties the TypeScript properties of a ``StepAdjustmentProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.StepAdjustment`` resource.
     */
    function scalingPolicyResourceStepAdjustmentPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_StepAdjustmentPropertyValidator(properties).assertSuccess();
        return {
            MetricIntervalLowerBound: cdk.numberToCloudFormation(properties.metricIntervalLowerBound),
            MetricIntervalUpperBound: cdk.numberToCloudFormation(properties.metricIntervalUpperBound),
            ScalingAdjustment: cdk.numberToCloudFormation(properties.scalingAdjustment),
        };
    }
    /**
     * Determine whether the given properties match those of a ``TargetTrackingConfigurationProperty``
     *
     * @param properties the TypeScript properties of a ``TargetTrackingConfigurationProperty``
     *
     * @returns the result of the validation.
     */
    function ScalingPolicyResource_TargetTrackingConfigurationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('customizedMetricSpecification', ScalingPolicyResource_CustomizedMetricSpecificationPropertyValidator)(properties.customizedMetricSpecification));
        errors.collect(cdk.propertyValidator('disableScaleIn', cdk.validateBoolean)(properties.disableScaleIn));
        errors.collect(cdk.propertyValidator('predefinedMetricSpecification', ScalingPolicyResource_PredefinedMetricSpecificationPropertyValidator)(properties.predefinedMetricSpecification));
        errors.collect(cdk.propertyValidator('targetValue', cdk.requiredValidator)(properties.targetValue));
        errors.collect(cdk.propertyValidator('targetValue', cdk.validateNumber)(properties.targetValue));
        return errors.wrap('supplied properties not correct for "TargetTrackingConfigurationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.TargetTrackingConfiguration`` resource
     *
     * @param properties the TypeScript properties of a ``TargetTrackingConfigurationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScalingPolicy.TargetTrackingConfiguration`` resource.
     */
    function scalingPolicyResourceTargetTrackingConfigurationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScalingPolicyResource_TargetTrackingConfigurationPropertyValidator(properties).assertSuccess();
        return {
            CustomizedMetricSpecification: scalingPolicyResourceCustomizedMetricSpecificationPropertyToCloudFormation(properties.customizedMetricSpecification),
            DisableScaleIn: cdk.booleanToCloudFormation(properties.disableScaleIn),
            PredefinedMetricSpecification: scalingPolicyResourcePredefinedMetricSpecificationPropertyToCloudFormation(properties.predefinedMetricSpecification),
            TargetValue: cdk.numberToCloudFormation(properties.targetValue),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``ScheduledActionResourceProps``
     *
     * @param properties the TypeScript properties of a ``ScheduledActionResourceProps``
     *
     * @returns the result of the validation.
     */
    function ScheduledActionResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.requiredValidator)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('autoScalingGroupName', cdk.validateString)(properties.autoScalingGroupName));
        errors.collect(cdk.propertyValidator('desiredCapacity', cdk.validateNumber)(properties.desiredCapacity));
        errors.collect(cdk.propertyValidator('endTime', cdk.validateString)(properties.endTime));
        errors.collect(cdk.propertyValidator('maxSize', cdk.validateNumber)(properties.maxSize));
        errors.collect(cdk.propertyValidator('minSize', cdk.validateNumber)(properties.minSize));
        errors.collect(cdk.propertyValidator('recurrence', cdk.validateString)(properties.recurrence));
        errors.collect(cdk.propertyValidator('startTime', cdk.validateString)(properties.startTime));
        return errors.wrap('supplied properties not correct for "ScheduledActionResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::AutoScaling::ScheduledAction`` resource
     *
     * @param properties the TypeScript properties of a ``ScheduledActionResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::AutoScaling::ScheduledAction`` resource.
     */
    function scheduledActionResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        ScheduledActionResourcePropsValidator(properties).assertSuccess();
        return {
            AutoScalingGroupName: cdk.stringToCloudFormation(properties.autoScalingGroupName),
            DesiredCapacity: cdk.numberToCloudFormation(properties.desiredCapacity),
            EndTime: cdk.stringToCloudFormation(properties.endTime),
            MaxSize: cdk.numberToCloudFormation(properties.maxSize),
            MinSize: cdk.numberToCloudFormation(properties.minSize),
            Recurrence: cdk.stringToCloudFormation(properties.recurrence),
            StartTime: cdk.stringToCloudFormation(properties.startTime),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-as-scheduledaction.html
     */
    class ScheduledActionResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::AutoScaling::ScheduledAction``.
         *
         * @param parent   the ``cdk.Construct`` this ``ScheduledActionResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``ScheduledActionResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: ScheduledActionResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'autoScalingGroupName', this);
            this.scheduledActionName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return scheduledActionResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    ScheduledActionResource.resourceTypeName = "AWS::AutoScaling::ScheduledAction";
    cloudformation.ScheduledActionResource = ScheduledActionResource;
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
