// @aws-cdk/aws-s3/test/notification-dests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iam = require("@aws-cdk/aws-iam");
const s3notifications = require("@aws-cdk/aws-s3-notifications");
const cdk = require("@aws-cdk/cdk");
/**
 * Since we can't take a dependency on @aws-cdk/sns, this is a simple wrapper
 * for AWS::SNS::Topic which implements IBucketNotificationDestination.
 */
class Topic extends cdk.Construct {
    constructor(parent, id) {
        super(parent, id);
        this.policy = new iam.PolicyDocument();
        this.notifyingBucketPaths = new Set();
        const resource = new cdk.Resource(this, 'Resource', { type: 'AWS::SNS::Topic' });
        const topicArn = resource.ref;
        new cdk.Resource(this, 'Policy', {
            type: 'AWS::SNS::TopicPolicy',
            properties: {
                Topics: [topicArn],
                PolicyDocument: this.policy
            }
        });
        this.topicArn = topicArn;
    }
    asBucketNotificationDestination(bucketArn, bucketId) {
        // add permission to each source bucket
        if (!this.notifyingBucketPaths.has(bucketId)) {
            this.policy.addStatement(new iam.PolicyStatement()
                .describe(`sid${this.policy.statementCount}`)
                .addServicePrincipal('s3.amazonaws.com')
                .addAction('sns:Publish')
                .addResource(this.topicArn)
                .addCondition('ArnLike', { "aws:SourceArn": bucketArn }));
            this.notifyingBucketPaths.add(bucketId);
        }
        return {
            arn: this.topicArn,
            type: s3notifications.BucketNotificationDestinationType.Topic
        };
    }
}
exports.Topic = Topic;
