// @aws-cdk/aws-lambda/lib/pipeline-action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codepipeline = require("@aws-cdk/aws-codepipeline-api");
const iam = require("@aws-cdk/aws-iam");
/**
 * CodePipeline invoke Action that is provided by an AWS Lambda function.
 *
 * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/actions-invoke-lambda-function.html
 */
class PipelineInvokeAction extends codepipeline.Action {
    constructor(parent, name, props) {
        super(parent, name, {
            stage: props.stage,
            runOrder: props.runOrder,
            category: codepipeline.ActionCategory.Invoke,
            provider: 'Lambda',
            artifactBounds: codepipeline.defaultBounds(),
            configuration: {
                FunctionName: props.lambda.functionName,
                UserParameters: props.userParameters
            }
        });
        // allow pipeline to list functions
        props.stage.pipeline.role.addToPolicy(new iam.PolicyStatement()
            .addAction('lambda:ListFunctions')
            .addAllResources());
        // allow pipeline to invoke this lambda functionn
        props.stage.pipeline.role.addToPolicy(new iam.PolicyStatement()
            .addAction('lambda:InvokeFunction')
            .addResource(props.lambda.functionArn));
        // allow lambda to put job results for this pipeline.
        const addToPolicy = props.addPutJobResultPolicy !== undefined ? props.addPutJobResultPolicy : true;
        if (addToPolicy) {
            props.lambda.addToRolePolicy(new iam.PolicyStatement()
                .addAllResources() // to avoid cycles (see docs)
                .addAction('codepipeline:PutJobSuccessResult')
                .addAction('codepipeline:PutJobFailureResult'));
        }
    }
}
exports.PipelineInvokeAction = PipelineInvokeAction;
