// @aws-cdk/aws-lambda/lib/runtime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RuntimeFamily;
(function (RuntimeFamily) {
    RuntimeFamily[RuntimeFamily["NodeJS"] = 0] = "NodeJS";
    RuntimeFamily[RuntimeFamily["Java"] = 1] = "Java";
    RuntimeFamily[RuntimeFamily["Python"] = 2] = "Python";
    RuntimeFamily[RuntimeFamily["DotNetCore"] = 3] = "DotNetCore";
    RuntimeFamily[RuntimeFamily["Go"] = 4] = "Go";
    RuntimeFamily[RuntimeFamily["Ruby"] = 5] = "Ruby";
})(RuntimeFamily = exports.RuntimeFamily || (exports.RuntimeFamily = {}));
/**
 * Lambda function runtime environment.
 *
 * If you need to use a runtime name that doesn't exist as a static member, you
 * can instantiate a `Runtime` object, e.g: `new Runtime('nodejs99.99')`.
 */
class Runtime {
    constructor(name, family, props = {}) {
        this.name = name;
        this.supportsInlineCode = !!props.supportsInlineCode;
        this.family = family;
    }
    toString() {
        return this.name;
    }
}
Runtime.NodeJS = new Runtime('nodejs', RuntimeFamily.NodeJS, { supportsInlineCode: true });
Runtime.NodeJS43 = new Runtime('nodejs4.3', RuntimeFamily.NodeJS, { supportsInlineCode: true });
Runtime.NodeJS610 = new Runtime('nodejs6.10', RuntimeFamily.NodeJS, { supportsInlineCode: true });
Runtime.NodeJS810 = new Runtime('nodejs8.10', RuntimeFamily.NodeJS, { supportsInlineCode: true });
Runtime.Python27 = new Runtime('python2.7', RuntimeFamily.Python, { supportsInlineCode: true });
Runtime.Python36 = new Runtime('python3.6', RuntimeFamily.Python, { supportsInlineCode: true });
Runtime.Java8 = new Runtime('java8', RuntimeFamily.Java);
Runtime.DotNetCore1 = new Runtime('dotnetcore1.0', RuntimeFamily.DotNetCore);
Runtime.DotNetCore2 = new Runtime('dotnetcore2.0', RuntimeFamily.DotNetCore);
Runtime.DotNetCore21 = new Runtime('dotnetcore2.1', RuntimeFamily.DotNetCore);
Runtime.Go1x = new Runtime('go1.x', RuntimeFamily.Go);
Runtime.Ruby25 = new Runtime('ruby2.5', RuntimeFamily.Ruby, { supportsInlineCode: true });
exports.Runtime = Runtime;
