// @aws-cdk/aws-route53/lib/hosted-zone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const hosted_zone_ref_1 = require("./hosted-zone-ref");
const route53_generated_1 = require("./route53.generated");
const util_1 = require("./util");
/**
 * Create a Route53 public hosted zone.
 */
class PublicHostedZone extends hosted_zone_ref_1.HostedZoneRef {
    constructor(parent, name, props) {
        super(parent, name);
        util_1.validateZoneName(props.zoneName);
        const hostedZone = new route53_generated_1.cloudformation.HostedZoneResource(this, 'Resource', Object.assign({}, determineHostedZoneProps(props)));
        this.hostedZoneId = hostedZone.ref;
        this.nameServers = hostedZone.hostedZoneNameServers;
        this.zoneName = props.zoneName;
    }
}
exports.PublicHostedZone = PublicHostedZone;
/**
 * Create a Route53 private hosted zone for use in one or more VPCs.
 *
 * Note that `enableDnsHostnames` and `enableDnsSupport` must have been enabled
 * for the VPC you're configuring for private hosted zones.
 */
class PrivateHostedZone extends hosted_zone_ref_1.HostedZoneRef {
    constructor(parent, name, props) {
        super(parent, name);
        /**
         * VPCs to which this hosted zone will be added
         */
        this.vpcs = [];
        util_1.validateZoneName(props.zoneName);
        const hostedZone = new route53_generated_1.cloudformation.HostedZoneResource(this, 'Resource', Object.assign({ vpcs: new cdk.Token(() => this.vpcs ? this.vpcs : undefined) }, determineHostedZoneProps(props)));
        this.hostedZoneId = hostedZone.ref;
        this.zoneName = props.zoneName;
        this.addVpc(props.vpc);
    }
    /**
     * Add another VPC to this private hosted zone.
     *
     * @param vpc the other VPC to add.
     */
    addVpc(vpc) {
        this.vpcs.push(toVpcProperty(vpc));
    }
}
exports.PrivateHostedZone = PrivateHostedZone;
function toVpcProperty(vpc) {
    return { vpcId: vpc.vpcId, vpcRegion: new cdk.AwsRegion() };
}
function determineHostedZoneProps(props) {
    const name = props.zoneName + '.';
    const hostedZoneConfig = props.comment ? { comment: props.comment } : undefined;
    const queryLoggingConfig = props.queryLogsLogGroupArn ? { cloudWatchLogsLogGroupArn: props.queryLogsLogGroupArn } : undefined;
    return { name, hostedZoneConfig, queryLoggingConfig };
}
