// @aws-cdk/aws-route53/lib/route53.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:04:18.573Z","fingerprint":"wlQacuzujA677hS8EBgWX3Wbc9WBbz1fl52uULAscpM="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``HealthCheckResourceProps``
     *
     * @param properties the TypeScript properties of a ``HealthCheckResourceProps``
     *
     * @returns the result of the validation.
     */
    function HealthCheckResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('healthCheckConfig', cdk.requiredValidator)(properties.healthCheckConfig));
        errors.collect(cdk.propertyValidator('healthCheckConfig', HealthCheckResource_HealthCheckConfigPropertyValidator)(properties.healthCheckConfig));
        errors.collect(cdk.propertyValidator('healthCheckTags', cdk.listValidator(HealthCheckResource_HealthCheckTagPropertyValidator))(properties.healthCheckTags));
        return errors.wrap('supplied properties not correct for "HealthCheckResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck`` resource
     *
     * @param properties the TypeScript properties of a ``HealthCheckResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck`` resource.
     */
    function healthCheckResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HealthCheckResourcePropsValidator(properties).assertSuccess();
        return {
            HealthCheckConfig: healthCheckResourceHealthCheckConfigPropertyToCloudFormation(properties.healthCheckConfig),
            HealthCheckTags: cdk.listMapper(healthCheckResourceHealthCheckTagPropertyToCloudFormation)(properties.healthCheckTags),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html
     */
    class HealthCheckResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::Route53::HealthCheck``.
         *
         * @param parent   the ``cdk.Construct`` this ``HealthCheckResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``HealthCheckResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: HealthCheckResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'healthCheckConfig', this);
            this.healthCheckId = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return healthCheckResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    HealthCheckResource.resourceTypeName = "AWS::Route53::HealthCheck";
    cloudformation.HealthCheckResource = HealthCheckResource;
    /**
     * Determine whether the given properties match those of a ``AlarmIdentifierProperty``
     *
     * @param properties the TypeScript properties of a ``AlarmIdentifierProperty``
     *
     * @returns the result of the validation.
     */
    function HealthCheckResource_AlarmIdentifierPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('region', cdk.requiredValidator)(properties.region));
        errors.collect(cdk.propertyValidator('region', cdk.validateString)(properties.region));
        return errors.wrap('supplied properties not correct for "AlarmIdentifierProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.AlarmIdentifier`` resource
     *
     * @param properties the TypeScript properties of a ``AlarmIdentifierProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.AlarmIdentifier`` resource.
     */
    function healthCheckResourceAlarmIdentifierPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HealthCheckResource_AlarmIdentifierPropertyValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            Region: cdk.stringToCloudFormation(properties.region),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HealthCheckConfigProperty``
     *
     * @param properties the TypeScript properties of a ``HealthCheckConfigProperty``
     *
     * @returns the result of the validation.
     */
    function HealthCheckResource_HealthCheckConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('alarmIdentifier', HealthCheckResource_AlarmIdentifierPropertyValidator)(properties.alarmIdentifier));
        errors.collect(cdk.propertyValidator('childHealthChecks', cdk.listValidator(cdk.validateString))(properties.childHealthChecks));
        errors.collect(cdk.propertyValidator('enableSni', cdk.validateBoolean)(properties.enableSni));
        errors.collect(cdk.propertyValidator('failureThreshold', cdk.validateNumber)(properties.failureThreshold));
        errors.collect(cdk.propertyValidator('fullyQualifiedDomainName', cdk.validateString)(properties.fullyQualifiedDomainName));
        errors.collect(cdk.propertyValidator('healthThreshold', cdk.validateNumber)(properties.healthThreshold));
        errors.collect(cdk.propertyValidator('ipAddress', cdk.validateString)(properties.ipAddress));
        errors.collect(cdk.propertyValidator('insufficientDataHealthStatus', cdk.validateString)(properties.insufficientDataHealthStatus));
        errors.collect(cdk.propertyValidator('inverted', cdk.validateBoolean)(properties.inverted));
        errors.collect(cdk.propertyValidator('measureLatency', cdk.validateBoolean)(properties.measureLatency));
        errors.collect(cdk.propertyValidator('port', cdk.validateNumber)(properties.port));
        errors.collect(cdk.propertyValidator('regions', cdk.listValidator(cdk.validateString))(properties.regions));
        errors.collect(cdk.propertyValidator('requestInterval', cdk.validateNumber)(properties.requestInterval));
        errors.collect(cdk.propertyValidator('resourcePath', cdk.validateString)(properties.resourcePath));
        errors.collect(cdk.propertyValidator('searchString', cdk.validateString)(properties.searchString));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        return errors.wrap('supplied properties not correct for "HealthCheckConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.HealthCheckConfig`` resource
     *
     * @param properties the TypeScript properties of a ``HealthCheckConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.HealthCheckConfig`` resource.
     */
    function healthCheckResourceHealthCheckConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HealthCheckResource_HealthCheckConfigPropertyValidator(properties).assertSuccess();
        return {
            AlarmIdentifier: healthCheckResourceAlarmIdentifierPropertyToCloudFormation(properties.alarmIdentifier),
            ChildHealthChecks: cdk.listMapper(cdk.stringToCloudFormation)(properties.childHealthChecks),
            EnableSNI: cdk.booleanToCloudFormation(properties.enableSni),
            FailureThreshold: cdk.numberToCloudFormation(properties.failureThreshold),
            FullyQualifiedDomainName: cdk.stringToCloudFormation(properties.fullyQualifiedDomainName),
            HealthThreshold: cdk.numberToCloudFormation(properties.healthThreshold),
            IPAddress: cdk.stringToCloudFormation(properties.ipAddress),
            InsufficientDataHealthStatus: cdk.stringToCloudFormation(properties.insufficientDataHealthStatus),
            Inverted: cdk.booleanToCloudFormation(properties.inverted),
            MeasureLatency: cdk.booleanToCloudFormation(properties.measureLatency),
            Port: cdk.numberToCloudFormation(properties.port),
            Regions: cdk.listMapper(cdk.stringToCloudFormation)(properties.regions),
            RequestInterval: cdk.numberToCloudFormation(properties.requestInterval),
            ResourcePath: cdk.stringToCloudFormation(properties.resourcePath),
            SearchString: cdk.stringToCloudFormation(properties.searchString),
            Type: cdk.stringToCloudFormation(properties.type),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HealthCheckTagProperty``
     *
     * @param properties the TypeScript properties of a ``HealthCheckTagProperty``
     *
     * @returns the result of the validation.
     */
    function HealthCheckResource_HealthCheckTagPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('key', cdk.requiredValidator)(properties.key));
        errors.collect(cdk.propertyValidator('key', cdk.validateString)(properties.key));
        errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "HealthCheckTagProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.HealthCheckTag`` resource
     *
     * @param properties the TypeScript properties of a ``HealthCheckTagProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HealthCheck.HealthCheckTag`` resource.
     */
    function healthCheckResourceHealthCheckTagPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HealthCheckResource_HealthCheckTagPropertyValidator(properties).assertSuccess();
        return {
            Key: cdk.stringToCloudFormation(properties.key),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``HostedZoneResourceProps``
     *
     * @param properties the TypeScript properties of a ``HostedZoneResourceProps``
     *
     * @returns the result of the validation.
     */
    function HostedZoneResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('hostedZoneConfig', HostedZoneResource_HostedZoneConfigPropertyValidator)(properties.hostedZoneConfig));
        errors.collect(cdk.propertyValidator('hostedZoneTags', cdk.listValidator(HostedZoneResource_HostedZoneTagPropertyValidator))(properties.hostedZoneTags));
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('queryLoggingConfig', HostedZoneResource_QueryLoggingConfigPropertyValidator)(properties.queryLoggingConfig));
        errors.collect(cdk.propertyValidator('vpcs', cdk.listValidator(HostedZoneResource_VPCPropertyValidator))(properties.vpcs));
        return errors.wrap('supplied properties not correct for "HostedZoneResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HostedZone`` resource
     *
     * @param properties the TypeScript properties of a ``HostedZoneResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HostedZone`` resource.
     */
    function hostedZoneResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HostedZoneResourcePropsValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            HostedZoneConfig: hostedZoneResourceHostedZoneConfigPropertyToCloudFormation(properties.hostedZoneConfig),
            HostedZoneTags: cdk.listMapper(hostedZoneResourceHostedZoneTagPropertyToCloudFormation)(properties.hostedZoneTags),
            QueryLoggingConfig: hostedZoneResourceQueryLoggingConfigPropertyToCloudFormation(properties.queryLoggingConfig),
            VPCs: cdk.listMapper(hostedZoneResourceVPCPropertyToCloudFormation)(properties.vpcs),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-route53-hostedzone.html
     */
    class HostedZoneResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::Route53::HostedZone``.
         *
         * @param parent   the ``cdk.Construct`` this ``HostedZoneResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``HostedZoneResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: HostedZoneResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'name', this);
            this.hostedZoneNameServers = new HostedZoneNameServers(this.getAtt('NameServers'));
            this.hostedZoneId = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return hostedZoneResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    HostedZoneResource.resourceTypeName = "AWS::Route53::HostedZone";
    cloudformation.HostedZoneResource = HostedZoneResource;
    /**
     * Determine whether the given properties match those of a ``HostedZoneConfigProperty``
     *
     * @param properties the TypeScript properties of a ``HostedZoneConfigProperty``
     *
     * @returns the result of the validation.
     */
    function HostedZoneResource_HostedZoneConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('comment', cdk.validateString)(properties.comment));
        return errors.wrap('supplied properties not correct for "HostedZoneConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.HostedZoneConfig`` resource
     *
     * @param properties the TypeScript properties of a ``HostedZoneConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.HostedZoneConfig`` resource.
     */
    function hostedZoneResourceHostedZoneConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HostedZoneResource_HostedZoneConfigPropertyValidator(properties).assertSuccess();
        return {
            Comment: cdk.stringToCloudFormation(properties.comment),
        };
    }
    /**
     * Determine whether the given properties match those of a ``HostedZoneTagProperty``
     *
     * @param properties the TypeScript properties of a ``HostedZoneTagProperty``
     *
     * @returns the result of the validation.
     */
    function HostedZoneResource_HostedZoneTagPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('key', cdk.requiredValidator)(properties.key));
        errors.collect(cdk.propertyValidator('key', cdk.validateString)(properties.key));
        errors.collect(cdk.propertyValidator('value', cdk.requiredValidator)(properties.value));
        errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
        return errors.wrap('supplied properties not correct for "HostedZoneTagProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.HostedZoneTag`` resource
     *
     * @param properties the TypeScript properties of a ``HostedZoneTagProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.HostedZoneTag`` resource.
     */
    function hostedZoneResourceHostedZoneTagPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HostedZoneResource_HostedZoneTagPropertyValidator(properties).assertSuccess();
        return {
            Key: cdk.stringToCloudFormation(properties.key),
            Value: cdk.stringToCloudFormation(properties.value),
        };
    }
    /**
     * Determine whether the given properties match those of a ``QueryLoggingConfigProperty``
     *
     * @param properties the TypeScript properties of a ``QueryLoggingConfigProperty``
     *
     * @returns the result of the validation.
     */
    function HostedZoneResource_QueryLoggingConfigPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('cloudWatchLogsLogGroupArn', cdk.requiredValidator)(properties.cloudWatchLogsLogGroupArn));
        errors.collect(cdk.propertyValidator('cloudWatchLogsLogGroupArn', cdk.validateString)(properties.cloudWatchLogsLogGroupArn));
        return errors.wrap('supplied properties not correct for "QueryLoggingConfigProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.QueryLoggingConfig`` resource
     *
     * @param properties the TypeScript properties of a ``QueryLoggingConfigProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.QueryLoggingConfig`` resource.
     */
    function hostedZoneResourceQueryLoggingConfigPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HostedZoneResource_QueryLoggingConfigPropertyValidator(properties).assertSuccess();
        return {
            CloudWatchLogsLogGroupArn: cdk.stringToCloudFormation(properties.cloudWatchLogsLogGroupArn),
        };
    }
    /**
     * Determine whether the given properties match those of a ``VPCProperty``
     *
     * @param properties the TypeScript properties of a ``VPCProperty``
     *
     * @returns the result of the validation.
     */
    function HostedZoneResource_VPCPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('vpcId', cdk.requiredValidator)(properties.vpcId));
        errors.collect(cdk.propertyValidator('vpcId', cdk.validateString)(properties.vpcId));
        errors.collect(cdk.propertyValidator('vpcRegion', cdk.requiredValidator)(properties.vpcRegion));
        errors.collect(cdk.propertyValidator('vpcRegion', cdk.validateString)(properties.vpcRegion));
        return errors.wrap('supplied properties not correct for "VPCProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.VPC`` resource
     *
     * @param properties the TypeScript properties of a ``VPCProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::HostedZone.VPC`` resource.
     */
    function hostedZoneResourceVPCPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        HostedZoneResource_VPCPropertyValidator(properties).assertSuccess();
        return {
            VPCId: cdk.stringToCloudFormation(properties.vpcId),
            VPCRegion: cdk.stringToCloudFormation(properties.vpcRegion),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
class HostedZoneNameServers extends cdk.Token {
    constructor() {
        super(...arguments);
        // @ts-ignore: private but unused on purpose.
        this.thisIsAHostedZoneNameServers = true;
    }
}
exports.HostedZoneNameServers = HostedZoneNameServers;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``RecordSetResourceProps``
     *
     * @param properties the TypeScript properties of a ``RecordSetResourceProps``
     *
     * @returns the result of the validation.
     */
    function RecordSetResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('aliasTarget', RecordSetResource_AliasTargetPropertyValidator)(properties.aliasTarget));
        errors.collect(cdk.propertyValidator('comment', cdk.validateString)(properties.comment));
        errors.collect(cdk.propertyValidator('failover', cdk.validateString)(properties.failover));
        errors.collect(cdk.propertyValidator('geoLocation', RecordSetResource_GeoLocationPropertyValidator)(properties.geoLocation));
        errors.collect(cdk.propertyValidator('healthCheckId', cdk.validateString)(properties.healthCheckId));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.validateString)(properties.hostedZoneId));
        errors.collect(cdk.propertyValidator('hostedZoneName', cdk.validateString)(properties.hostedZoneName));
        errors.collect(cdk.propertyValidator('multiValueAnswer', cdk.validateBoolean)(properties.multiValueAnswer));
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('region', cdk.validateString)(properties.region));
        errors.collect(cdk.propertyValidator('resourceRecords', cdk.listValidator(cdk.validateString))(properties.resourceRecords));
        errors.collect(cdk.propertyValidator('setIdentifier', cdk.validateString)(properties.setIdentifier));
        errors.collect(cdk.propertyValidator('ttl', cdk.validateString)(properties.ttl));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        errors.collect(cdk.propertyValidator('weight', cdk.validateNumber)(properties.weight));
        return errors.wrap('supplied properties not correct for "RecordSetResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSet`` resource
     *
     * @param properties the TypeScript properties of a ``RecordSetResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSet`` resource.
     */
    function recordSetResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetResourcePropsValidator(properties).assertSuccess();
        return {
            Name: cdk.stringToCloudFormation(properties.name),
            Type: cdk.stringToCloudFormation(properties.type),
            AliasTarget: recordSetResourceAliasTargetPropertyToCloudFormation(properties.aliasTarget),
            Comment: cdk.stringToCloudFormation(properties.comment),
            Failover: cdk.stringToCloudFormation(properties.failover),
            GeoLocation: recordSetResourceGeoLocationPropertyToCloudFormation(properties.geoLocation),
            HealthCheckId: cdk.stringToCloudFormation(properties.healthCheckId),
            HostedZoneId: cdk.stringToCloudFormation(properties.hostedZoneId),
            HostedZoneName: cdk.stringToCloudFormation(properties.hostedZoneName),
            MultiValueAnswer: cdk.booleanToCloudFormation(properties.multiValueAnswer),
            Region: cdk.stringToCloudFormation(properties.region),
            ResourceRecords: cdk.listMapper(cdk.stringToCloudFormation)(properties.resourceRecords),
            SetIdentifier: cdk.stringToCloudFormation(properties.setIdentifier),
            TTL: cdk.stringToCloudFormation(properties.ttl),
            Weight: cdk.numberToCloudFormation(properties.weight),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-recordset.html
     */
    class RecordSetResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::Route53::RecordSet``.
         *
         * @param parent   the ``cdk.Construct`` this ``RecordSetResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``RecordSetResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: RecordSetResource.resourceTypeName, properties });
            cdk.requireProperty(properties, 'name', this);
            cdk.requireProperty(properties, 'type', this);
            this.recordSetDomainName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return recordSetResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    RecordSetResource.resourceTypeName = "AWS::Route53::RecordSet";
    cloudformation.RecordSetResource = RecordSetResource;
    /**
     * Determine whether the given properties match those of a ``AliasTargetProperty``
     *
     * @param properties the TypeScript properties of a ``AliasTargetProperty``
     *
     * @returns the result of the validation.
     */
    function RecordSetResource_AliasTargetPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('dnsName', cdk.requiredValidator)(properties.dnsName));
        errors.collect(cdk.propertyValidator('dnsName', cdk.validateString)(properties.dnsName));
        errors.collect(cdk.propertyValidator('evaluateTargetHealth', cdk.validateBoolean)(properties.evaluateTargetHealth));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.requiredValidator)(properties.hostedZoneId));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.validateString)(properties.hostedZoneId));
        return errors.wrap('supplied properties not correct for "AliasTargetProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSet.AliasTarget`` resource
     *
     * @param properties the TypeScript properties of a ``AliasTargetProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSet.AliasTarget`` resource.
     */
    function recordSetResourceAliasTargetPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetResource_AliasTargetPropertyValidator(properties).assertSuccess();
        return {
            DNSName: cdk.stringToCloudFormation(properties.dnsName),
            EvaluateTargetHealth: cdk.booleanToCloudFormation(properties.evaluateTargetHealth),
            HostedZoneId: cdk.stringToCloudFormation(properties.hostedZoneId),
        };
    }
    /**
     * Determine whether the given properties match those of a ``GeoLocationProperty``
     *
     * @param properties the TypeScript properties of a ``GeoLocationProperty``
     *
     * @returns the result of the validation.
     */
    function RecordSetResource_GeoLocationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('continentCode', cdk.validateString)(properties.continentCode));
        errors.collect(cdk.propertyValidator('countryCode', cdk.validateString)(properties.countryCode));
        errors.collect(cdk.propertyValidator('subdivisionCode', cdk.validateString)(properties.subdivisionCode));
        return errors.wrap('supplied properties not correct for "GeoLocationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSet.GeoLocation`` resource
     *
     * @param properties the TypeScript properties of a ``GeoLocationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSet.GeoLocation`` resource.
     */
    function recordSetResourceGeoLocationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetResource_GeoLocationPropertyValidator(properties).assertSuccess();
        return {
            ContinentCode: cdk.stringToCloudFormation(properties.continentCode),
            CountryCode: cdk.stringToCloudFormation(properties.countryCode),
            SubdivisionCode: cdk.stringToCloudFormation(properties.subdivisionCode),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``RecordSetGroupResourceProps``
     *
     * @param properties the TypeScript properties of a ``RecordSetGroupResourceProps``
     *
     * @returns the result of the validation.
     */
    function RecordSetGroupResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('comment', cdk.validateString)(properties.comment));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.validateString)(properties.hostedZoneId));
        errors.collect(cdk.propertyValidator('hostedZoneName', cdk.validateString)(properties.hostedZoneName));
        errors.collect(cdk.propertyValidator('recordSets', cdk.listValidator(RecordSetGroupResource_RecordSetPropertyValidator))(properties.recordSets));
        return errors.wrap('supplied properties not correct for "RecordSetGroupResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup`` resource
     *
     * @param properties the TypeScript properties of a ``RecordSetGroupResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup`` resource.
     */
    function recordSetGroupResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetGroupResourcePropsValidator(properties).assertSuccess();
        return {
            Comment: cdk.stringToCloudFormation(properties.comment),
            HostedZoneId: cdk.stringToCloudFormation(properties.hostedZoneId),
            HostedZoneName: cdk.stringToCloudFormation(properties.hostedZoneName),
            RecordSets: cdk.listMapper(recordSetGroupResourceRecordSetPropertyToCloudFormation)(properties.recordSets),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-route53-recordsetgroup.html
     */
    class RecordSetGroupResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::Route53::RecordSetGroup``.
         *
         * @param parent   the ``cdk.Construct`` this ``RecordSetGroupResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``RecordSetGroupResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: RecordSetGroupResource.resourceTypeName, properties });
            this.recordSetGroupName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return recordSetGroupResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    RecordSetGroupResource.resourceTypeName = "AWS::Route53::RecordSetGroup";
    cloudformation.RecordSetGroupResource = RecordSetGroupResource;
    /**
     * Determine whether the given properties match those of a ``AliasTargetProperty``
     *
     * @param properties the TypeScript properties of a ``AliasTargetProperty``
     *
     * @returns the result of the validation.
     */
    function RecordSetGroupResource_AliasTargetPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('dnsName', cdk.requiredValidator)(properties.dnsName));
        errors.collect(cdk.propertyValidator('dnsName', cdk.validateString)(properties.dnsName));
        errors.collect(cdk.propertyValidator('evaluateTargetHealth', cdk.validateBoolean)(properties.evaluateTargetHealth));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.requiredValidator)(properties.hostedZoneId));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.validateString)(properties.hostedZoneId));
        return errors.wrap('supplied properties not correct for "AliasTargetProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.AliasTarget`` resource
     *
     * @param properties the TypeScript properties of a ``AliasTargetProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.AliasTarget`` resource.
     */
    function recordSetGroupResourceAliasTargetPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetGroupResource_AliasTargetPropertyValidator(properties).assertSuccess();
        return {
            DNSName: cdk.stringToCloudFormation(properties.dnsName),
            EvaluateTargetHealth: cdk.booleanToCloudFormation(properties.evaluateTargetHealth),
            HostedZoneId: cdk.stringToCloudFormation(properties.hostedZoneId),
        };
    }
    /**
     * Determine whether the given properties match those of a ``GeoLocationProperty``
     *
     * @param properties the TypeScript properties of a ``GeoLocationProperty``
     *
     * @returns the result of the validation.
     */
    function RecordSetGroupResource_GeoLocationPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('continentCode', cdk.validateString)(properties.continentCode));
        errors.collect(cdk.propertyValidator('countryCode', cdk.validateString)(properties.countryCode));
        errors.collect(cdk.propertyValidator('subdivisionCode', cdk.validateString)(properties.subdivisionCode));
        return errors.wrap('supplied properties not correct for "GeoLocationProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.GeoLocation`` resource
     *
     * @param properties the TypeScript properties of a ``GeoLocationProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.GeoLocation`` resource.
     */
    function recordSetGroupResourceGeoLocationPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetGroupResource_GeoLocationPropertyValidator(properties).assertSuccess();
        return {
            ContinentCode: cdk.stringToCloudFormation(properties.continentCode),
            CountryCode: cdk.stringToCloudFormation(properties.countryCode),
            SubdivisionCode: cdk.stringToCloudFormation(properties.subdivisionCode),
        };
    }
    /**
     * Determine whether the given properties match those of a ``RecordSetProperty``
     *
     * @param properties the TypeScript properties of a ``RecordSetProperty``
     *
     * @returns the result of the validation.
     */
    function RecordSetGroupResource_RecordSetPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('aliasTarget', RecordSetGroupResource_AliasTargetPropertyValidator)(properties.aliasTarget));
        errors.collect(cdk.propertyValidator('comment', cdk.validateString)(properties.comment));
        errors.collect(cdk.propertyValidator('failover', cdk.validateString)(properties.failover));
        errors.collect(cdk.propertyValidator('geoLocation', RecordSetGroupResource_GeoLocationPropertyValidator)(properties.geoLocation));
        errors.collect(cdk.propertyValidator('healthCheckId', cdk.validateString)(properties.healthCheckId));
        errors.collect(cdk.propertyValidator('hostedZoneId', cdk.validateString)(properties.hostedZoneId));
        errors.collect(cdk.propertyValidator('hostedZoneName', cdk.validateString)(properties.hostedZoneName));
        errors.collect(cdk.propertyValidator('multiValueAnswer', cdk.validateBoolean)(properties.multiValueAnswer));
        errors.collect(cdk.propertyValidator('name', cdk.requiredValidator)(properties.name));
        errors.collect(cdk.propertyValidator('name', cdk.validateString)(properties.name));
        errors.collect(cdk.propertyValidator('region', cdk.validateString)(properties.region));
        errors.collect(cdk.propertyValidator('resourceRecords', cdk.listValidator(cdk.validateString))(properties.resourceRecords));
        errors.collect(cdk.propertyValidator('setIdentifier', cdk.validateString)(properties.setIdentifier));
        errors.collect(cdk.propertyValidator('ttl', cdk.validateString)(properties.ttl));
        errors.collect(cdk.propertyValidator('type', cdk.requiredValidator)(properties.type));
        errors.collect(cdk.propertyValidator('type', cdk.validateString)(properties.type));
        errors.collect(cdk.propertyValidator('weight', cdk.validateNumber)(properties.weight));
        return errors.wrap('supplied properties not correct for "RecordSetProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.RecordSet`` resource
     *
     * @param properties the TypeScript properties of a ``RecordSetProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::Route53::RecordSetGroup.RecordSet`` resource.
     */
    function recordSetGroupResourceRecordSetPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RecordSetGroupResource_RecordSetPropertyValidator(properties).assertSuccess();
        return {
            AliasTarget: recordSetGroupResourceAliasTargetPropertyToCloudFormation(properties.aliasTarget),
            Comment: cdk.stringToCloudFormation(properties.comment),
            Failover: cdk.stringToCloudFormation(properties.failover),
            GeoLocation: recordSetGroupResourceGeoLocationPropertyToCloudFormation(properties.geoLocation),
            HealthCheckId: cdk.stringToCloudFormation(properties.healthCheckId),
            HostedZoneId: cdk.stringToCloudFormation(properties.hostedZoneId),
            HostedZoneName: cdk.stringToCloudFormation(properties.hostedZoneName),
            MultiValueAnswer: cdk.booleanToCloudFormation(properties.multiValueAnswer),
            Name: cdk.stringToCloudFormation(properties.name),
            Region: cdk.stringToCloudFormation(properties.region),
            ResourceRecords: cdk.listMapper(cdk.stringToCloudFormation)(properties.resourceRecords),
            SetIdentifier: cdk.stringToCloudFormation(properties.setIdentifier),
            TTL: cdk.stringToCloudFormation(properties.ttl),
            Type: cdk.stringToCloudFormation(properties.type),
            Weight: cdk.numberToCloudFormation(properties.weight),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
