// @aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-target-group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_target_group_1 = require("../shared/base-target-group");
const enums_1 = require("../shared/enums");
const imported_1 = require("../shared/imported");
const util_1 = require("../shared/util");
/**
 * Define a Network Target Group
 */
class NetworkTargetGroup extends base_target_group_1.BaseTargetGroup {
    /**
     * Import an existing listener
     */
    static import(parent, id, props) {
        return new ImportedNetworkTargetGroup(parent, id, props);
    }
    constructor(parent, id, props) {
        super(parent, id, props, {
            protocol: enums_1.Protocol.Tcp,
            port: props.port,
        });
        this.listeners = [];
        if (props.proxyProtocolV2) {
            this.setAttribute('proxy_protocol_v2.enabled', 'true');
        }
        this.addTarget(...(props.targets || []));
    }
    /**
     * Add a load balancing target to this target group
     */
    addTarget(...targets) {
        for (const target of targets) {
            const result = target.attachToNetworkTargetGroup(this);
            this.addLoadBalancerTarget(result);
        }
    }
    /**
     * Register a listener that is load balancing to this target group.
     *
     * Don't call this directly. It will be called by listeners.
     */
    registerListener(listener) {
        this.loadBalancerAssociationDependencies.push(listener);
        this.listeners.push(listener);
    }
    /**
     * Full name of first load balancer
     */
    get firstLoadBalancerFullName() {
        if (this.listeners.length === 0) {
            throw new Error('The TargetGroup needs to be attached to a LoadBalancer before you can call this method');
        }
        return base_target_group_1.loadBalancerNameFromListenerArn(this.listeners[0].listenerArn);
    }
}
exports.NetworkTargetGroup = NetworkTargetGroup;
/**
 * An imported network target group
 */
class ImportedNetworkTargetGroup extends imported_1.BaseImportedTargetGroup {
    registerListener(_listener) {
        // Nothing to do, we know nothing of our members
    }
    loadBalancerDependency() {
        return new util_1.LazyDependable([]);
    }
}
