// @aws-cdk/aws-ecr/lib/ecr.generated.js
"use strict";
// Copyright 2012-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// Generated from the AWS CloudFormation Resource Specification
// See: docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html
// @cfn2ts:meta@ {"generated":"2018-12-13T10:04:18.656Z","fingerprint":"w5Zd3tgFYdOSUYv9fe0V0uqtgIS1fw8f91GF3XlYJ5w="}
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:max-line-length | This is generated code - line lengths are difficult to control
const cdk = require("@aws-cdk/cdk");
var cloudformation;
(function (cloudformation) {
    /**
     * Determine whether the given properties match those of a ``RepositoryResourceProps``
     *
     * @param properties the TypeScript properties of a ``RepositoryResourceProps``
     *
     * @returns the result of the validation.
     */
    function RepositoryResourcePropsValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('lifecyclePolicy', RepositoryResource_LifecyclePolicyPropertyValidator)(properties.lifecyclePolicy));
        errors.collect(cdk.propertyValidator('repositoryName', cdk.validateString)(properties.repositoryName));
        errors.collect(cdk.propertyValidator('repositoryPolicyText', cdk.validateObject)(properties.repositoryPolicyText));
        return errors.wrap('supplied properties not correct for "RepositoryResourceProps"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECR::Repository`` resource
     *
     * @param properties the TypeScript properties of a ``RepositoryResourceProps``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECR::Repository`` resource.
     */
    function repositoryResourcePropsToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RepositoryResourcePropsValidator(properties).assertSuccess();
        return {
            LifecyclePolicy: repositoryResourceLifecyclePolicyPropertyToCloudFormation(properties.lifecyclePolicy),
            RepositoryName: cdk.stringToCloudFormation(properties.repositoryName),
            RepositoryPolicyText: cdk.objectToCloudFormation(properties.repositoryPolicyText),
        };
    }
    /**
     * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-repository.html
     */
    class RepositoryResource extends cdk.Resource {
        /**
         * Creates a new ``AWS::ECR::Repository``.
         *
         * @param parent   the ``cdk.Construct`` this ``RepositoryResource`` is a part of
         * @param name     the name of the resource in the ``cdk.Construct`` tree
         * @param properties the properties of this ``RepositoryResource``
         */
        constructor(parent, name, properties) {
            super(parent, name, { type: RepositoryResource.resourceTypeName, properties });
            this.repositoryArn = this.getAtt('Arn').toString();
            this.repositoryName = this.ref.toString();
        }
        get propertyOverrides() {
            return this.untypedPropertyOverrides;
        }
        renderProperties(properties) {
            return repositoryResourcePropsToCloudFormation(cdk.resolve(properties));
        }
    }
    /**
     * The CloudFormation resource type name for this resource class.
     */
    RepositoryResource.resourceTypeName = "AWS::ECR::Repository";
    cloudformation.RepositoryResource = RepositoryResource;
    /**
     * Determine whether the given properties match those of a ``LifecyclePolicyProperty``
     *
     * @param properties the TypeScript properties of a ``LifecyclePolicyProperty``
     *
     * @returns the result of the validation.
     */
    function RepositoryResource_LifecyclePolicyPropertyValidator(properties) {
        if (!cdk.canInspect(properties)) {
            return cdk.VALIDATION_SUCCESS;
        }
        const errors = new cdk.ValidationResults();
        errors.collect(cdk.propertyValidator('lifecyclePolicyText', cdk.validateString)(properties.lifecyclePolicyText));
        errors.collect(cdk.propertyValidator('registryId', cdk.validateString)(properties.registryId));
        return errors.wrap('supplied properties not correct for "LifecyclePolicyProperty"');
    }
    /**
     * Renders the AWS CloudFormation properties of an ``AWS::ECR::Repository.LifecyclePolicy`` resource
     *
     * @param properties the TypeScript properties of a ``LifecyclePolicyProperty``
     *
     * @returns the AWS CloudFormation properties of an ``AWS::ECR::Repository.LifecyclePolicy`` resource.
     */
    function repositoryResourceLifecyclePolicyPropertyToCloudFormation(properties) {
        if (!cdk.canInspect(properties)) {
            return properties;
        }
        RepositoryResource_LifecyclePolicyPropertyValidator(properties).assertSuccess();
        return {
            LifecyclePolicyText: cdk.stringToCloudFormation(properties.lifecyclePolicyText),
            RegistryId: cdk.stringToCloudFormation(properties.registryId),
        };
    }
})(cloudformation = exports.cloudformation || (exports.cloudformation = {}));
